// uu_plus4u5filesg01/dist-esm/forms/autocomplete-items.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import Config from "../config/config.js";
import Css from "../helpers/css.js";

const ClassNames = {};

const ClassDefs = {
  popover: `
    width: 100%;
    max-height: 350px;
    overflow-y: auto;
  `,
  link: `
    && {
      color: rgba(0,0,0, .87);
      display: block;
      padding: 8px 16px;
      width: 100%;
  
      &:hover {
        text-decoration: none;
        background-color: #E0E0E0;
      }
    }
  `
};

const getClassName = className => {
  if (!ClassNames[className]) {
    ClassNames[className] = Css.css(ClassDefs[className]);
  }

  return ClassNames[className];
};

const AutocompleteItems = props => {
  // no items to shown => render null
  if (!props.shown || !Array.isArray(props.items) || !props.items.length) return null;

  let items = props.items;

  // filter items by filterValue
  if (props.filterValue) {
    items = items.filter(item => item.indexOf(props.filterValue) !== -1);
  }

  // filter items by filterItems
  if (Array.isArray(props.filterItems)) {
    items = items.filter(item => props.filterItems.indexOf(item) === -1);
  }

  // validate if is something to show
  if (items.length === 0) {
    return null;
  }

  return (
    <UU5.Bricks.Popover shown fitHeightToViewport forceRender className={getClassName("popover")}>
      {items.map(item => (
        <UU5.Bricks.Link
          className={getClassName("link")}
          colorSchema="default"
          key={item}
          mainAttrs={{
            onMouseDown: e => {
              e.preventDefault();
              props.onClick(item);
            }
          }}
        >
          {item}
        </UU5.Bricks.Link>
      ))}
    </UU5.Bricks.Popover>
  );
};

AutocompleteItems.displayName = Config.TAG + "AutocompleteItems";
AutocompleteItems.tagName = Config.TAG + "AutocompleteItems";

AutocompleteItems.propTypes = {
  items: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
  filterValue: UU5.PropTypes.string,
  filterItems: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
  shown: UU5.PropTypes.bool,
  onClick: UU5.PropTypes.func
};

AutocompleteItems.defaultProps = {
  filterValue: "",
  shown: false
};

export default AutocompleteItems;
