// uu_plus4u5filesg01/dist-esm/forms/preview-file-form.js
import * as UU5 from "uu5g04";
import Calls from "../calls.js";
import Config from "../config/config.js";

import Lsi from "../config/lsi.js";
import LsiHelper from "../helpers/lsi-helper.js";

import "./preview-file-form.less";

const PreviewFileForm = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Config.TAG + "PreviewFileForm",
    classNames: {
      main: Config.CSS + "preview-file-form"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    code: UU5.PropTypes.string,
    done: UU5.PropTypes.func,
    onCancel: UU5.PropTypes.func,
    rev: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.number]),
    baseUri: UU5.PropTypes.string,
    cmdGetBinaryData: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    if (this.props.cmdGetBinaryData && this.props.code) {
      this._updateImageSrc();
    }
    return {
      error: false,
      src: null
    };
  },

  componentWillReceiveProps(nextProps) {
    if (nextProps.cmdGetBinaryData !== this.props.cmdGetBinaryData || nextProps.code !== this.props.code) {
      this._updateImageSrc(nextProps);
    }
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  getForm() {
    return this._form;
  },

  hasUnsavedChanges() {
    return false;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _updateImageSrcDone(data) {
    this.setState(state => {
      if (state.src && state.src.match(/^blob:/)) {
        // release old image url
        window.URL.revokeObjectURL(state.src);
      }
      // create new url to blob
      return { src: data.match(/^blob:/) ? window.Url.createObjectURL(data) : data };
    });
  },

  _updateImageSrc(props = this.props) {
    let src;
    let data = { code: props.code };
    if (typeof props.cmdGetBinaryData === "function") {
      props.cmdGetBinaryData(data).then(this._updateImageSrcDone, this._updateImageSrcDone);
    } else {
      src = Calls.getCommandUri(props.baseUri, props.cmdGetBinaryData, data);
      src = src + "&" + props.rev;
      this._updateImageSrcDone(src);
    }
  },

  _onSave() {
    this.props.done && this.props.done();
  },

  _renderImage() {
    return (
      <UU5.Bricks.Image
        mainAttrs={{ onError: () => this.setState({ error: true }) }}
        responsive
        src={this.state.src}
        authenticate
      />
    );
  },

  _renderError() {
    return <UU5.Bricks.Div content={LsiHelper.getLsiComponent(Lsi.LSI_LABEL_FILE_MANAGER_PREVIEW_ERROR)} />;
  },

  _getChild() {
    return (
      <UU5.Bricks.Div {...this.getMainPropsToPass()}>
        {this.state.error ? this._renderError() : this._renderImage()}
      </UU5.Bricks.Div>
    );
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this._getChild();
  }
  //@@viewOff:render
});

export default PreviewFileForm;
