// uu_plus4u5g01/dist-esm/app/internal/about-delete-form.js
import React from "react";
import createReactClass from "create-react-class";
import * as UU5 from "uu5g04";
import * as Plus4U5 from "uu_plus4u5g01";

import Cfg from "../config/config.js";

import "./about-delete-form.less";

export const AboutDeleteForm = createReactClass({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.TAG + "AboutDeleteForm",
    classNames: {
      main: Cfg.CSS + " " + Cfg.CSS + "about-delete-form",
      content: Cfg.CSS + "about-delete-form-content"
    },
    opt: {
      nestingLevelWrapper: true
    },
    lsi: () => Plus4U5.Environment.Lsi.App.authors
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {},
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  setSubmitDisabled(b) {
    this.setState({ submitDisabled: b });
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:private
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <UU5.Forms.Form {...this.getMainPropsToPass()} onSave={this.props.onSave} onCancel={this.props.onCancel}>
        <UU5.Bricks.Div content={this.props.content} className={this.getClassName("content")} />
        <UU5.Forms.Controls
          disabled={this.state.submitDisabled} // TODO set only to button submit when UU5 allow setting it only to single button
          buttonCancelProps={{
            content: this.getLsiComponent("buttonCancel")
          }}
          buttonSubmitProps={{
            colorSchema: "danger",
            content: this.getLsiComponent("buttonDelete")
          }}
        />
      </UU5.Forms.Form>
    );
  }
  //@@viewOff:render
});

export default AboutDeleteForm;
