// uu_plus4u5g01/dist-esm/app/internal/authors.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import Author from "./author.js";

import "./authors.less";

export const Authors = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Internal.Authors",
    classNames: {
      main: "plus4u5-app-authors",
      header: "plus4u5-app-authors-header",
      authors: "plus4u5-app-authors-authors"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    leadingAuthors: UU5.PropTypes.arrayOf(
      UU5.PropTypes.shape({
        name: UU5.PropTypes.string.isRequired,
        content: UU5.PropTypes.string,
        uuIdentity: UU5.PropTypes.string,
        signature: UU5.PropTypes.string,
        role: UU5.PropTypes.any.isRequired
      })
    ),
    otherAuthors: UU5.PropTypes.arrayOf(
      UU5.PropTypes.shape({
        name: UU5.PropTypes.string,
        content: UU5.PropTypes.string,
        uuIdentity: UU5.PropTypes.string,
        role: UU5.PropTypes.any
      })
    ),
    header: UU5.PropTypes.any,
    underline: UU5.PropTypes.bool,
    level: UU5.PropTypes.oneOf(["0", "1", "2", "3", "4", "5", "6", 0, 1, 2, 3, 4, 5, 6])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      leadingAuthors: [],
      otherAuthors: null,
      header: null,
      underline: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _getAuthors(authors, size) {
    return authors.map((author, i) => <Author key={i} {...author} size={size} />);
  },
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Section
        {...this.getMainPropsToPass()}
        header={this.props.header}
        underline={this.props.underline}
        content={null}
        level={this.props.level}
      >
        {this.props.leadingAuthors && (
          <UU5.Bricks.Div
            className={this.getClassName().authors}
            content={this._getAuthors(this.props.leadingAuthors, "m")}
          />
        )}
        {this.props.otherAuthors && (
          <UU5.Bricks.Div
            className={this.getClassName().authors}
            content={this._getAuthors(this.props.otherAuthors, "s")}
          />
        )}
      </UU5.Bricks.Section>
    );
  }
  //@@viewOff:render
});

export default Authors;
