// uu_plus4u5g01/dist-esm/app/internal/author.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";
import "uu_plus4u5g01-bricks";

export const Author = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Author",
    classNames: {
      main: "plus4u5-app-author",
      size: "plus4u5-app-author-size",
      image: "plus4u5-app-author-image",
      signature: "plus4u5-app-author-signature",
      name: "plus4u5-app-author-name",
      position: "plus4u5-app-author-position",
      content: "plus4u5-app-author-content"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    signature: UU5.PropTypes.string,
    src: UU5.PropTypes.string,
    uuIdentity: UU5.PropTypes.string,
    role: UU5.PropTypes.any,
    size: UU5.PropTypes.oneOf(["s", "m"])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      signature: null,
      src: null,
      uuIdentity: null,
      role: null,
      size: "m"
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _getMainAttrs() {
    const attrs = this.getMainPropsToPass();
    attrs.content = null;
    attrs.className += ` ${this.getClassName().size}-${this.props.content ? "l" : this.props.size}`;
    return attrs;
  },
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Column {...this._getMainAttrs()}>
        <Plus4U5.Bricks.UserPhoto
          className={this.getClassName().image}
          type="circle"
          width={null}
          uuIdentity={this.props.uuIdentity || (!this.props.src ? "0-0" : null)}
          src={this.props.src}
          shadow
        />

        {this.props.signature && (
          <UU5.Bricks.Image className={this.getClassName().signature} src={this.props.signature} />
        )}

        <UU5.Bricks.Span className={this.getClassName().name} content={this.props.name} />
        <UU5.Bricks.Span className={this.getClassName().position} content={this.props.role} />

        {this.props.content && <UU5.Bricks.P className={this.getClassName().content} content={this.props.content} />}
      </UU5.Bricks.Column>
    );
  }
  //@@viewOff:render
});

export default Author;
