// uu_plus4u5g01/dist-esm/app/internal/executive-panel-float.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import Cfg from "../config/config.js";
import Css from "./css.js";

import MenuPanelList from "../menu-panel-list.js";
import MenuPanelContextMenu from "./menu-panel-context-menu.js";

const itemPropTypes = {
  icon: UU5.PropTypes.string,
  type: UU5.PropTypes.oneOf(["text", "checkbox"]),
  onClick: UU5.PropTypes.func,
  items: UU5.PropTypes.arrayOf(UU5.PropTypes.object),
  content: UU5.PropTypes.any
};

export const ExecutivePanelFloat = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.TAG + "ExecutivePanelFloat",
    classNames: {
      main: () => {
        let classNames = [Cfg.CSS + "executive-panel"];
        return classNames.join(" ");
      },
      header: () => Css.css("text-align: center;line-height: 30px;margin-right: 3px;"),
      menuSettings: () => Css.css("position: absolute;top: 0px;right: 8px;")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    header: UU5.PropTypes.any,
    settings: UU5.PropTypes.arrayOf(UU5.PropTypes.shape(itemPropTypes)),
    items: UU5.PropTypes.arrayOf(UU5.PropTypes.shape(itemPropTypes)),
    position: UU5.PropTypes.string,
    width: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.number])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      header: undefined,
      settings: undefined,
      items: undefined,
      width: 184,
      position: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _registerCtxMenu(ref) {
    this._ctxMenu = ref;
  },

  _onClickDropdown(component, event) {
    event.stopPropagation();
    event.preventDefault();

    let openSettings = {
      aroundElement: component,
      position: "bottom-right"
    };

    this._ctxMenu.isOpen() ? this._ctxMenu.close() : this._ctxMenu.open(openSettings);
  },

  _getContextMenu() {
    return <MenuPanelContextMenu ref_={this._registerCtxMenu} items={this.props.settings} />;
  },

  _getHeader() {
    let result = ( 
      <div className={this.getClassName("header")}>
        {this.props.header}
        <div className={this.getClassName("menuSettings")}>
          { 
            this.props.settings ? 
                <UU5.Bricks.Button
                className={this.getClassName("settingsMenuButton")}
                onClick={this._onClickDropdown}
                bgStyle="transparent"
                colorSchema="default"
                size="s"
              >
                <UU5.Bricks.Icon icon="mdi-dots-vertical" />
              </UU5.Bricks.Button>
            : null
          }
        </div>
      </div>
    );

    

    return result;
  },
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.FloatingBox
        {...this.getMainPropsToPass()}
        header={this._getHeader()}
        width={this.props.width}
        position={this.props.position}
        borderRadius={this.props.borderRadius}
        elevation={this.props.elevation}
      >
        {
          this.props.settings ? this._getContextMenu() : null
        }
        <MenuPanelList items={this.props.items} />
      </UU5.Bricks.FloatingBox>
    );
  }
  //@@viewOff:render
});

export default ExecutivePanelFloat;
