// uu_plus4u5g01/dist-esm/app/internal/executive-panel-vertical.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import Cfg from "../config/config.js";

import MenuPanel from "../menu-panel.js";
import MenuPanelTree from "../menu-panel-tree.js";

const itemPropTypes = {
  icon: UU5.PropTypes.string,
  type: UU5.PropTypes.oneOf(["text", "checkbox"]),
  onClick: UU5.PropTypes.func,
  items: UU5.PropTypes.arrayOf(UU5.PropTypes.object),
  content: UU5.PropTypes.any
};

export const ExecutivePanelVertical = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.TAG + "ExecutivePanelVertical",
    classNames: {
      main: Cfg.CSS + "executive-panel-vertical"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    settings: UU5.PropTypes.arrayOf(UU5.PropTypes.shape(itemPropTypes)),
    items: UU5.PropTypes.arrayOf(UU5.PropTypes.shape(itemPropTypes)),
    borderRadius: UU5.PropTypes.string,
    elevation: UU5.PropTypes.oneOf(["0", "1", "2", "3", "4", "5", 0, 1, 2, 3, 4, 5]),
    header: UU5.PropTypes.any,
    expanded: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      settings: undefined,
      items: undefined,
      borderRadius: undefined,
      elevation: undefined,
      header: undefined,
      expanded: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:private
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <MenuPanel
        {...this.getMainPropsToPass()}
        settings={this.props.settings}
        header={this.props.header}
        borderRadius={this.props.borderRadius}
        elevation={this.props.elevation}
        backgroundColorSchema="grey"
        expanded={this.props.expanded}
      >
        <MenuPanelTree items={this.props.items} content={this.props.content || this.props.children} />
      </MenuPanel>
    );
  }
  //@@viewOff:render
});

export default ExecutivePanelVertical;
