// uu5tilesg01/dist-node/config/lsi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var Lsi = {
  actionBar: {
    searchInactive: {
      en: "Inactive ...",
      cs: "Neaktivní ..."
    },
    searchPlaceholder: {
      en: "Search text ...",
      cs: "Hledat text ..."
    },
    searchFilterLabel: {
      en: "Search",
      cs: "Hledat"
    }
  },
  bulkActionBar: {
    showSelection: {
      en: "Show Selection",
      cs: "Zobrazit výběr"
    },
    showAll: {
      en: "Show All",
      cs: "Zobrazit vše"
    },
    title: {
      en: "Bulk Actions",
      cs: "Hromadné akce"
    }
  },
  filterBar: {
    title: {
      en: "Filters",
      cs: "Filtry"
    },
    addFilter: {
      en: "Add Filter",
      cs: "Přidat filtr"
    },
    cancelFilters: {
      en: "Cancel Filters",
      cs: "Zrušit filtry"
    },
    showSelectedLabel: {
      en: "Showing all selected items ",
      cs: "Zobrazení všech označených položek "
    },
    showSelectedCancel: {
      en: "Cancel",
      cs: "Zrušit zobrazení"
    }
  },
  sorterBar: {
    title: {
      en: "Advanced Sort",
      cs: "Pokročilé řazení"
    },
    ascending: {
      en: "Ascending",
      cs: "vzestupně"
    },
    descending: {
      en: "Descending",
      cs: "sestupně"
    },
    addSorter: {
      en: "Add Sort",
      cs: "Přidat řazení"
    },
    cancelAdvancedSort: {
      en: "Cancel Advanced Sort",
      cs: "Zrušit pokročilé řazení"
    }
  },
  infoBar: {
    unsorted: {
      en: "Unsorted",
      cs: "Neseřazeno"
    },
    advancedSort: {
      en: "Advanced Sort",
      cs: "Pokročilé řazení"
    },
    ascending: {
      en: "Ascending",
      cs: "vzestupně"
    },
    descending: {
      en: "Descending",
      cs: "sestupně"
    },
    counter: {
      en: "Showing %d of %d",
      cs: "Zobrazeno %d z %d"
    },
    select: {
      en: "Select All",
      cs: "Označ vše"
    },
    deselect: {
      en: "Cancel Selection",
      cs: "Zrušit označení"
    },
    selectionDropdown: {
      en: "Selection",
      cs: "Označení"
    }
  },
  list: {},
  listController: {},
  simpleFilterPanel: {
    filterTypeLabel: {
      en: "Select filter type",
      cs: "Vyber typ filtru"
    },
    filterTypePlaceholder: {
      en: "Select filter",
      cs: "Vyber filtr"
    },
    filterValueLabel: {
      en: "And its value",
      cs: "A jeho hodnotu"
    },
    filterValuePlaceholder: {
      en: "Filter value",
      cs: "Hodnota filtru"
    },
    addButton: {
      en: "Add",
      cs: "Přidat"
    }
  },
  tile: {}
};
var _default = Lsi;
exports.default = _default;
