// uu_contentkitg01/dist-esm/editable/bg-style-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../core/_config.js";
import Lsi from "../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Editable.BgStyleInput")
};
//@@viewOff:statics

const BgStyleInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool,
    label: UU5.PropTypes.object,
    items: UU5.PropTypes.array
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false,
    label: Lsi.bgStyle,
    items: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Forms.SwitchSelector
        {...UU5.Common.VisualComponent.getAttrs(props)}
        label={props.label}
        value={props.componentProps.bgStyle}
        onChange={opt => props.onChangeProps({ bgStyle: opt.value })}
        items={
          props.items || [
            { value: "filled", content: Lsi.filled },
            { value: "outline", content: Lsi.outline },
            { value: "transparent", content: Lsi.transparent },
            { value: "underline", content: Lsi.underline }
          ]
        }
      />
    );
    //@@viewOff:render
  }
});

export default BgStyleInput;
