// uu_contentkitg01/dist-esm/editable/border-radius-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../core/_config.js";
import Lsi from "../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Editable.BorderRadiusInput")
};
//@@viewOff:statics

const BorderRadiusInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool,
    items: UU5.PropTypes.array
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false,
    items: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Forms.SwitchSelector
        {...UU5.Common.VisualComponent.getAttrs(props)}
        label={Lsi.borderRadius}
        value={UU5.Common.Tools.fillUnit(props.componentProps.borderRadius)}
        onChange={opt => props.onChangeProps({ borderRadius: opt.value })}
        items={
          props.items || [
            { value: "0", content: Lsi.none },
            { value: "2px" },
            { value: "4px" },
            { value: "8px" },
            { value: "16px" },
            { value: "50%", content: Lsi.rounded }
          ]
        }
      />
    );
    //@@viewOff:render
  }
});

export default BorderRadiusInput;
