// uu_contentkitg01/dist-esm/editable/elevation-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../core/_config.js";
import Lsi from "../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Editable.ElevationInput")
};
//@@viewOff:statics

const ElevationInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool,
    propName: UU5.PropTypes.string,
    label: UU5.PropTypes.object
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false,
    propName: "elevation",
    label: Lsi.elevation
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Forms.SwitchSelector
        {...UU5.Common.VisualComponent.getAttrs(props)}
        label={props.label}
        value={props.componentProps[props.propName] + ""}
        onChange={opt => props.onChangeProps({ [props.propName]: +opt.value })}
        items={[
          { value: "-1" },
          { value: "0" },
          { value: "1" },
          { value: "2" },
          { value: "3" },
          { value: "4" },
          { value: "5" }
        ]}
      />
    );
    //@@viewOff:render
  }
});

export default ElevationInput;
