// uu_contentkitg01/dist-esm/editable/info-message.js
//@@viewOn:imports
import UU5 from "uu5g04";
import "uu5g04-bricks";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../core/_config.js";
import Css from "./css";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Editable.InfoMessage"),
  classNames: {
    message: () => Css.css`
      font-size: 14px;
      font-style: italic;
      padding-top: 6px;
      border-radius: 0.3em;
      color: rgba(0, 0, 0, 0.34);
    `
  }
};
//@@viewOff:statics

const InfoMessage = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    message: UU5.PropTypes.object,
    disabled: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    message: null,
    disabled: false
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Bricks.Text
        {...UU5.Common.VisualComponent.getAttrs(props, STATICS.classNames.message())}
        content={props.message}
      />
    );
    //@@viewOff:render
  }
});

export default InfoMessage;
