// uu_contentkitg01/dist-esm/editable/number-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../core/_config.js";

import InfoMessage from "./info-message.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.NumberInput")
};
//@@viewOff:statics

const NumberInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool,
    message: UU5.PropTypes.object,
    propName: UU5.PropTypes.string.isRequired,
    numberProps: UU5.PropTypes.object
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false,
    message: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <>
        <UU5.Forms.Number
          {...UU5.Common.VisualComponent.getAttrs(props)}
          {...props.numberProps}
          value={props.componentProps[props.propName]}
          onBlur={({ value }) => props.onChangeProps({ [props.propName]: !value ? null : +value })}
        />
        {props.message && <InfoMessage message={props.message} disabled={props.disabled} />}
      </>
    );
    //@@viewOff:render
  }
});

export default NumberInput;
