// uu_contentkitg01/dist-esm/editable/width-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../core/_config.js";
import Lsi from "../config/lsi.js";

import InfoMessage from "./info-message.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Editable.WidthInput")
};
//@@viewOff:statics

const WidthInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool,
    inputColWidth: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false,
    inputColWidth: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <>
        <UU5.Forms.Text
          {...UU5.Common.VisualComponent.getAttrs(props)}
          label={Lsi.width}
          value={props.componentProps.width}
          onChange={opt => props.onChangeProps({ width: opt.value })}
          disabled={props.disabled}
          inputColWidth={props.inputColWidth}
        />
        <InfoMessage message={Lsi.help.units} disabled={props.disabled} />
      </>
    );
    //@@viewOff:render
  }
});

export default WidthInput;
