// uu5richtextg01/dist-esm/richtext/plugins/uu5/blockquote.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

import EditorBlockLazy from "../../editor-block-lazy.js";

const PLUGIN_NAME = "blockquote";

// cannot use "blockquote" as a block name, draft use it too. It causes render uu5blockquote inside draft blockquote
const BLOCK_NAME = "BLOCKQUOTE";

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "block",
    buttonDefs: {
      name: "blockquote",
      icon: "mdi-format-quote-open",
      label: () => UU5.Environment.Lsi.RichText.plugins.blockquote.label,
      type: "block",
      value: BLOCK_NAME
    },
    toEditorItemDefs: {
      name: BLOCK_NAME,
      tagName: "UU5.Bricks.Blockquote",
      htmlAliases: {
        tagName: "blockquote"
      },
      component: properties => {
        return (
          <UU5.Bricks.Blockquote {...properties.blockProps}>
            <EditorBlockLazy {...properties} />
          </UU5.Bricks.Blockquote>
        );
      },
      element: "blockquote"
    }
  })
);

export default pluginDef;
