// uu5richtextg01/dist-esm/richtext/plugins/uu5/error.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

import EditorBlockLazy from "../../editor-block-lazy.js";

import "./error.less";

const ErrorPluginDef = Immutable.Map(
  Immutable.fromJS({
    name: "error",
    opts: {
      blockMap: {
        error: {
          component: properties => {
            const dataProps = properties.blockProps;
            return (
              <UU5.Bricks.Div>
                <UU5.Bricks.Code {...dataProps} className="uu5-richtext-editor-error uu5-richtext-editor-error-block">
                  <EditorBlockLazy {...properties} />
                </UU5.Bricks.Code>
              </UU5.Bricks.Div>
            );
          },
          opt: { preserveEnters: true }
        }
      },
      entityMap: {
        error: {
          component: properties => {
            const dataProps = properties.contentState.getEntity(properties.entityKey).getData();
            return (
              <UU5.Bricks.Code {...dataProps} className="uu5-richtext-editor-error">
                {properties.children}
              </UU5.Bricks.Code>
            );
          },
          opt: { preserveEnters: true }
        }
      }
    }
  })
);

export default ErrorPluginDef;
