// uu5richtextg01/dist-esm/richtext/plugins/uu5/font-family.js
import UU5 from "uu5g04";
import Immutable from "immutable";

const PLUGIN_NAME = "fontFamily";

const DEFAULT_FONT = "ClearSans";
const FONTS = ["ClearSans", "Arial", "Courier", "Times New Roman"];

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "inline",
    buttonDefs: [
      {
        name: "fontFamily",
        icon: "mdi-format-text",
        label: () => UU5.Environment.Lsi.RichText.plugins.fontFamily.label,
        preserveStyle: true,
        items: FONTS.map(font => {
          return {
            label: font,
            value: font === DEFAULT_FONT ? "" : `fontFamily:${font}`
          };
        })
      }
    ],
    toEditorItemDefs: {
      tagName: "UU5.Bricks.Span",
      htmlAliases: {
        tagName: "span",
        style: { fontFamily: true }
      },
      style: { fontFamily: true },
      prefix: PLUGIN_NAME
    },
    toObjectDefs: { prefix: PLUGIN_NAME, tagName: "UU5.Bricks.Span", style: { fontFamily: "${1}" } }
  })
);

export default pluginDef;
