// uu5richtextg01/dist-esm/richtext/plugins/uu5/index.js
import UU5 from "uu5g04";
import Immutable from "immutable";

const PLUGIN_NAME = "index";

const SUB_STYLE_NAME = PLUGIN_NAME + ".sub";
const SUPER_STYLE_NAME = PLUGIN_NAME + ".super";

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "inline",
    buttonDefs: [
      {
        name: "sub",
        icon: "mdi-format-subscript",
        label: () => UU5.Environment.Lsi.RichText.plugins.index.subscriptLabel,
        value: SUB_STYLE_NAME
      },
      {
        name: "sup",
        icon: "mdi-format-superscript",
        label: () => UU5.Environment.Lsi.RichText.plugins.index.superscriptLabel,
        value: SUPER_STYLE_NAME
      }
    ],
    toEditorItemDefs: [
      {
        name: SUB_STYLE_NAME,
        tagName: "UU5.Bricks.Span",
        htmlAliases: [
          { tagName: "sub" },
          { tagName: "span", style: { verticalAlign: "sub", fontSize: "smaller" } },
          { tagName: "span", style: { verticalAlign: "sub", fontSize: "0.8em" } }
        ],
        style: { verticalAlign: "sub", fontSize: "0.8em" },
        styleFn: styleObj => {
          if (styleObj.fontSize) {
            let parts = styleObj.fontSize.match(/([0-9.]+)(.*)/);
            if (parts) {
              return { fontSize: parseFloat(parts[1]) * 0.8 + parts[2], verticalAlign: "sub" };
            }
          }
          return { fontSize: "0.8em", verticalAlign: "sub" };
        }
      },
      {
        name: SUB_STYLE_NAME,
        tagName: "UU5.Bricks.Span",
        style: { verticalAlign: "sub", fontSize: "smaller" },
        skipToObjectRule: true
      },
      {
        name: SUPER_STYLE_NAME,
        tagName: "UU5.Bricks.Span",
        htmlAliases: [
          { tagName: "sup" },
          { tagName: "span", style: { verticalAlign: "super", fontSize: "smaller" } },
          { tagName: "span", style: { verticalAlign: "super", fontSize: "0.8em" } }
        ],
        style: { verticalAlign: "super", fontSize: "0.8em" },
        styleFn: styleObj => {
          if (styleObj.fontSize) {
            let parts = styleObj.fontSize.match(/([0-9.]+)(.*)/);
            if (parts) {
              return { fontSize: parseFloat(parts[1]) * 0.8 + parts[2], verticalAlign: "super" };
            }
          }
          return { fontSize: "0.8em", verticalAlign: "super" };
        }
      },
      {
        name: SUPER_STYLE_NAME,
        tagName: "UU5.Bricks.Span",
        style: { verticalAlign: "super", fontSize: "smaller" },
        skipToObjectRule: true
      }
    ]
  })
);

export default pluginDef;
