// uu5richtextg01/dist-esm/richtext/plugins/uu5/italic.js
import UU5 from "uu5g04";
import Immutable from "immutable";

const PLUGIN_NAME = "italic";
const STYLE_NAME = "ITALIC"; // needed to be bold to reuse standard shortcut CTRL + B defined by draft js

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "inline",
    buttonDefs: {
      name: "italic",
      icon: "mdi-format-italic",
      label: () => UU5.Environment.Lsi.RichText.plugins.italic.label,
      value: STYLE_NAME
    },
    toEditorItemDefs: [
      {
        name: STYLE_NAME,
        tagName: "UU5.Bricks.Em",
        htmlAliases: [{ tagName: "i" }, { tagName: "em" }, { tagName: "span", style: { fontStyle: "italic" } }]
      },
      {
        tagName: "UU5.Bricks.Span",
        style: { fontStyle: "italic" },
        name: STYLE_NAME
      },
      {
        name: STYLE_NAME,
        tagName: "i"
      },
      {
        name: STYLE_NAME,
        tagName: "em"
      }
    ],
    toObjectDefs: {
      name: STYLE_NAME,
      type: "inline",
      tagName: "i"
    }
  })
);

export default pluginDef;
