// uu5richtextg01/dist-esm/richtext/plugins/uu5/plugins.js
// inline styles
import BoldPlugin from "./bold.js";
import ColorPlugin from "./color.js";
import FontFamily from "./font-family.js";
import FontSize from "./font-size.js";
import IndexPlugin from "./index.js";
import ItalicPlugin from "./italic.js";
import StrikePlugin from "./strike.js";
import UnderlinePlugin from "./underline.js";
import RemoveFormatPlugin from "./remove-format.js";

// block stlyes
import TextAlignPlugin from "./text-align.js";
import IndentPlugin from "./block-indent.js";

// blocks
// import BlockquotePlugin from "./blockquote.js";
import ListPlugin from "./list.js";
import ParagraphPlugin from "./paragraph.js";
import PPlugin from "./p.js";
import UnstyledPlugin from "./unstyled.js";

// entities
// import ButtonPlugin from "./button.js";
import CodePlugin from "./code.js";
import IconPlugin from "./icon.js";
import LinkPlugin from "./link.js";
import UU5StringPlugin from "./uu5string.js";
import uu5StringEntity from "./uu5string-entity.js";
import uu5StringBlock from "./uu5string-block.js";
import ErrorPlugin from "./error.js";

import WhitespacePlugin from "./whitespace.js";

export default {
  fontSize: FontSize,
  bold: BoldPlugin,
  index: IndexPlugin,
  italic: ItalicPlugin,
  strike: StrikePlugin,
  underline: UnderlinePlugin,
  removeFormat: RemoveFormatPlugin,
  textAlign: TextAlignPlugin,
  unstyled: UnstyledPlugin,
  paragraph: ParagraphPlugin,
  p: PPlugin,
  link: LinkPlugin,
  list: ListPlugin,
  color: ColorPlugin,
  fontFamily: FontFamily,
  indent: IndentPlugin,
  code: CodePlugin,
  uu5string: UU5StringPlugin,
  uu5stringEntity: uu5StringEntity,
  uu5StringBlock: uu5StringBlock,
  icon: IconPlugin,
  error: ErrorPlugin,
  whitespace: WhitespacePlugin
};
