// uu5richtextg01/dist-esm/richtext/plugins/uu5/unstyled.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

import EditorBlockLazy from "../../editor-block-lazy.js";

const PLUGIN_NAME = "unstyled";
const BR_TAG_NAME = "br";

const toObject = function(type, name, props, wrapper, wrapperProps, isEmpty) {
  if (type === "block" && name === PLUGIN_NAME) {
    if (isEmpty) {
      return { tag: this.brTagName, isSelfClosed: true };
    } else {
      return {
        tag: this.tagName,
        props: props
      };
    }
  }
};

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "block",
    toEditorItemDefs: [
      {
        name: PLUGIN_NAME,
        tagName: "UU5.Bricks.Div",
        component: properties => {
          return (
            <UU5.Bricks.Div {...properties.blockProps}>
              <EditorBlockLazy {...properties} />
            </UU5.Bricks.Div>
          );
        },
        element: "div",
        htmlAliases: [{ tagName: "div", className: /uu5-common-div/, style: true, order: 1 }]
      },
      {
        name: PLUGIN_NAME,
        tagName: BR_TAG_NAME
      }
    ],
    toObjectDefs: {
      name: PLUGIN_NAME,
      tagName: "UU5.Bricks.Div",
      brTagName: BR_TAG_NAME,
      fn: toObject
    }
  })
);

export default pluginDef;
