// uu5richtextg01/dist-esm/richtext/plugins/uu5/uu5string-block.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

import { removeSpecialNodes } from "../../bricks/uu5string.js";

const PLUGIN_NAME = "uu5StringBlock";

const toEditorItem = () => {
  return {
    display: "block",
    type: "uu5string",
    data: {},
    getChildren: uu5stringObject => {
      return {
        children: UU5.Common.UU5String.contentToString(uu5stringObject.children.map(child => removeSpecialNodes(child)))
      };
    }
  };
};

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "block",
    toEditorItemDefs: {
      tagName: "UU5.Bricks.uu5StringBlock",
      fn: toEditorItem,
      preserveEnters: true,
      htmlAliases: [
        {
          tagName: "div",
          style: { fontFamily: "richtext-uu5string, monospace" },
          result: { tagName: "UU5.Bricks.uu5StringBlock", style: undefined },
          order: 201
        },
        {
          tagName: "code",
          className: /uu5-richtext-editor-uu5string-block/,
          result: { tagName: "UU5.Bricks.uu5StringBlock", style: undefined },
          order: 201
        }
      ]
    }
  })
);

export default pluginDef;
