// uu5richtextg01/dist-esm/richtext/plugins/uu5/uu5string-entity.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

import { removeSpecialNodes } from "../../bricks/uu5string.js";

const PLUGIN_NAME = "uu5StringEntity";

const toEditorItem = () => {
  return {
    display: "entity",
    type: "uu5string",
    data: {},
    getChildren: uu5stringObject => {
      return {
        children: UU5.Common.UU5String.contentToString(uu5stringObject.children.map(child => removeSpecialNodes(child)))
      };
    }
  };
};

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "entity",
    toEditorItemDefs: {
      tagName: "UU5.RichText.uu5StringEntity",
      fn: toEditorItem,
      preserveEnters: true,
      htmlAliases: [
        {
          tagName: "span",
          style: { fontFamily: "richtext-uu5string, monospace" },
          result: { tagName: "UU5.RichText.uu5StringEntity", style: undefined },
          order: 200
        },
        {
          tagName: "code",
          style: { fontFamily: "richtext-uu5string, monospace" },
          result: { tagName: "UU5.RichText.uu5StringEntity", style: undefined },
          order: 200
        }
      ]
    }
  })
);

export default pluginDef;
