// uu5richtextg01/dist-esm/richtext/plugins/html/error.js
import Immutable from "immutable";

import EditorBlockLazy from "../../editor-block-lazy.js";

import "./error.less";

const ErrorPluginDef = Immutable.Map(
  Immutable.fromJS({
    name: "error",
    opts: {
      blockMap: {
        error: {
          component: properties => {
            const dataProps = properties.blockProps;
            return (
              <div className="uu5-richtext-html-editor-error">
                <code {...dataProps}>
                  <EditorBlockLazy {...properties} />
                </code>
              </div>
            );
          }
        }
      },
      entityMap: {
        error: {
          component: properties => {
            const dataProps = properties.contentState.getEntity(properties.entityKey).getData();
            return (
              <code {...dataProps} className="uu5-richtext-html-editor-error">
                {properties.children}
              </code>
            );
          }
        }
      }
    }
  })
);

export default ErrorPluginDef;
