// uu5richtextg01/dist-esm/richtext/plugins/html/html-fragment-block.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

const PLUGIN_NAME = "htmlFragmentBlock";

const toEditorItem = (tag, props = {}, children, wrapper, wrapperProps) => {
  return {
    display: "block",
    type: "uu5string",
    data: {},
    getChildren: uu5stringObject => {
      return {
        children: UU5.Common.UU5String.contentToString(uu5stringObject.children)
      };
    }
  };
};

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "block",
    toEditorItemDefs: {
      tagName: "UU5.Bricks.htmlFragmentBlock",
      fn: toEditorItem,
      preserveEnters: true,
      htmlAliases: [
        {
          tagName: "div",
          style: { fontFamily: "richtext-html-fragment, monospace" },
          result: { tagName: "UU5.Bricks.htmlFragmentBlock", style: undefined },
          order: 201
        },
        {
          tagName: "code",
          className: /uu5-richtext-html-editor-fragment/,
          result: { tagName: "UU5.Bricks.htmlFragmentBlock", style: undefined },
          order: 201
        }
      ]
    }
  })
);

export default pluginDef;
