// uu5richtextg01/dist-esm/richtext/plugins/html/html-fragment-entity.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

const PLUGIN_NAME = "htmlFragmentEntity";

const toEditorItem = (tag, props = {}, children, wrapper, wrapperProps) => {
  return {
    display: "entity",
    type: "uu5string",
    data: {},
    getChildren: uu5stringObject => {
      return {
        children: UU5.Common.UU5String.contentToString(uu5stringObject.children)
      };
    }
  };
};

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "entity",
    toEditorItemDefs: {
      tagName: "UU5.RichText.htmlFragmentEntity",
      fn: toEditorItem,
      preserveEnters: true,
      htmlAliases: [
        {
          tagName: "span",
          style: { fontFamily: "richtext-html-fragment, monospace" },
          result: { tagName: "UU5.RichText.htmlFragmentEntity", style: undefined },
          order: 200
        },
        {
          tagName: "code",
          style: { fontFamily: "richtext-html-fragment, monospace" },
          result: { tagName: "UU5.RichText.htmlFragmentEntity", style: undefined },
          order: 200
        }
      ]
    }
  })
);

export default pluginDef;
