// uu5richtextg01/dist-esm/richtext/plugins/html/image.js
import * as UU5 from "uu5g04";
import Immutable from "immutable";

import ImageButton from "./internal/image-button.js";

const PLUGIN_NAME = "image";

const pluginDef = Immutable.Map(
  Immutable.fromJS({
    name: PLUGIN_NAME,
    type: "entity",
    buttonDefs: {
      name: "image",
      icon: "mdi-image",
      label: () => UU5.Environment.Lsi.RichText.htmlPlugins.image.label,
      button: ImageButton,
      props: { ruleName: "image" },
      getProps: function() {
        return {
          onUpload: this.onUpload
        };
      }
    },
    toEditorItemDefs: {
      tagName: "img",
      component: properties => {
        const dataProps = properties.contentState.getEntity(properties.entityKey).getData().props || {};
        return (
          <span>
            <img className="my-image" {...dataProps} />
            {properties.children}
          </span>
        );
      },
      fn: (tagName, props) => {
        return {
          type: PLUGIN_NAME,
          display: "entity",
          children: ["\u200b"],
          props: props
        };
      },
      htmlAliases: { tagName: "img", style: true }
    }
  })
);

export default pluginDef;
