// uu5richtextg01/dist-esm/richtext/plugins/html/plugins.js
// inline styles
import BoldPlugin from "./bold.js";
import ColorPlugin from "./color.js";
import FontFamily from "./font-family.js";
import FontSize from "./font-size.js";
import IndexPlugin from "./index.js";
import ItalicPlugin from "./italic.js";
import StrikePlugin from "./strike.js";
import UnderlinePlugin from "./underline.js";
import RemoveFormatPlugin from "../uu5/remove-format.js";
import WhiteSpacePlugin from "../uu5/whitespace.js";

// block stlyes
import TextAlignPlugin from "../uu5/text-align.js";
import IndentPlugin from "../uu5/block-indent.js";

// blocks
import BlockquotePlugin from "./blockquote.js";
import ListPlugin from "./list.js";
import ParagraphPlugin from "./paragraph.js";
import UnstyledPlugin from "./unstyled.js";
import HtmlFragment from "./html-fragment.js";
import HtmlFragmentBlock from "./html-fragment-block.js";
import HtmlFragmentEntity from "./html-fragment-entity.js";

// entities
import CodePlugin from "./code.js";
import ImagePlugin from "./image.js";
import LinkPlugin from "./link.js";

import ErrorPlugin from "./error.js";

export default {
  fontSIze: FontSize,
  bold: BoldPlugin,
  index: IndexPlugin,
  italic: ItalicPlugin,
  strike: StrikePlugin,
  underline: UnderlinePlugin,
  removeFormat: RemoveFormatPlugin,
  textAlign: TextAlignPlugin,
  unstyled: UnstyledPlugin,
  paragraph: ParagraphPlugin,
  link: LinkPlugin,
  list: ListPlugin,
  color: ColorPlugin,
  fontFamily: FontFamily,
  indent: IndentPlugin,
  code: CodePlugin,
  blockquote: BlockquotePlugin,
  image: ImagePlugin,
  htmlFragment: HtmlFragment,
  whitespace: WhiteSpacePlugin,
  htmlFragmentBlock: HtmlFragmentBlock,
  htmlFragmentEntity: HtmlFragmentEntity,
  error: ErrorPlugin
};
