// uu_contentkitg01/dist-esm/core/_config.js
import * as UU5 from "uu5g04";
import Lsi from "../config/lsi";

const Cfg = {
  LIB: "UuContentKit",
  CSS: "uucontentkit",

  tag: function(component) {
    return this.LIB + "." + component;
  },
  css: function(component) {
    return this.CSS + "-" + component;
  },

  LSI_LABEL_FILE_VIEWER_NO_DATA: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Data source not specified.</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Není zadán zdroj dat.</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  LSI_FORM_BUTTON_CANCEL: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Cancel</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Zrušit</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  LSI_FORM_BUTTON_OK: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Confirm</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Potvrdit</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  LSI_LINK_EDITABLE_LABEL: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Name</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Název</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  LSI_LINK_EDITABLE_PAGE: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Page code</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Kód stránky</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  LSI_DCC_TABLE_CONTEXT_MENU_ADD_ROW_BEFORE: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Insert row before</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Vložit řádek před</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),
  LSI_DCC_TABLE_CONTEXT_MENU_ADD_ROW_AFTER: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Insert row after</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Vložit řádek za</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),
  LSI_DCC_TABLE_CONTEXT_MENU_REMOVE_ROW: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Delete row</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Smazat řádek</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),
  LSI_DCC_TABLE_CONTEXT_MENU_ADD_COL_BEFORE: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Insert col before</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Vložit sloupec před</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),
  LSI_DCC_TABLE_CONTEXT_MENU_ADD_COL_AFTER: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Insert col after</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Vložit sloupec za</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),
  LSI_DCC_TABLE_CONTEXT_MENU_REMOVE_COL: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Delete col</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Smazat sloupec</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  LSI_DCC_TABLE_COL_WIDTH: (
    <UU5.Bricks.Lsi>
      <UU5.Bricks.Lsi.Item language="en">Column width</UU5.Bricks.Lsi.Item>
      <UU5.Bricks.Lsi.Item language="cs">Šířky sloupců</UU5.Bricks.Lsi.Item>
    </UU5.Bricks.Lsi>
  ),

  EFFECT_OPTIONS: ["none", "zoom", "grayscale", "blur", "sepia", "invert", "tilt", "morph"],
  EFFECT: [
    { value: "none" },
    { value: "zoom" },
    { value: "grayscale" },
    { value: "blur" },
    { value: "sepia" },
    { value: "invert" },
    { value: "tilt" },
    { value: "morph" }
  ],

  SHAPE: [
    "default",
    "rounded",
    "square",
    "roundedSquare",
    "circle",
    "rect",
    "rect4x3",
    "rect3x2",
    "rect5x4",
    "rect2x1",
    "rect3x1",
    "rect4x1",
    "rect45x10",
    "rect16x9",
    "rect16x10",
    "rect3x4",
    "rect4x5",
    "rect2x3",
    "rect1x2",
    "roundedRect",
    "roundedRect4x3",
    "roundedRect3x2",
    "roundedRect5x4",
    "roundedRect2x1",
    "roundedRect3x1",
    "roundedRect4x1",
    "roundedRect45x10",
    "roundedRect16x9",
    "roundedRect16x10",
    "roundedRect3x4",
    "roundedRect4x5",
    "roundedRect2x3",
    "roundedRect1x2"
  ],

  // ASPECT_RATIO: {
  //   "16:9": "256px",
  //   "3:2": "216px",
  //   "4:3": "192px",
  //   "2:3": "144px",
  //   "1:10": "40px",
  //   "10:1": "400px",
  //   "1:1": "216px",
  //   custom: ""
  // },

  ASPECT_RATIO_VERTICAL: {
    "1:1": "252px",
    "3:2": "378px",
    "4:3": "336px",
    "5:4": "315px",
    "16:9": "448px",
    "16:10": "403px",
    "1:2": "126px",
    "2:3": "168px",
    "3:4": "189px",
    "4:5": "202px",
    custom: ""
  },

  ASPECT_RATIO_HORIZONTAL: {
    "2:1": "288px",
    "3:1": "432px",
    "4:1": "576px",
    "45:10": "648px",
    custom: ""
  },

  SCREEN_SIZES: ["xs", "s", "m", "l", "xl"],

  urlRegex: /^(?:http(s)?:\/\/)/,
  elevation: {
    name: "elevation",
    type: "switchSelector",
    label: Lsi.elevation,
    getProps: () => ({
      items: [{ value: -1 }, { value: 0 }, { value: 1 }, { value: 2 }, { value: 3 }, { value: 4 }, { value: 5 }]
    })
  },

  bgStyle: {
    name: "bgStyle",
    type: "switchSelector",
    label: Lsi.bgStyle,
    getProps: () => ({
      items: [
        { value: "filled", content: Lsi.filled },
        { value: "outline", content: Lsi.outline },
        { value: "transparent", content: Lsi.transparent },
        { value: "underline", content: Lsi.underline }
      ]
    })
  },

  borderRadius: {
    name: "borderRadius",
    type: "switchSelector",
    label: Lsi.borderRadius,
    getProps: () => ({
      items: [
        { value: "0", content: Lsi.none },
        { value: "2px" },
        { value: "4px" },
        { value: "8px" },
        { value: "16px" }
      ]
    })
  }
};

export default Cfg;
