// uu_contentkitg01/dist-esm/core/uu5string-editor-wrapper.js
import * as UU5 from "uu5g04";
import "uu5g04-forms";

import Cfg from "./_config.js";

import "./uu5string-editor-wrapper.less";

export const Uu5stringEditorWrapper = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Forms.InputMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.tag("Core.Uu5stringEditorWrapper"),
    nestingLevelList: UU5.Environment.getNestingLevelList("bigBoxCollection", "box"),
    classNames: {
      main: Cfg.css("uu5string-editor-wrapper"),
      label: Cfg.css("uu5string-editor-wrapper-label"),
      content: Cfg.css("uu5string-editor-wrapper-content")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: UU5.PropTypes.string,
    rows: UU5.PropTypes.number,
    maxRows: UU5.PropTypes.number,
    onBlur: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      value: ""
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  reset_() {
    this._input.setValue(this.props.value);
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  getValue_() {
    return this._input.getValue();
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getStringEditor() {
    return UU5.Common.Tools.findComponent("UU5.CodeKit.Uu5StringEditor", {
      ref_: ref => (this._input = ref),
      className: this.getClassName("content"),
      value: this.props.value,
      rows: this.props.rows,
      maxRows: this.props.maxRows,
      onBlur: this.props.onBlur,
      format: "pretty"
    });
  },
  _getChild() {
    let inputId = this.getId() + "-input";
    return (
      <UU5.Bricks.Div {...this.getMainPropsToPass()}>
        {this.getLabel(inputId)}
        {this.getInputWrapper([this._getStringEditor(inputId)])}
      </UU5.Bricks.Div>
    );
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this._getChild();
  }
  //@@viewOff:render
});

export default Uu5stringEditorWrapper;
