// uu_contentkitg01/dist-esm/config/lsi.js
import UU5 from "uu5g04";

const Lsi = {
  blockEditableIconContentError: {
    cs: "Povoleny jsou maximalně 2 znaky.",
    en: "There are allowed only 2 characters."
  },
  blockEditableLabelIcon: {
    cs: "Zvolit ikonu",
    en: "Choose icon"
  },

  blockEditableLabelIconContent: {
    cs: "Vlastní obsah",
    en: "Custom content"
  },

  blockEditableLabelIconDefault: {
    cs: "Defaultní ikona",
    en: "Default icon"
  },
  blockEditableLabelText: {
    cs: "Text",
    en: "Text"
  },

  blockEditableIconPlaceholder: {
    cs: "Ponechte prázdné pro výchozí ikonu",
    en: "Leave blank for default icon"
  },

  LSI_TABLE_SHOW_HIDE_COL_WIDTH: {
    en: "Show/Hide column width",
    cs: "Zobrazit/Skrýt šířky sloupců"
  },
  LSI_TABLE_COVER_ROW_HEADER: {
    en: "Header row",
    cs: "Hlavičkový řádek"
  },
  LSI_TABLE_COVER_COL_HEADER: {
    en: "Header column",
    cs: "Hlavičkový sloupec"
  },
  LSI_TABLE_COVER_TRANSPOSE: {
    en: "Transpose the table",
    cs: "Transponovat tabulku"
  },
  LSI_TABLE_COVER_RESPONSIVE: {
    en: "Responsible table",
    cs: "Responsivní tabulka"
  },

  LSI_MORE_PROPERTIES_BUTTON: {
    en: "Other settings",
    cs: "Další nastavení"
  },
  LSI_BUTTON_OK: {
    en: "Save",
    cs: "Uložit"
  },
  LSI_BUTTON_CANCEL: {
    en: "Cancel",
    cs: "Zrušit"
  },

  LSI_LABEL_TABLE_NO_DATA: {
    // TODO props ??
    en: "Table has no data",
    cs: "Tabulka nemá zadaná data."
  },

  LSI_LABEL_INVALID_DATA: {
    en: "Invalid data.",
    cs: "Nevalidní data."
  },

  LSI_FORM_BUTTON_CANCEL: {
    en: "Cancel",
    cs: "Zrušit"
  },

  LSI_FORM_BUTTON_OK: {
    en: "Confirm",
    cs: "Potvrdit"
  },

  LSI_DCC_COMPONENT_ADD_LIST_ITEM_BEFORE: {
    en: "Add item before",
    cs: "Přidat položku před"
  },
  LSI_DCC_COMPONENT_ADD_LIST_ITEM_AFTER: {
    en: "Add item after",
    cs: "Přidat položku za"
  },
  LSI_DCC_COMPONENT_REMOVE_LIST_ITEM: {
    en: "Remove item",
    cs: "Smazat položku"
  },

  advance: <UU5.Bricks.Lsi lsi={{ cs: "Pokročilé", en: "Advanced" }} />,
  icon: <UU5.Bricks.Lsi lsi={{ cs: "Ikonka", en: "Icon" }} />,
  linkProperties: <UU5.Bricks.Lsi lsi={{ cs: "Vlastnosti odkazu", en: "Link proprties" }} />,
  link: <UU5.Bricks.Lsi lsi={{ cs: "Odkaz", en: "Link" }} />,
  display: <UU5.Bricks.Lsi lsi={{ cs: "Vzhled", en: "Visual" }} />,
  componentProperties: <UU5.Bricks.Lsi lsi={{ cs: "Základní vlastnosti", en: "Basic properties" }} />,
  sharedProps: <UU5.Bricks.Lsi lsi={{ cs: "Sdílené vlastnosti", en: "Shared properties" }} />,
  iconSettings: <UU5.Bricks.Lsi lsi={{ cs: "Vlastnosti ikony", en: "Icon properties" }} />,
  code: <UU5.Bricks.Lsi lsi={{ cs: "Kód", en: "Code" }} />,
  fragment: <UU5.Bricks.Lsi lsi={{ cs: "Fragment", en: "Fragment" }} />,
  target: <UU5.Bricks.Lsi lsi={{ cs: "Otevřít v", en: "Open in" }} />,
  title: <UU5.Bricks.Lsi lsi={{ cs: "Titulek", en: "Title" }} />,
  content: <UU5.Bricks.Lsi lsi={{ cs: "Obsah", en: "Content" }} />,
  text: <UU5.Bricks.Lsi lsi={{ cs: "Text", en: "Text" }} />,
  name: <UU5.Bricks.Lsi lsi={{ cs: "Název", en: "Name" }} />,
  color: <UU5.Bricks.Lsi lsi={{ cs: "Barva", en: "Color" }} />,
  elevation: <UU5.Bricks.Lsi lsi={{ en: "Elevation" }} />,
  headerLevel: <UU5.Bricks.Lsi lsi={{ cs: "Úroveň hlavičky", en: "Header level" }} />,
  level: <UU5.Bricks.Lsi lsi={{ cs: "Úroveň", en: "Level" }} />,
  elevationHover: <UU5.Bricks.Lsi lsi={{ cs: "Elevation po najetí myši", en: "Elevation on hover" }} />,
  borderRadius: <UU5.Bricks.Lsi lsi={{ cs: "Zaoblení rohů", en: "Border radius" }} />,
  linkType: <UU5.Bricks.Lsi lsi={{ cs: "Typ linku", en: "Link type" }} />,
  type: <UU5.Bricks.Lsi lsi={{ cs: "Typ", en: "Type" }} />,
  aspectRatio: <UU5.Bricks.Lsi lsi={{ cs: "Poměr stran", en: "Aspect ratio" }} />,
  width: <UU5.Bricks.Lsi lsi={{ cs: "Šířka", en: "Width" }} />,
  height: <UU5.Bricks.Lsi lsi={{ cs: "Výška", en: "Height" }} />,
  image: <UU5.Bricks.Lsi lsi={{ cs: "Obrázek", en: "Image" }} />,
  imageSettings: <UU5.Bricks.Lsi lsi={{ cs: "Vlastnosti obrázku", en: "Image properties" }} />,
  file: <UU5.Bricks.Lsi lsi={{ cs: "Soubor", en: "File" }} />,
  effect: <UU5.Bricks.Lsi lsi={{ cs: "Efekt", en: "Effect" }} />,
  effectImage: <UU5.Bricks.Lsi lsi={{ cs: "Efekt obrázku", en: "Image effect" }} />,
  shape: <UU5.Bricks.Lsi lsi={{ cs: "Tvar", en: "Shape" }} />,
  imageShape: <UU5.Bricks.Lsi lsi={{ cs: "Tvar obrázku", en: "Image shape" }} />,
  fit: <UU5.Bricks.Lsi lsi={{ cs: "Přizpůsobení", en: "Fit" }} />,
  stretch: <UU5.Bricks.Lsi lsi={{ cs: "roztáhnout", en: "stretch" }} />,
  defaultSize: <UU5.Bricks.Lsi lsi={{ cs: "zachovat originální velikost", en: "maintain original size" }} />,
  alignment: <UU5.Bricks.Lsi lsi={{ cs: "Zarovnání", en: "Alignment" }} />,
  imageAlignment: <UU5.Bricks.Lsi lsi={{ cs: "Zarovnání obrázku", en: "Image alignment" }} />,
  contentAlignment: <UU5.Bricks.Lsi lsi={{ cs: "Zarovnání obsahu", en: "Content alignment" }} />,
  textAlignment: <UU5.Bricks.Lsi lsi={{ cs: "Zarovnání text", en: "Text alignment" }} />,
  arrangement: <UU5.Bricks.Lsi lsi={{ cs: "Uspořádání", en: "Arrangement" }} />,
  arrangementMobile: <UU5.Bricks.Lsi lsi={{ cs: "Uspořádání na mobilu", en: "Arrangement on mobile" }} />,
  showLink: <UU5.Bricks.Lsi lsi={{ cs: "Zobrazit odkaz", en: "Show link" }} />,
  showCode: <UU5.Bricks.Lsi lsi={{ cs: "Zobrazit kód", en: "Show code" }} />,
  showDate: <UU5.Bricks.Lsi lsi={{ cs: "Zobrazit datum", en: "Show date" }} />,
  lightbox: <UU5.Bricks.Lsi lsi={{ en: "Lightbox" }} />,
  imageCode: <UU5.Bricks.Lsi lsi={{ cs: "Kód obrázku", en: "Image code" }} />,
  imageThumbCode: <UU5.Bricks.Lsi lsi={{ cs: "Kód náhledu", en: "Thumbnail code" }} />,
  imageSelect: <UU5.Bricks.Lsi lsi={{ cs: "Vybrat obrázek", en: "Select image" }} />,
  setContent: <UU5.Bricks.Lsi lsi={{ cs: "Nastavit obsah", en: "Set content" }} />,
  fromSelect: <UU5.Bricks.Lsi lsi={{ cs: "Vybrat ze seznamu", en: "Select from list" }} />,
  imageUpload: <UU5.Bricks.Lsi lsi={{ cs: "Nahrát obrázek", en: "Upload image" }} />,
  fileUpload: <UU5.Bricks.Lsi lsi={{ cs: "Nahrát soubor", en: "Upload file" }} />,
  rotate: <UU5.Bricks.Lsi lsi={{ cs: "Otočení", en: "Rotate" }} />,
  translate: <UU5.Bricks.Lsi lsi={{ cs: "Posunutí", en: "Translate" }} />,
  linkSmall: <UU5.Bricks.Lsi lsi={{ cs: "odkaz", en: "link" }} />,
  block: <UU5.Bricks.Lsi lsi={{ cs: "blok", en: "block" }} />,
  buttonSmall: <UU5.Bricks.Lsi lsi={{ cs: "tlačítko", en: "button" }} />,
  touchIcon: <UU5.Bricks.Lsi lsi={{ cs: "touch ikona", en: "touch icon" }} />,
  basicInfo: <UU5.Bricks.Lsi lsi={{ en: "Basic props", cs: "Základní vlastnosti" }} />,
  slace: <UU5.Bricks.Lsi lsi={{ en: "Scale", cs: "Přibližování" }} />,
  buttonSettings: <UU5.Bricks.Lsi lsi={{ en: "Button properties", cs: "Vlastnosti tlačítka" }} />,
  component: <UU5.Bricks.Lsi lsi={{ en: "Component", cs: "Komponenta" }} />,
  header: <UU5.Bricks.Lsi lsi={{ en: "Header", cs: "Nadpis" }} />,
  gradient: <UU5.Bricks.Lsi lsi={{ en: "Gradient" }} />,
  bgColor: <UU5.Bricks.Lsi lsi={{ en: "Background color", cs: "Barva pozadí" }} />,
  bgStyle: <UU5.Bricks.Lsi lsi={{ en: "Background style", cs: "Styl pozadí" }} />,
  bgStyleButton: <UU5.Bricks.Lsi lsi={{ en: "Button style", cs: "Styl tlačítka" }} />,
  verticalAlign: <UU5.Bricks.Lsi lsi={{ en: "Vertical align", cs: "Vertikální zarovnání" }} />,
  richText: <UU5.Bricks.Lsi lsi={{ en: "Content", cs: "Obsah" }} />,
  buttonText: <UU5.Bricks.Lsi lsi={{ en: "Button text", cs: "Text tlačítka" }} />,
  buttonHref: <UU5.Bricks.Lsi lsi={{ en: "Button link", cs: "Odkaz" }} />,
  buttonSize: <UU5.Bricks.Lsi lsi={{ en: "Button size", cs: "Velikost tlačítka" }} />,
  buttonTarget: <UU5.Bricks.Lsi lsi={{ en: "Button target", cs: "Cílové okno tlačítka" }} />,
  buttonBgColor: <UU5.Bricks.Lsi lsi={{ en: "Button background color", cs: "Barva pozadí tlačítka" }} />,
  scale: <UU5.Bricks.Lsi lsi={{ en: "Scale", cs: "Přibližování" }} />,
  scaleOnHover: <UU5.Bricks.Lsi lsi={{ en: "Scale on hover", cs: "Přibližovat po najetí myši" }} />,
  scaleWidth: <UU5.Bricks.Lsi lsi={{ en: "Scale width", cs: "Šířka přiblížené dlaždice" }} />,
  scaleHeight: <UU5.Bricks.Lsi lsi={{ en: "Scale height", cs: "Výška přiblížené dlaždice" }} />,
  transitionTime: <UU5.Bricks.Lsi lsi={{ en: "Transition time", cs: "Délka efektu" }} />,
  date: <UU5.Bricks.Lsi lsi={{ cs: "Datum", en: "Date" }} />,
  colWidth: <UU5.Bricks.Lsi lsi={{ cs: "Šířka sloupce", en: "Column width" }} />,
  colWidthImage: <UU5.Bricks.Lsi lsi={{ cs: "Šířka sloupce obrázku", en: "Image column width" }} />,
  mode: <UU5.Bricks.Lsi lsi={{ cs: "Režim", en: "Mode" }} />,
  rowMinHeight: <UU5.Bricks.Lsi lsi={{ cs: "Minimální výška řádku", en: "Row minimal height" }} />,
  rowMaxHeight: <UU5.Bricks.Lsi lsi={{ cs: "Maximální výška řádku", en: "Row maximal height" }} />,
  lastRowMaxHeight: <UU5.Bricks.Lsi lsi={{ cs: "Maximální výška posledního řádku", en: "Last row maximal height" }} />,
  fileManager: <UU5.Bricks.Lsi lsi={{ cs: "Správce souborů", en: "File manager" }} />,
  setFile: <UU5.Bricks.Lsi lsi={{ cs: "Nastavit soubor", en: "Set file" }} />,
  setImage: <UU5.Bricks.Lsi lsi={{ cs: "Nastavit obrázek", en: "Set image" }} />,
  fileCode: <UU5.Bricks.Lsi lsi={{ cs: "Kód souboru", en: "File code" }} />,
  previewImage: <UU5.Bricks.Lsi lsi={{ cs: "Nastavit náhled", en: "Set preview" }} />,
  noneAlign: <UU5.Bricks.Lsi lsi={{ cs: "žádné", en: "none" }} />,
  left: <UU5.Bricks.Lsi lsi={{ cs: "vlevo", en: "left" }} />,
  right: <UU5.Bricks.Lsi lsi={{ cs: "vpravo", en: "right" }} />,
  center: <UU5.Bricks.Lsi lsi={{ cs: "uprostřed", en: "center" }} />,
  top: <UU5.Bricks.Lsi lsi={{ cs: "nahoru", en: "top" }} />,
  bottom: <UU5.Bricks.Lsi lsi={{ cs: "dolu", en: "bottom" }} />,
  none: <UU5.Bricks.Lsi lsi={{ cs: "žádné", en: "none" }} />,
  noneLightbox: <UU5.Bricks.Lsi lsi={{ cs: "žádný", en: "none" }} />,
  global: <UU5.Bricks.Lsi lsi={{ cs: "globální", en: "global" }} />,
  local: <UU5.Bricks.Lsi lsi={{ cs: "lokální", en: "local" }} />,
  attachment: <UU5.Bricks.Lsi lsi={{ cs: "ke stažení", en: "download" }} />,
  inline: <UU5.Bricks.Lsi lsi={{ cs: "otevřít v obsahu stránky", en: "open in page content" }} />,
  vertical: <UU5.Bricks.Lsi lsi={{ cs: "Vertikální ", en: "Vertical" }} />,
  horizontal: <UU5.Bricks.Lsi lsi={{ cs: "Horizontální", en: "Horizontal" }} />,
  alt: <UU5.Bricks.Lsi lsi={{ cs: "Alternativní text", en: "Alternative text" }} />,
  margin: <UU5.Bricks.Lsi lsi={{ cs: "Odsazení", en: "Margin" }} />,
  marginTop: <UU5.Bricks.Lsi lsi={{ cs: "Horní odsazení", en: "Margin top" }} />,
  marginBottom: <UU5.Bricks.Lsi lsi={{ cs: "Spodní odsazení", en: "Margin bottom" }} />,
  all: <UU5.Bricks.Lsi lsi={{ cs: "všechna rozlišení", en: "all screen sizes" }} />,
  different: <UU5.Bricks.Lsi lsi={{ cs: "různá rozlišení", en: "different screen sizes" }} />,
  custom: <UU5.Bricks.Lsi lsi={{ cs: "vlastní", en: "custom" }} />,
  addImages: <UU5.Bricks.Lsi lsi={{ cs: "Přidat obrázky", en: "Add images" }} />,
  url: <UU5.Bricks.Lsi lsi={{ cs: "URL", en: "URL" }} />,
  imageUrl: <UU5.Bricks.Lsi lsi={{ cs: "URL obrázku", en: "Image URL" }} />,
  selectImages: <UU5.Bricks.Lsi lsi={{ cs: "Vybrat obrázky", en: "Select images" }} />,
  arrangementTextImage: <UU5.Bricks.Lsi lsi={{ cs: "text a obrázek", en: "text and image" }} />,
  arrangementImageText: <UU5.Bricks.Lsi lsi={{ cs: "obrázek a text", en: "image and text" }} />,
  minHeight: <UU5.Bricks.Lsi lsi={{ cs: "Minimální výška řádku", en: "Row minimal height" }} />,
  maxHeight: <UU5.Bricks.Lsi lsi={{ cs: "Maximální výška řádku", en: "Row maximal height" }} />,
  minWidth: <UU5.Bricks.Lsi lsi={{ cs: "Minimální šířka sloupce", en: "Column minimal width" }} />,
  maxWidth: <UU5.Bricks.Lsi lsi={{ cs: "Maximální šířka řádku", en: "Column maximal width" }} />,
  columnHeight: <UU5.Bricks.Lsi lsi={{ cs: "Výška sloupce", en: "Column height" }} />,
  randomOrder: <UU5.Bricks.Lsi lsi={{ cs: "Náhodné pořadí obrázků", en: "Random image order" }} />,
  circular: <UU5.Bricks.Lsi lsi={{ cs: "kruhový", en: "circular" }} />,
  final: <UU5.Bricks.Lsi lsi={{ cs: "final", en: "final" }} />,
  rewind: <UU5.Bricks.Lsi lsi={{ cs: "přetočit", en: "rewind" }} />,
  hideControls: <UU5.Bricks.Lsi lsi={{ cs: "Žádné ikony pro pohyb", en: "Hide icons for moving" }} />,
  hideIndicators: <UU5.Bricks.Lsi lsi={{ cs: "Schovat navigační bar", en: "Hide navigation bar" }} />,
  stepByOne: <UU5.Bricks.Lsi lsi={{ cs: "Postupovat po jednom obrázku", en: "Step by one image" }} />,
  nextIcon: <UU5.Bricks.Lsi lsi={{ cs: "Ikona pro následující obrázek", en: "Icon for next image" }} />,
  previousIcon: <UU5.Bricks.Lsi lsi={{ cs: "Ikona pro předchozí obrázek", en: "Icon for previous image" }} />,
  displayedImages: <UU5.Bricks.Lsi lsi={{ cs: "Počet zobrazených obrázek", en: "Number of images displayed" }} />,
  interval: <UU5.Bricks.Lsi lsi={{ cs: "Interval v milisekundách", en: "Interval in milliseconds" }} />,
  addButton: <UU5.Bricks.Lsi lsi={{ cs: "Přidat", en: "Add" }} />,
  typeTile: <UU5.Bricks.Lsi lsi={{ cs: "Typ dlaždice", en: "Tile type" }} />,
  largeTile: <UU5.Bricks.Lsi lsi={{ cs: "velká", en: "large" }} />,
  smallTile: <UU5.Bricks.Lsi lsi={{ cs: "malá", en: "small" }} />,
  imageTile: <UU5.Bricks.Lsi lsi={{ cs: "obrázková", en: "image" }} />,

  blank: <UU5.Bricks.Lsi lsi={{ cs: "Nové kartě", en: "New card" }} />,
  self: <UU5.Bricks.Lsi lsi={{ cs: "Stávající kartě", en: "Current card" }} />,

  buttonTextColor: { en: "Button text color", cs: "Barva textu tlačítka" },
  className: <UU5.Bricks.Lsi lsi={{ cs: "CSS Třída", en: "CSS Class" }} />,
  style: <UU5.Bricks.Lsi lsi={{ cs: "Styl", en: "Style" }} />,

  filters: <UU5.Bricks.Lsi lsi={{ cs: "Filtry", en: "Filters" }} />,
  grayscale: <UU5.Bricks.Lsi lsi={{ cs: "Odstín šedé barvy", en: "Grayscale" }} />,
  blur: <UU5.Bricks.Lsi lsi={{ cs: "Rozmazání", en: "Blur" }} />,
  brightness: <UU5.Bricks.Lsi lsi={{ cs: "Jas", en: "Brightness" }} />,
  contrast: <UU5.Bricks.Lsi lsi={{ cs: "Kontrast", en: "Contrast" }} />,
  invert: <UU5.Bricks.Lsi lsi={{ cs: "Invertovat", en: "Invert" }} />,
  opacity: <UU5.Bricks.Lsi lsi={{ cs: "Průhlednost", en: "Opacity" }} />,

  imagePosition: <UU5.Bricks.Lsi lsi={{ cs: "Pozice obrázku", en: "Image position" }} />,
  imageSize: <UU5.Bricks.Lsi lsi={{ cs: "Velikost obrázku", en: "Image size" }} />,
  imageRepeat: <UU5.Bricks.Lsi lsi={{ cs: "Opakování obrázku", en: "Image repeat" }} />,
  repeat: <UU5.Bricks.Lsi lsi={{ cs: "opakovat", en: "repeat" }} />,
  noRepeat: <UU5.Bricks.Lsi lsi={{ cs: "neopakovat", en: "no repeat" }} />,
  parallax: <UU5.Bricks.Lsi lsi={{ en: "Parallax" }} />,

  filled: <UU5.Bricks.Lsi lsi={{ cs: "vyplněný", en: "filled" }} />,
  outline: <UU5.Bricks.Lsi lsi={{ cs: "ohraničený", en: "outline" }} />,
  transparent: <UU5.Bricks.Lsi lsi={{ cs: "transparentní", en: "transparent" }} />,
  underline: <UU5.Bricks.Lsi lsi={{ cs: "podtržený", en: "underline" }} />,
  rounded: <UU5.Bricks.Lsi lsi={{ cs: "kulaté", en: "rounded" }} />,

  failedToUploadFile: <UU5.Bricks.Lsi lsi={{ cs: "Nepodařilo se nahrát soubor", en: "Failed to upload file" }} />,
  failedToDeleteFile: <UU5.Bricks.Lsi lsi={{ cs: "Nepodařilo se smazat soubor", en: "Failed to delete file" }} />,
  failedToUploadImage: <UU5.Bricks.Lsi lsi={{ cs: "Nepodařilo se nahrát obrázek", en: "Failed to upload image" }} />,
  failedToDeleteImage: <UU5.Bricks.Lsi lsi={{ cs: "Nepodařilo se smazay obrázek", en: "Failed to delete image" }} />,
  fileCodeCopied: (
    <UU5.Bricks.Lsi lsi={{ cs: "Kód byl zkopírován do schránky.", en: "Code was copied to clipboard." }} />
  ),
  setFileAlert: <UU5.Bricks.Lsi lsi={{ cs: "Soubor byl nastaven.", en: "File was set." }} />,
  validUrl: <UU5.Bricks.Lsi lsi={{ cs: "Není zadána celá URL", en: "The entire URL is not specified" }} />,

  loadFileDataFailed: { cs: "Nepodařilo se načíst soubor.", en: "Failed to load file" },
  codeIsNotSet: { cs: "Není nastaven kód souboru.", en: "File code is not set." },
  useFile: { cs: "Použít soubor", en: "Use file" },
  linkPlaceholder: { cs: "Kód nebo url", en: "Code or url" },
  tagSelectError: <UU5.Bricks.Lsi lsi={{ cs: "Nevalidní vstup", en: "Invalid input" }} />,
  noFileSelected: { cs: "Není vybrán žádný soubor", en: "No file selected" },
  noImageSelected: { cs: "Není vybrán žádný obrázek", en: "No image selected" },
  fileCodeSelected: { cs: "Vybraný soubor", en: "Selected file" },
  imageCodeSelected: { cs: "Vybraný obrázek", en: "Selected image" },
  fileUploaderPH: { cs: "Sem přetáhněte obrázky k nahrání", en: "Drop images to upload here" },
  altPH: { cs: "Alternativní text", en: "Alternative text" },
  headerPH: { cs: "Hlavička", en: "Header" },
  headerLevelTooltip: { cs: "Úroveň hlavičky", en: "Header level" },

  bulkAction: { cs: "Smazat", en: "Delete" },
  gallerySettings: { cs: "Nastavení galerie", en: "Gallery settings" },
  imagePanelSettings: { cs: "Nastavení obrázkového panelu", en: "Image panel settings" },
  newsTileSettings: { cs: "Nastavení novinkové dlaždice", en: "News tile settings" },
  addImagesNoComp: { cs: "Přidat obrázky", en: "Add images" },
  addImagesUrl: { cs: "Přidat obrázek pomocí URL", en: "Add image by URL" },
  delete: { cs: "Smazat", en: "Delete" },
  setImagePlain: { cs: "Nastavit obrázek", en: "Set image" },
  addNewTile: { cs: "Přidat novou dlaždici", en: "Add new tile" },
  setNewsTileList: { cs: "Nastavení novinkového seznam", en: "News list settings" },

  help: {
    fragment: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Umožní odscrollovat obrazovku uživatele na daný element cílové stránky.",
          en: "Allows you to scroll the user's screen to a given page element."
        }}
      />
    ),
    contentDisposition: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Umožnuje nastavit zda se soubor má rovnou stáhnout nebo zobrazit v obsahu stránky po kliknutí.",
          en:
            "Allows you set whether the file should be downloaded directly or displayed in the content of the page after clicking."
        }}
      />
    ),
    lightbox: (
      <UU5.Bricks.Lsi
        lsi={{
          cs:
            "Umožní zobrazit obrázek v modálním okně v jeho plné velikosti a přepínat mezi všemi obrázky na stránce, které mají zapnutý lightbox.",
          en:
            "Allows you to view the image in a modal window to its full size and switch between all the images on the page that have enabled lightbox."
        }}
      />
    ),
    fit: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: `<uu5string/>default - obrázek vyplní celou šířku a zachová si poměr stran <br/>coverNoScaleUp - obrázek je vycenstrovaný a zmenšený, aby pokryl celou oblast a zachoval si poměr stran`,
          en: `<uu5string/>default - the image fills the whole width of the area preserving its own aspect ratio <br/>coverNoScaleUp - the image is centered and shrinked to cover the whole area preserving its own aspect ratio`
        }}
      />
    ),
    alt: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Zobrazí se v případě, že se nepodaří načíst vybraný obrázek.",
          en: "Appears if the selected image fails to load."
        }}
      />
    ),
    altTooltip: {
      cs: "Zobrazí se v případě, že se nepodaří načíst vybraný obrázek.",
      en: "Appears if the selected image fails to load."
    },
    margin: (
      <UU5.Bricks.Lsi
        lsi={{
          cs:
            "Odsazení lze zadat pro všechny strany stejné, např. 16px, nebo pro každou zvlášť, např. 8px 16px 8px 8px. Lze používat různé jednotky (px, em, vw, % a další).",
          en:
            "The margin can be specified same for all sides, e.g. 16px, or can be different for all sides, e.g. 8px 16px 8px 8px. Different units can be used (px, em, vw, % and others)."
        }}
      />
    ),
    marginOneSide: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Pro definování odsazení lze použít různé jednotky (px, em, vw, % a další).",
          en: "Different units can be used to define the margin (px, em, vw, % and others)."
        }}
      />
    ),
    units: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Lze použít různé jednotky (px, em, vw, % a další). Pokud není jednotka specifikována, použijí se px.",
          en: "Different units can be used (px, em, vw, % and others). If the unit is not specified, px is used."
        }}
      />
    ),
    lightboxGallery: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: `<uu5string/>lokální - bude možné procházet obrázky definovaný pouze v této galerii<br/>
                globální - umožní procházet mezi všemi obrázky na stránce<br/>
                žádný - nebude možné otevřít lightbox
          `,
          en: `<uu5string/>local - it will be possible to cycle only through the images defined specifically for this gallery<br/>
                global - allow to switch between all the images on the page that have enabled lightbox<br/>
                none - it won't be possible to open the lightbox at all.
          `
        }}
      />
    ),
    thumbMode: (
      <UU5.Bricks.Lsi
        lsi={{
          cs:
            "Track zajišťuje, že všechny obrázky v jednotlivých řádcích mají stejnou výšku a přizpůsobí svojí šířku tak, že nevzniknou velké mezery mezi nima.",
          en:
            "Track ensures that all the images in the individual rows have the same height and adjusts their width so that the large gaps do not occur between them."
        }}
      />
    ),
    interval: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Pokud je hodnota nastavena na 0, tak obrázky se nebudou automaticky posouvat.",
          en: "If the value is set to 0, the images will not automatically move forth."
        }}
      />
    ),
    carouselWidth: (
      <UU5.Bricks.Lsi
        lsi={{
          cs:
            "Šířka komponenty (px). Pokud je zadána společně s výškou, tak má šířka přednost. Druhý rozměr se vypočítá tak, aby se zachovaly poměry stran obrázku.",
          en:
            "Width of component (px). It takes precedence over the property height if it is entered together. The second dimension is calculated to preserve the aspect ratios of the image."
        }}
      />
    ),
    carouselHeight: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Výška komponenty (px). Druhý rozměr se vypočítá tak, aby se zachovaly poměry stran obrázku.",
          en: "Height of component (px). The second dimension is calculated to preserve the aspect ratios of the image."
        }}
      />
    ),
    infoCarouselHeight: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Výška komponenty (px).",
          en: "Height of component (px)."
        }}
      />
    ),
    translate: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Defaultní jednotka je %. Jedná se o posunutí po ose Y.",
          en: "The default unit is %. It is a displacement along the Y axis."
        }}
      />
    ),
    rotate: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Hodnota je ve stupních.",
          en: "The value is in degrees."
        }}
      />
    ),
    imageColumn: (
      <UU5.Bricks.Lsi
        lsi={{
          cs:
            "Výšku, minimální šířku a maximální šířku lze zadat pouze jako číslo a daná hodnota reprezentuje počet pixelů.",
          en:
            "Height, minimum width and maximum width can only be entered as a number and the value represents the number of pixels."
        }}
      />
    ),
    imageRow: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Minimální a maximální výšku řádku lze zadat pouze jako číslo a daná hodnota reprezentuje počet pixelů.",
          en:
            "The minimum and maximum row heights can only be entered as a number and the value represents the number of pixels."
        }}
      />
    ),
    previewImage: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: `<uu5string/>contain - zobrazí celý obrázek se zachovaným poměrem stran<br/>
                cover - obrázek si zachová poměr stran a vyplní celý obsah<br/>
                icon - zobrazí ikonu dle typu souboru nebo vlastního výběru
          `,
          en: `<uu5string/>contain - displays the entire image with the aspect ratio maintained<br/>
                cover - image retains the aspect ratio and fills in the entire content<br/>
                icon - displays an icon according to the file type or custom selection
          `
        }}
      />
    ),
    parallax: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Vytvoří efekt scrollování obsahu stránky přes vybraný obrázek.",
          en: "Creates the effect of scrolling page content over the selected image."
        }}
      />
    ),
    gradient: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: '<uu5string/>Barvy vytvořené pomocí gradientu se mění průběžně s polohou a vytvářejí plynulé přechody barev. Přetíží nastavení barvy pozadí. O gradientu si můžete přečíct více <UU5.Bricks.Link target="_blank" content="zde." href="https://www.w3schools.com/css/css3_gradients.asp"/>',
          en: '<uu5string/>The colors produced by a gradient vary continuously with position, producing smooth color transitions. It will override setting of background color. You can read more about gradient <UU5.Bricks.Link target="_blank" content=\"here.\" href=\"https://www.w3schools.com/css/css3_gradients.asp\"/>'
        }}
      />
    ),
    sharedProps: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: "Umožňuje nastavit globálně sdílené vlastnosti jednotlivých tiles. Přímé přenastavení na jednotlivých tiles přetíží toto nastavení.",
          en: "Lets you set globally shared properties of each tile. Direct adjustment to individual tiles override this setting."
        }}
      />
    ),
    verticalAlign: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: `<uu5string/>Vertikální zarovnání komponent. Bude mít vliv pouze v případě, že bude vedle sebe vložených více komponent, které jsou definované jako inline-block.<br/>
                top - komponenta je na řádku zarovnána k hornímu okraji nejvyšší komponenty<br/>
                center - komponenta je na řádku zarovnána na střed vůči ostatním komponentám<br/>
                bottom - komponenta je na řádku zarovnána k spodnímu okraji nejníže položené komponentě<br/>
          `,
          en: `<uu5string/>Vertical alignment of components. It will only take effect if multiple components are inserted next to each other, which are defined as inline-block.<br/>
                top - the component is aligned with the top of the tallest component on the line<br/>
                center - the component is aligned in the middle with other component on the line<br/>
                bottom - the component is aligned with the lowest component on the line<br/>
          `
        }}
      />
    ),
    svgWarning: (
      <UU5.Bricks.Lsi
        lsi={{
          cs: `Vybraný obrázek není svg. Může způsobit neočekávané chování.`,
          en: `Selected image is not an svg. It can cause unexpected behavior.`
        }}
      />
    )
  },

  editation: {
    link: <UU5.Bricks.Lsi lsi={{ cs: "Upravit odkaz", en: "Edit link" }} />,
    fileLink: <UU5.Bricks.Lsi lsi={{ cs: "Upravit odkaz na soubor", en: "Edit file link" }} />,
    fileBlockLink: <UU5.Bricks.Lsi lsi={{ cs: "Upravit blokový odkazu na soubor", en: "Edit file block link" }} />,
    image: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázek", en: "Edit image" }} />,
    imageGallery: <UU5.Bricks.Lsi lsi={{ cs: "Upravit galerii", en: "Edit gallery" }} />,
    imageRow: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázkový řádek", en: "Edit image row" }} />,
    imageColumn: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázkový sloupec", en: "Edit image column" }} />,
    imageCarousel: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázkový carousel", en: "Edit image carousel" }} />,
    horizontalSeparator: (
      <UU5.Bricks.Lsi lsi={{ cs: "Upravit horizontální oddělovač", en: "Edit horizontal separator" }} />
    ),
    imagePanel: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázkový panel", en: "Edit image panel" }} />,
    imageItem: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázkovou položku", en: "Edit image item" }} />,
    backgroundImage: <UU5.Bricks.Lsi lsi={{ cs: "Upravit obrázek na pozadí", en: "Edit background image" }} />,
    newsTile: <UU5.Bricks.Lsi lsi={{ cs: "Upravit novinkovou dlaždici", en: "Edit news tile" }} />,
    newsTileList: <UU5.Bricks.Lsi lsi={{ cs: "Upravit seznam novinkových dlaždic", en: "Edit news tile list" }} />,
    infoTile: <UU5.Bricks.Lsi lsi={{ cs: "Upravit informační dlaždice", en: "Edit info tile" }} />,
    infoCarousel: <UU5.Bricks.Lsi lsi={{ cs: "Upravit informační carousel", en: "Edit info carousel" }} />
  }
};

export const PLACEHOLDERS = {
  imagePanel: {
    header: `Enjoy the summer`,
    content: `Give yourself some free time to prevent burning out or losing your spirit. Go out and experience new scenery, people, cultures, cuisine and activities. You will feel refreshed and full of stamina afterwards.`,
    button: `Start now`
  },
  fileBlockLink: {
    text: "No file selected"
  },
  infoCarousel: {
    header1: `Commuting problems`,
    content1: `Getting to your office every day from home can be stressful and time consuming. Getting stuck in traffic jams is not a very pleasant way to spend your free time. There are better ways to handle this.`,
    header2: `Stress in the city`,
    content2: `Living directly in the city where you work is a great way to solve every day long distance travels. Life in the city is stressful in a different way and using the overcrowded public transport can be unpleasant experience as well.`,
    header3: `Teleworking`,
    content3: `Great solution is working from home. There are downsides as well though. Being closed in your house alone the whole day can be also stressful and not many people can find the will to focus on their tasks without procrastinating.`
  }
};

export default Lsi;
