// uu_plus4u5g01/dist-esm/app/about.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./about.less";

export const About = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.About",
    classNames: {
      main: "plus4u5-app-about",
      content: "plus4u5-app-about-content"
    },
    opt: {
      dummyLevel: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Section {...this.getMainPropsToPass()} content={null} level={this.props.level || 3}>
        <UU5.Bricks.P className={this.getClassName().content} content={this.props.content} />
      </UU5.Bricks.Section>
    );
  }
  //@@viewOff:render
});

export default About;
