// uu_plus4u5g01/dist-esm/app/app-lsi.js
export const AppLsi = {
  about: {
    label: {
      cs: "O nás",
      en: "About"
    }
  },
  authors: {
    leadingAuthors: {
      cs: "Hlavní autoři",
      en: "Leading Authors"
    },
    otherAuthors: {
      cs: "Vedlejší autoři",
      en: "Other Authors"
    },
    buttonAdd: {
      cs: "Přidat",
      en: "Add"
    },
    buttonCancel: {
      cs: "Zrušit",
      en: "Cancel"
    },
    buttonDelete: {
      cs: "Smazat",
      en: "Delete"
    },
    buttonSave: {
      cs: "Uložit",
      en: "Save"
    },
    addAuthorName: {
      cs: "Jméno a příjmení",
      en: "Name and Surname"
    },
    addAuthorUuId: {
      cs: "uuIdentity",
      en: "uuIdentity"
    },
    addAuthorRole: {
      cs: "Role",
      en: "Role"
    },
    addAuthorDesc: {
      cs: "Popis",
      en: "Description"
    },
    addAuthorPhoto: {
      cs: "Fotografie",
      en: "Photo"
    },
    textLanguageAdd: {
      cs: "Přidat jazyk",
      en: "Add Language"
    },
    addAuthorModalHeader: {
      cs: "Vyhledávání autora",
      en: "Search Author"
    },
    searchAuthorLabel: {
      cs: "Uživatel Plus4U",
      en: "Plus4U User"
    },
    searchAuthorNoResult: {
      cs: "Uživatel nebyl nalezen",
      en: "User was not found"
    },
    searchAuthorHint: {
      cs: "Zadejte jméno, příjmení nebo +4U ID uživatele Plus4U",
      en: "Enter Plus4U user name, surname or +4U ID"
    },
    searchAuthorInvalidInput: {
      cs: "Zadejte minimálně 3 znaky",
      en: "Please enter at least 3 characters"
    },
    authorsEditableDeleteHeader: {
      cs: "Smazat autora",
      en: "Delete Author"
    },
    authorsEditableEditHeader: {
      cs: "Upravit autora",
      en: "Update Author"
    },
    authorsEditableDeleteText: {
      cs: "Skutečně si přejete smazat tohoto autora?",
      en: "Are you sure you want to delete this author?"
    },
    authorsEditableInvalidImage: {
      cs: "Soubor musí být obrázek",
      en: "File must be an image"
    },
    textTooLong4000: {
      cs: "Text je příliš dlouhý (4000 znaků max).",
      en: "Text is too long (max 4000 characters)."
    },
    deletePageConfirm: {
      cs: "Opravdu chcete trvale smazat stránku?",
      en: "Do you really want to permanently delete the page?"
    }
  },
  copyBaseUri: {
    buttonText: {
      cs: "Kopírovat base URI",
      en: "Copy base URI"
    },
    baseUriNotFound: {
      cs: "Base URI aplikace se nepodařilo nalézt",
      en: "Application's base URI couldn't be found"
    }
  },
  languageSelector: {
    changePermanently: {
      cs: "Uložit jazyk",
      en: "Save language"
    },
    confirmChangePermanently: {
      cs: "Chcete uložit jazyk natrvalo? Změna se projeví ve všech aplikacích ve službě Plus4U.",
      en: "Do you want to store language settings? It will change the default language in all Plus4U applications."
    },
    confirmChangePermanentlyHeader: {
      cs: "Uložit nastavení jazyka",
      en: "Save language settings"
    },
    confirm: {
      cs: "OK",
      en: "OK"
    },
    refuse: {
      cs: "Storno",
      en: "Cancel"
    }
  },
  licence: {
    header: {
      cs: "Vlastník licence",
      en: "License"
    },
    organisation: {
      cs: "Organizace",
      en: "Organisation"
    },
    authorities: {
      cs: "Kompetentní osoby",
      en: "Authorities"
    },
    awid: {
      en: "AWID"
    }
  },
  loginButton: {
    login: {
      cs: "Přihlásit",
      en: "Log in"
    },
    registration: {
      cs: "Registrovat",
      en: "Register"
    },
    logout: {
      cs: "Odhlásit",
      en: "Log out"
    }
  },
  login: {
    loginText: {
      cs: "Nejprve se prosím přihlaste",
      en: "Please log in first"
    },
    registerText: {
      cs: "Pokud nemáte založen účet, můžete se zaregistrovat.",
      en: "If you don’t have an account yet, feel free to make a simple registration."
    },
    registerButton: {
      cs: "Vytvořit účet",
      en: "Create account"
    }
  },
  navigationBarLoader: {
    data: {
      cs: "Data ze serveru jsou prazdná",
      en: "Data from server are empty"
    },
    load: {
      cs: "Nepodařilo se načíst",
      en: "Couldn't load"
    },
    welcomeIn: {
      cs: "v",
      en: "in"
    }
  },
  navigator: {
    more: {
      cs: "Více",
      en: "More"
    },
    loadError: {
      cs: "Nejsou zobrazeny všechny ikony. Pokud problém přetrvává, kontaktujte Helpdesk.",
      en: "Some icons are not displayed. Contact Helpdesk if the problem persists.",
      sk: "Nie sú zobrazené všetky ikony. Ak problém pretrváva, kontaktujte Helpdesk."
    },
    dataError: {
      cs: "Nejsou zobrazeny všechny ikony z důvodu nevalidní konfigurace. Kontaktujte Helpdesk.",
      en: "Some icons are not displayed due to invalid configuration. Contact Helpdesk.",
      sk: "Nie sú zobrazené všetky ikony kvôli nevalidnej konfigurácii. Kontaktujte Helpdesk."
    }
  },
  resources: {
    header: {
      cs: "Použité zdroje",
      en: "Resources"
    }
  },
  spaError: {
    textError: {
      cs: "Omlouváme se, v aplikaci se vyskytla neočekávaná chyba.",
      en: "We're sorry, there was an unexpected error in the app."
    },
    continue: {
      cs: "Pokračovat na hlavní stránku",
      en: "Continue to main page"
    }
  },
  spaNotAuthenticated: {
    loginText: {
      cs: "Zobrazení obsahu vyžaduje přihlášeného uživatele. Přihlašte se nebo vytvořte nový účet.",
      en: "You are not logged in to view this content. Please log in or create an account."
    },
    registerText: {
      cs: "Pokud nemáte založen účet, můžete se zaregistrovat.",
      en: "If you don’t have an account yet, feel free to make a simple registration."
    },
    registerLink: {
      cs: "Vytvořit účet",
      en: "Create account"
    },
    helpDesk: {
      cs: "Pokud máte nějaký problém, obraťte se prosím na náš ", //link to Helpdsek
      en: "If you have a problem, please contact our " //link to Helpdsek
    }
  },
  support: {
    description: {
      en: "If you have a problem or idea to improve this application, tell us:",
      cs: "Pokud máte nějaký problém nebo nápad na zlepšení této aplikace, napište nám:"
    },
    descriptionLogin: {
      en: "If you have a problem or idea to improve this application, log in:",
      cs: "Pokud máte nějaký problém nebo nápad na zlepšení této aplikace, přihlašte se:"
    },
    productPortal: {
      en: "Product portal",
      cs: "Portál produktu"
    },
    flsButton: {
      en: "First line support",
      cs: "Primární podpora"
    }
  },
  technologies: {
    header: {
      cs: "Použité technologie",
      en: "Used Technologies"
    },
    technologies: {
      cs:
        '<uu5string/>V aplikaci/knihovně byly použity následující technologie společnosti <UU5.Bricks.Link href="https://unicorn.com/" content="Unicorn" target="_blank" />:',
      en:
        '<uu5string/>Following <UU5.Bricks.Link href="https://unicorn.com/" content="Unicorn" target="_blank" /> technologies were used to create the application/library:'
    },
    technologiesApplication: {
      cs:
        '<uu5string/>V aplikaci byly použity následující technologie společnosti <UU5.Bricks.Link href="https://unicorn.com/" content="Unicorn" target="_blank" />:',
      en:
        '<uu5string/>Following <UU5.Bricks.Link href="https://unicorn.com/" content="Unicorn" target="_blank" /> technologies were used to create the application:'
    },
    technologiesLibrary: {
      cs:
        '<uu5string/>V knihovně byly použity následující technologie společnosti <UU5.Bricks.Link href="https://unicorn.com/" content="Unicorn" target="_blank" />:',
      en:
        '<uu5string/>Following <UU5.Bricks.Link href="https://unicorn.com/" content="Unicorn" target="_blank" /> technologies were used to create the library:'
    },
    technologiesAnd: {
      cs: " a ",
      en: " and "
    }
  },
  search: {
    inputPlaceholder: {
      cs: "Hledat",
      en: "Search"
    }
  }
};

export default AppLsi;
