// uu_plus4u5g01/dist-esm/app/authors.js
import * as UU5 from "uu5g04";
import InternalAuthors from "./internal/authors.js";
import LazyExports from "./internal/lazy-exports.js";

const EditableAuthors = UU5.Common.Component.lazy(async () => {
  let ReactDnd = SystemJS.import("react-dnd");
  let backendName = UU5.Common.Tools.isMobileOrTablet ? "react-dnd-touch-backend" : "react-dnd-html5-backend";
  let ReactDndHtml5Backend = SystemJS.import(backendName);
  let [dndExports, backendExports] = await Promise.all([ReactDnd, ReactDndHtml5Backend]);
  LazyExports["react-dnd"] = dndExports;
  LazyExports[backendName] = backendExports;
  return import("./internal/editable-authors.js");
});

export const Authors = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.EditableMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Authors",
    classNames: {
      main: "plus4u5-app-authors"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    leadingAuthors: UU5.PropTypes.arrayOf(
      UU5.PropTypes.shape({
        name: UU5.PropTypes.string.isRequired,
        content: UU5.PropTypes.string,
        uuIdentity: UU5.PropTypes.string,
        signature: UU5.PropTypes.string,
        role: UU5.PropTypes.any.isRequired,
        src: UU5.PropTypes.string
      })
    ),
    otherAuthors: UU5.PropTypes.arrayOf(
      UU5.PropTypes.shape({
        name: UU5.PropTypes.string,
        content: UU5.PropTypes.string,
        uuIdentity: UU5.PropTypes.string,
        signature: UU5.PropTypes.string,
        role: UU5.PropTypes.any,
        src: UU5.PropTypes.string
      })
    ),
    header: UU5.PropTypes.any,
    underline: UU5.PropTypes.bool,
    level: UU5.PropTypes.oneOf(["0", "1", "2", "3", "4", "5", "6", 0, 1, 2, 3, 4, 5, 6]),
    getPhotoSrc: UU5.PropTypes.func,
    onChange: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      leadingAuthors: [],
      otherAuthors: null,
      header: null,
      level: 3,
      underline: false,
      getPhotoSrc: null,
      onChange: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onEndEditation(editedProps) {
    let { onChange } = this.props;

    let propsForOnChange = { ...this.props, ...editedProps };
    let defaultProps = this.constructor.getDefaultProps() || {};
    for (let k in defaultProps) {
      if (defaultProps[k] === propsForOnChange[k]) delete propsForOnChange[k];
    }
    for (let k in propsForOnChange) {
      if (typeof propsForOnChange[k] === "function" || propsForOnChange[k] === undefined) delete propsForOnChange[k];
    }
    delete propsForOnChange.parent;

    let propsToSave = { ...propsForOnChange };
    propsToSave.leadingAuthors = this._omitPhotoFiles(propsToSave.leadingAuthors);
    propsToSave.otherAuthors = this._omitPhotoFiles(propsToSave.otherAuthors);
    this.endEditation(propsToSave);

    if (typeof onChange === "function") onChange(propsForOnChange);
  },

  _omitPhotoFiles(authors) {
    return authors.map(author => {
      if (author.photoFile) {
        author = { ...author };
        delete author.photoFile;
      }
      return author;
    });
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { leadingAuthors, otherAuthors, header, level, underline, getPhotoSrc } = this.props;

    return this.isInlineEdited() ? (
      <UU5.Common.Suspense fallback="">
        <EditableAuthors
          {...this.getMainPropsToPass()}
          leadingAuthors={leadingAuthors}
          otherAuthors={otherAuthors}
          header={header}
          level={level}
          underline={underline}
          onEndEditation={this._onEndEditation}
          getPhotoSrc={getPhotoSrc}
        />
      </UU5.Common.Suspense>
    ) : (
      <InternalAuthors
        {...this.getMainPropsToPass()}
        leadingAuthors={leadingAuthors}
        otherAuthors={otherAuthors}
        header={header}
        level={level}
        underline={underline}
      />
    );
  }
  //@@viewOff:render
});

export default Authors;
