// uu_plus4u5g01/dist-esm/app/bottom.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./bottom.less";

export const Bottom = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Bottom",
    classNames: {
      main: "plus4u5-app-bottom uu5-noindex",
      content: "plus4u5-app-bottom-content"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getPropsToPass() {
    return UU5.Common.Tools.merge(this.getMainPropsToPass(), { content: null });
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let content = this.props.content || this.props.children || "© Copyright " + new Date().getFullYear();
    return (
      <UU5.Bricks.Div {...this._getPropsToPass()}>
        <UU5.Bricks.Span content={content} className={this.getClassName("content")} />
      </UU5.Bricks.Div>
    );
  }
  //@@viewOff:render
});

export default Bottom;
