// uu_plus4u5g01/dist-esm/app/copy-base-uri.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";

import "./copy-base-uri.less";

export const CopyBaseUri = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.CopyBaseUri",
    classNames: {
      main: "plus4u5-app-copy-base-uri"
    },
    lsi: () => Plus4U5.Environment.Lsi.App.copyBaseUri
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    icon: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      icon: "mdi-content-copy"
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _copyBaseUri() {
    let result;
    let appBasePath = UU5.Environment.getAppBasePath();
    let baseUri = appBasePath ? location.protocol + "//" + location.host + appBasePath : null;

    if (baseUri) {
      result = UU5.Common.Tools.copyToClipboard(baseUri);
    } else {
      result = this.showError(this.getLsiValue("baseUriNotFound"), null, null);
    }

    return result;
  },

  _getIcon() {
    let result = null;

    if (this.props.icon !== null) {
      result = <UU5.Bricks.Icon icon={this.props.icon} />;
    }

    return result;
  },

  _getContent() {
    return this.getChildren() || this.getLsiComponent("buttonText");
  },

  _getMainPropsToPass() {
    let props = this.getMainPropsToPass();

    props.onClick = this._copyBaseUri;
    props.colorSchema = this.props.colorSchema;

    return props;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Button {...this._getMainPropsToPass()}>
        {this._getIcon()}
        {this._getContent()}
      </UU5.Bricks.Button>
    );
  }
  //@@viewOff:render
});

export default CopyBaseUri;
