// uu_plus4u5g01/dist-esm/app/frame.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./frame.less";

export const Frame = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ScreenSizeMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Frame",
    nestingLevel: "container",
    classNames: {
      main: "plus4u5-app-frame",
      xs: "plus4u5-app-frame-xs",
      s: "plus4u5-app-frame-s",
      m: "plus4u5-app-frame-m",
      l: "plus4u5-app-frame-l",
      xl: "plus4u5-app-frame-xl"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainPropsToPass() {
    let props = this.getMainPropsToPass();
    return UU5.Common.Tools.mergeDeep(props, {
      className: props.className + " " + this.getClassName(this.getScreenSize())
    });
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() === "container" ? this.props.children : null;
  }
  //@@viewOff:render
});

export default Frame;
