// uu_plus4u5g01/dist-esm/app/language.js
import * as UU5 from "uu5g04";

const storageKey = "plus4u5UserLanguage";

function saveLanguageToStorage(language) {
  if (language !== getLanguageFromStorage()) {
    localStorage.setItem(storageKey, language);
  }
}

function getLanguageFromStorage() {
  return localStorage.getItem(storageKey);
}

function setLanguage(language, languages) {
  const foundLanguage = findLanguage(language, languages);
  if (foundLanguage) {
    UU5.Common.Tools.setLanguage(foundLanguage);
  }
}

function findLanguage(language, languages) {
  if (language !== UU5.Common.Tools.getLanguage()) {
    if (!languages || languages.indexOf(language) > -1) {
      return language;
    } else {
      let foundLanguage = language.match(/^\w+/)[0];
      if (languages.indexOf(foundLanguage) > -1) {
        return foundLanguage;
      } else {
        return null;
      }
    }
  }
}

export const Language = UU5.Common.Component.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Language"
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    languages: UU5.PropTypes.arrayOf(UU5.PropTypes.string)
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      languages: undefined
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    let storedLanguage = getLanguageFromStorage();

    if (storedLanguage) {
      setLanguage(storedLanguage, this.props.languages);
    }

    return {
      identity: !storedLanguage
    };
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    if (this.state.identity) {
      return (
        <UU5.Common.Identity>
          {({ identity }) => {
            if (identity) {
              return (
                <UU5.Common.Loader
                  uri="https://cmd.plus4u.net/PLUS4U-BT/uu-plus4uppl/PersonalCard/getAttributes/exec"
                  data={{ contact_list: false }}
                  headers={{ accept: "application/json" }}
                  authenticate
                >
                  {({ isLoading, isError, data }) => {
                    if (!isLoading && !isError && data && data.data && data.data.language) {
                      saveLanguageToStorage(data.data.language);
                      setLanguage(data.data.language, this.props.languages);
                    }

                    return null;
                  }}
                </UU5.Common.Loader>
              );
            } else {
              return null;
            }
          }}
        </UU5.Common.Identity>
      );
    } else {
      return null;
    }
  }
  //@@viewOff:render
});

export default Language;
