// uu_plus4u5g01/dist-esm/app/left.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./left.less";

export const Left = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Left",
    classNames: {
      main: "plus4u5-app-left"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Div {...this.getMainPropsToPass()}>
        Plus4U5.App.Left TODO ...
        <UU5.Bricks.Ol>
          <UU5.Bricks.Li>
            <UU5.Bricks.LinkUnicorn />
          </UU5.Bricks.Li>
          <UU5.Bricks.Li>
            <UU5.Bricks.LinkPlus4U />
          </UU5.Bricks.Li>
          <UU5.Bricks.Li>
            <UU5.Bricks.LinkUnicornCollege />
          </UU5.Bricks.Li>
          <UU5.Bricks.Li>
            <UU5.Bricks.LinkUAF />
          </UU5.Bricks.Li>
          <UU5.Bricks.Li>
            <UU5.Bricks.LinkUU5 />
          </UU5.Bricks.Li>
        </UU5.Bricks.Ol>
      </UU5.Bricks.Div>
    );
  }
  //@@viewOff:render
});

export default Left;
