// uu_plus4u5g01/dist-esm/app/loading.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./loading.less";

export const Loading = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Loading",
    classNames: {
      main: "plus4u5-app-loading uu5-noindex",
      spinner: "plus4u5-app-loading-spinner",
      content: "plus4u5-app-loading-content"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentDidMount() {
    UU5.Common.Tools.warning("Component Plus4u5.App.loading is deprecated! Use Plus4u5.App.SpaLoading instead.");
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let content = this.getContent() || this.props.children || <UU5.Bricks.Icon icon="plus4u5-plus4u" />;

    return (
      <UU5.Bricks.Loading {...this.getMainPropsToPass()} content={null}>
        <UU5.Bricks.Div className={this.getClassName().spinner}>
          <UU5.Bricks.Div className="first" />
          <UU5.Bricks.Div className="second" />
          <UU5.Bricks.Div className="third" />
        </UU5.Bricks.Div>
        <UU5.Bricks.Div className={this.getClassName().content} content={content} />
      </UU5.Bricks.Loading>
    );
  }
  //@@viewOff:render
});

export default Loading;
