// uu_plus4u5g01/dist-esm/app/login.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import { AuthenticationService } from "uu_appg01_oidc";
import * as Plus4U5 from "uu_plus4u5g01";
import LoginButton from "./login-button.js";

import "./login.less";

export const Login = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.LsiMixin,
    UU5.Common.IdentityMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Login",
    nestingLevelList: UU5.Environment.getNestingLevelList("box", "inline"),
    classNames: {
      main: "plus4u5-app-login uu5-noindex",
      text: "plus4u5-app-login-text",
      buttons: "plus4u5-app-login-buttons",
      loginText: "plus4u5-app-login-login-text",
      registerText: "plus4u5-app-login-register-text",
      loginButton: "plus4u5-app-login-login-button",
      registerButton: "plus4u5-app-login-register-button"
    },
    opt: {
      nestingLevelWrapper: true
    },
    lsi: () => Plus4U5.Environment.Lsi.App.login
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    loginProps: UU5.PropTypes.shape({
      scope: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.arrayOf(UU5.PropTypes.string)]),
      authenticationContext: UU5.PropTypes.object,
      usePopup: UU5.PropTypes.bool,
      acrValues: UU5.PropTypes.string,
      maxAge: UU5.PropTypes.number,
      language: UU5.PropTypes.string
    }),
    registrationProps: UU5.PropTypes.shape({
      scope: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.arrayOf(UU5.PropTypes.string)]),
      authenticationContext: UU5.PropTypes.object,
      usePopup: UU5.PropTypes.bool,
      acrValues: UU5.PropTypes.string,
      maxAge: UU5.PropTypes.number,
      language: UU5.PropTypes.string
    })
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      loginProps: undefined,
      registrationProps: undefined
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _register() {
    let url = this.getLsiItem(Plus4U5.Environment.registrationUrl);
    window.open(url, "_blank");
    return this;
  },

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() && this.isNotAuthenticated() ? (
      <UU5.Bricks.Div {...this.getMainPropsToPass()}>
        <UU5.Bricks.Div className={this.getClassName("text")}>
          <UU5.Bricks.P className={this.getClassName("loginText")} content={this.getLsiValue("loginText")} />
          <UU5.Bricks.P className={this.getClassName("registerText")} content={this.getLsiValue("registerText")} />
        </UU5.Bricks.Div>
        <UU5.Bricks.Div className={this.getClassName("buttons")}>
          <LoginButton {...this.props.loginProps} className={this.getClassName("loginButton")} />
          {AuthenticationService.getServiceName() === "oidcg02" ? (
            <LoginButton
              content={this.getLsiValue("registerButton")}
              {...this.props.registrationProps}
              key="reg-oidcg02"
              className={this.getClassName("registerButton")}
              prompt="registration"
            />
          ) : (
            <UU5.Bricks.Button
              key="reg-plus4u"
              className={this.getClassName("registerButton")}
              onClick={this._register}
              content={this.getLsiValue("registerButton")}
            />
          )}
        </UU5.Bricks.Div>
      </UU5.Bricks.Div>
    ) : null;
  }
  //@@viewOff:render
});

export default Login;
