// uu_plus4u5g01/dist-esm/app/navigation-bar.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import NavigationBarLoader from "./navigation-bar-loader.js";

import "./navigation-bar.less";

export const NavigationBar = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.PureRenderMixin,
    UU5.Common.IdentityMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.NavigationBar",
    nestingLevelList: UU5.Environment.getNestingLevelList("bigBoxCollection", "box"),
    classNames: {
      main: "plus4u5-app-navigation-bar"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    mainEntityUri: UU5.PropTypes.string.isRequired,
    navigationBarUri: UU5.PropTypes.string,
    target: UU5.PropTypes.oneOf(["auto", "_blank", "blank", "_top", "top", "_self", "self", "_parent", "parent"]),
    plus4uBt: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      navigationBarUri: null,
      plus4uBt: false,
      target: "_blank"
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  reload() {
    this._loader && this._loader.reload();
    return this;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _fixUesUri(uri) {
    if (!uri) return uri;
    return uri.substring(0, 4) === "ues:" ? uri : "ues:" + uri;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let child = null;
    let { plus4uBt, target, mainEntityUri, navigationBarUri } = this.props;
    if (this.getNestingLevel() && this.isAuthenticated()) {
      child = (
        <NavigationBarLoader
          {...this.getMainPropsToPass()}
          mainEntityUri={this._fixUesUri(mainEntityUri)}
          navigationBarUri={this._fixUesUri(navigationBarUri)}
          plus4uBt={plus4uBt}
          target={target}
          uuIdentityName={this.getIdentity().name}
          ref_={loader => (this._loader = loader)}
        />
      );
    }

    return child;
  }
  //@@viewOff:render
});

export default NavigationBar;
