// uu_plus4u5g01/dist-esm/app/notification-badge.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import NotificationBadgeLoader from "./notification-badge-loader_.js";

import "./notification-badge.less";

export const NotificationBadge = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.PureRenderMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.IdentityMixin,
    UU5.Common.CcrWriterMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.NotificationBadge",
    nestingLevel: "smallBox",
    classNames: {
      main: "plus4u5-app-notification-badge uu5-noindex"
    },
    opt: {
      nestingLevelWrapper: true,
      ccrKey: "Plus4U5.App.NotificationBadge"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  reload() {
    this._loader && this._loader.reload();
    return this;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let child = null;

    if (this.getNestingLevel() && this.isAuthenticated()) {
      child = (
        <NotificationBadgeLoader
          {...this.getMainPropsToPass()}
          uuIdentity={this.getIdentity().uuIdentity}
          ref_={loader => (this._loader = loader)}
        />
      );
    }

    return child;
  }
  //@@viewOff:render
});

export default NotificationBadge;
