// uu_plus4u5g01/dist-esm/app/page.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import { Session } from "uu_appg01_oidc";

import Top from "./top.js";
import Bottom from "./bottom.js";
import Button from "./button.js";
import SpaError from "./spa-error.js";
import Language from "./language.js";

import "./page.less";

const PAGE_CONTENT = UU5.PropTypes.oneOfType([
  UU5.PropTypes.shape({
    tag: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.element]),
    props: UU5.PropTypes.arrayOf(UU5.PropTypes.object)
  }),
  UU5.PropTypes.node,
  UU5.PropTypes.element,
  UU5.PropTypes.string,
  UU5.PropTypes.number
]); //content (bodyItem, node, element, string, number)

export const Page = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Page",
    nestingLevel: "page",
    classNames: {
      main: "plus4u5-app-page",
      topWrapper: "plus4u5-app-page-top-wrapper",
      bottomWrapper: "plus4u5-app-page-bottom-wrapper",
      leftWrapper: "plus4u5-app-page-left-wrapper",
      rightWrapper: "plus4u5-app-page-right-wrapper",
      contentWrapper: "plus4u5-app-page-content-wrapper",
      systemLayerWrapper: "plus4u5-app-page-system-layer-wrapper",
      appLayerWrapper: "plus4u5-app-page-app-layer-wrapper",
      languageSelector: "plus4u5-app-page-language-selector"
    },
    defaults: {
      systemLayerContent: [<Button key="slc-button" />, <UU5.Bricks.ButtonToTop key="slc-buttontotop" />]
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    displayedLanguages: UU5.PropTypes.oneOfType([UU5.PropTypes.arrayOf(UU5.PropTypes.string), UU5.PropTypes.string]),
    type: UU5.PropTypes.oneOf([
      "0",
      "1",
      "2",
      "3",
      "4",
      "5",
      "6",
      "7",
      "8",
      "9",
      "10",
      "11",
      "12",
      "13",
      "14",
      0,
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14
    ]),
    fullPage: UU5.PropTypes.bool,

    topWrapperProps: UU5.PropTypes.object,
    bottomWrapperProps: UU5.PropTypes.object,
    leftWrapperProps: UU5.PropTypes.object,
    rightWrapperProps: UU5.PropTypes.object,
    contentWrapperProps: UU5.PropTypes.object,
    appLayerWrapperProps: UU5.PropTypes.object,
    systemLayerWrapperProps: UU5.PropTypes.object,

    top: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    bottom: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    left: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    leftOpen: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    leftClosed: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    right: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    rightOpen: PAGE_CONTENT, //content (bodyItem, node, element, string, number)
    rightClosed: PAGE_CONTENT, //content (bodyItem, node, element, string, number)

    alertBus: UU5.PropTypes.oneOfType([
      UU5.PropTypes.shape({
        tag: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.element]),
        props: UU5.PropTypes.arrayOf(UU5.PropTypes.object)
      }),
      UU5.PropTypes.node,
      UU5.PropTypes.element
    ]), //content (bodyItem, node, element)
    modal: UU5.PropTypes.oneOfType([
      UU5.PropTypes.shape({
        tag: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.element]),
        props: UU5.PropTypes.arrayOf(UU5.PropTypes.object)
      }),
      UU5.PropTypes.node,
      UU5.PropTypes.element
    ]), //content (bodyItem, node, element)
    appLayerContent: UU5.PropTypes.any, //content
    systemLayerContent: UU5.PropTypes.any, //content

    leftWidth: UU5.PropTypes.string, // 'xs-0 s-20 m-15 l-15 xl-30' [!|][xs|s|m|l|xl][-nn][-nn]
    rightWidth: UU5.PropTypes.string, // 'xs-0 s-20 m-15 l-15 xl-30'
    loginProps: UU5.PropTypes.shape({
      authenticationContext: UU5.PropTypes.object,
      prompt: UU5.PropTypes.oneOf(["login", "registration"]),
      scope: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
      usePopup: UU5.PropTypes.bool
    })
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      displayedLanguages: null,
      fullPage: true
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return {
      hasError: false
    };
  },

  componentWillMount() {
    this.constructor.page = this;

    this.props.userLayerContent &&
      UU5.Common.Tools.warning("Property 'userLayerContent' is deprecated! Use 'appLayerContent' instead.");
    this.props.userLayerWrapperProps &&
      UU5.Common.Tools.warning("Property 'userLayerWrapperProps' is deprecated! Use 'appLayerWrapperProps' instead.");
  },

  componentWillReceiveProps() {
    this.setState({ errorData: undefined });
  },

  componentDidCatch(error) {
    this.setState({ errorData: { stack: error.stack, message: error.message } });
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _expandWrapperProps(props, className) {
    const wrapperProps = UU5.Common.Tools.mergeDeep({}, props) || {};
    wrapperProps.className = wrapperProps.className || "";
    wrapperProps.className += " " + this.getClassName(className);
    return wrapperProps;
  },

  _getSystemLayerContent() {
    let comps =
      this.props.systemLayerContent === undefined
        ? this.getDefault().systemLayerContent
        : this.props.systemLayerContent;
    if (Array.isArray(comps)) {
      comps = comps.map(comp => {
        let result;
        if (comp && comp.key === "slc-ls")
          result = UU5.Common.Element.clone(comp, { displayedLanguages: this.props.displayedLanguages });
        else if (comp && comp.key === "slc-button" && this.props.loginProps)
          result = UU5.Common.Element.clone(comp, this.props.loginProps);
        else result = comp;
        return result;
      });
    }
    return comps;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    if (this.state.errorData) {
      return <SpaError error={this.state.errorData} />;
    }

    const topWrapperProps = this._expandWrapperProps(this.props.topWrapperProps, "topWrapper");
    const bottomWrapperProps = this._expandWrapperProps(this.props.bottomWrapperProps, "bottomWrapper");
    const leftWrapperProps = this._expandWrapperProps(this.props.leftWrapperProps, "leftWrapper");
    const rightWrapperProps = this._expandWrapperProps(this.props.rightWrapperProps, "rightWrapper");
    const contentWrapperProps = this._expandWrapperProps(this.props.contentWrapperProps, "contentWrapper");
    const appLayerWrapperProps = this._expandWrapperProps(
      this.props.appLayerWrapperProps || this.props.userLayerWrapperProps,
      "appLayerWrapper"
    );
    const systemLayerWrapperProps = this._expandWrapperProps(this.props.systemLayerWrapperProps, "systemLayerWrapper");

    return this.getNestingLevel() === "page" ? (
      <UU5.Common.Session session={Session.currentSession}>
        <UU5.Bricks.Page
          {...this.props}
          {...this.getMainPropsToPass()}
          type={this.props.type || 1}
          fullPage={this.props.fullPage}
          topWrapperProps={topWrapperProps}
          bottomWrapperProps={bottomWrapperProps}
          leftWrapperProps={leftWrapperProps}
          rightWrapperProps={rightWrapperProps}
          contentWrapperProps={contentWrapperProps}
          appLayerWrapperProps={appLayerWrapperProps}
          systemLayerWrapperProps={systemLayerWrapperProps}
          top={this.props.top === undefined ? <Top /> : this.props.top}
          bottom={this.props.bottom === undefined ? <Bottom /> : this.props.bottom}
          left={this.props.left}
          right={this.props.right}
          leftOpen={this.props.leftOpen}
          leftClosed={this.props.leftClosed}
          rightOpen={this.props.rightOpen}
          rightClosed={this.props.rightClosed}
          leftWidth={this.props.leftWidth}
          rightWidth={this.props.rightWidth}
          alertBus={this.props.alertBus === undefined ? <UU5.Bricks.AlertBus /> : this.props.alertBus}
          modal={this.props.modal === undefined ? <UU5.Bricks.Modal /> : this.props.modal}
          systemLayerContent={this._getSystemLayerContent()}
          appLayerContent={this.props.appLayerContent || this.props.userLayerContent}
        >
          {this.props.children}
          <UU5.Bricks.SessionWatch key="sessionWatch" />
        </UU5.Bricks.Page>
      </UU5.Common.Session>
    ) : null;
  }
  //@@viewOff:render
});

Page.getSystemLayerDefault = extraComponents => Page.defaults.systemLayerContent.concat(extraComponents || []);

export default Page;
