// uu_plus4u5g01/dist-esm/app/resources.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";

import "./resources.less";

export const Resources = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.SectionMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Resources",
    classNames: {
      main: props => {
        let className = "plus4u5-app-resources";

        if (props.width) {
          className +=
            " " +
            UU5.Common.Css.css(`
            width: ${props.width}px;
            margin-left: auto;
            margin-right: auto;
          `);
        }

        return className;
      },
      resources: "plus4u5-app-resources-resources",
      resource: "plus4u5-app-resources-resource"
    },
    lsi: () => Plus4U5.Environment.Lsi.App.resources,
    opt: {
      dummyLevel: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    resources: UU5.PropTypes.array,
    width: UU5.PropTypes.number
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      resources: [],
      width: 336
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getResources() {
    return this.props.resources.map((text, i) => (
      <UU5.Bricks.Div className={this.getClassName().resource} key={i} content={text} />
    ));
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let header;
    if (this.props.header === "") {
      header = null;
    } else if (this.props.header) {
      header = this.props.header;
    } else {
      header = this.getLsiComponent("header");
    }

    return (
      <UU5.Bricks.Section
        {...this.getMainPropsToPass()}
        level={this.props.level || 4}
        header={header}
        content={this._getResources()}
      />
    );
  }
  //@@viewOff:render
});

export default Resources;
