// uu_plus4u5g01/dist-esm/app/right.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./right.less";

export const Right = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Right",
    nestingLevel: "boxCollection",
    classNames: {
      main: "plus4u5-app-right"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Div {...this.getMainPropsToPass()}>
        <UU5.Bricks.Button content={<UU5.Bricks.Icon icon="uu-glyphicon-calendar" />} colorSchema="success" />
        <UU5.Bricks.Button content={<UU5.Bricks.Icon icon="glyphicon-envelope" />} colorSchema="primary" />
        <UU5.Bricks.Button content={<UU5.Bricks.Icon icon="glyphicon-bell" />} colorSchema="grey-rich" />
        <UU5.Bricks.Button content={<UU5.Bricks.Icon icon="glyphicon-wrench" />} colorSchema="warning" />
      </UU5.Bricks.Div>
    );
  }
  //@@viewOff:render
});

export default Right;
