// uu_plus4u5g01/dist-esm/app/search-button.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./search-button.less";

export const SearchButton = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.NestingLevelMixin, UU5.Common.CcrReaderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.SearchButton",
    nestingLevel: "smallBox",
    classNames: {
      main: "plus4u5-app-search-button"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onClickHandler() {
    let modal = this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE).getModal();
    if (modal) {
      modal.open({
        header: <UU5.Bricks.Lsi lsi={{ cs: "Hledat", en: "Search" }} />,
        content: "TODO: ....",
        footer: "TODO: ....",
        sticky: false
      });
    }
    return this;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() ? (
      <UU5.Bricks.Button
        {...this.getMainPropsToPass()}
        content={<UU5.Bricks.Icon icon="mdi-magnify" />}
        onClick={this._onClickHandler}
      />
    ) : null;
  }
  //@@viewOff:render
});

export default SearchButton;
