// uu_plus4u5g01/dist-esm/app/spa-not-authenticated.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";
import "uu_plus4u5g01-bricks";
import Button from "./button.js";
import LoginButton from "./login-button.js";

import "./spa-not-authenticated.less";

export const SpaNotAuthenticated = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin, UU5.Common.ColorSchemaMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.SpaNotAuthenticated",
    classNames: {
      main: "plus4u5-app-spanotauthenticated uu5-noindex",
      content: "plus4u5-app-spanotauthenticated-content",
      loginText: "plus4u5-app-spanotauthenticated-login-text",
      helpDesk: "plus4u5-app-spanotauthenticated-help-desk",
      bgTransparent: "plus4u5-app-spanotauthenticated-bg-style-transparent",
      bgFilled: "plus4u5-app-spanotauthenticated-bg-style-filled",
      registerLink: "plus4u5-app-spanotauthenticated-register-link"
    },
    lsi: () => Plus4U5.Environment.Lsi.App.spaNotAuthenticated
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    languageSelectorProps: UU5.PropTypes.object,
    productInfoProps: UU5.PropTypes.object,
    bgStyle: UU5.PropTypes.oneOf(["filled", "transparent"])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      languageSelectorProps: undefined,
      productInfoProps: undefined,
      bgStyle: "filled"
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getMainPropsToPass() {
    let mainProps = this.getMainPropsToPass();
    if (this.props.bgStyle === "transparent") {
      mainProps.className += " " + this.getClassName("bgTransparent");
    } else {
      mainProps.className += " " + this.getClassName("bgFilled");
    }
    return mainProps;
  },

  _getColor() {
    let color;
    if (this.props.bgStyle === "filled" && this.props.colorSchema === "black") {
      color = "white";
    } else if (this.props.bgStyle === "filled" && this.props.colorSchema === "white") {
      color = "black";
    } else if (this.props.colorSchema === "black") {
      color = "black";
    } else {
      color = "white";
    }
    return color;
  },

  _login(loginButton) {
    loginButton.onLoginDefault({ language: this._languageSelector.getLanguage() });
  },
  _registerLanguageSelector(languageSelector) {
    this._languageSelector = languageSelector;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let children = this.getChildren();
    return (
      <UU5.Bricks.Div {...this._getMainPropsToPass()} content={null}>
        <Button />
        <div className={this.getClassName("content")}>
          {UU5.Common.Tools.findComponent("UuProductCatalogue.Bricks.ProductInfo", this.props.productInfoProps)}
          <div className={this.getClassName("loginText")}>
            {this.getLsiComponent("loginText")}
            <br /> <LoginButton onLogin={this._login} colorSchema="green" />
            <br />
            <UU5.Bricks.Link
              colorSchema="custom"
              href="https://www.plus4u.net/registrace/"
              className={this.getClassName("registerLink")}
              content={this.getLsiComponent("registerLink")}
            />
          </div>

          {children && <div>{children}</div>}
          <UU5.Bricks.LanguageSelector
            className="plus4u5-app-spanotauthenticated-earth-icon"
            ref_={this._registerLanguageSelector}
            bgStyle="filled"
            headerMode="label"
            bodyMode="label-code"
            displayedLanguages={["cs", "en"]}
            {...this.props.languageSelectorProps}
            colorSchema="custom"
          />
          <span className={this.getClassName("helpDesk")}>
            {this.getLsiComponent("helpDesk")}
            <UU5.Bricks.Link colorSchema={this._getColor()} content="Helpdesk" href="https://www.plus4u.net/contact" />
          </span>
        </div>
      </UU5.Bricks.Div>
    );
  }
  //@@viewOff:render
});

export default SpaNotAuthenticated;
