// uu_plus4u5g01/dist-esm/app/spa.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import Login from "./login.js";
import Page from "./page.js";

import "./spa.less";

export const Spa = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.IdentityMixin,
    UU5.Common.ContentMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Spa",
    nestingLevel: "spa",
    classNames: {
      main: "plus4u5-app-spa"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    notAuthenticatedContent: UU5.PropTypes.any
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      notAuthenticatedContent: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() ? (
      <Page
        {...this.getMainPropsToPass()}
        type={1}
        name={this.isAuthenticated() ? "authenticated" : "notAuthenticated"}
      >
        {this.isAuthenticated()
          ? UU5.Common.Children.toArray(this.props.children)
          : this.props.notAuthenticatedContent || <Login />}
      </Page>
    ) : null;
  }
  //@@viewOff:render
});

export default Spa;
