// uu_plus4u5g01/dist-esm/app/top.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import Button from "./button.js";
import NotificationBadge from "./notification-badge.js";
import LanguageSelector from "./language-selector.js";

import "./top.less";

export const Top = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [
    UU5.Common.BaseMixin,
    UU5.Common.ElementaryMixin,
    UU5.Common.NestingLevelMixin,
    UU5.Common.CcrReaderMixin,
    UU5.Common.ContentMixin
  ],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.App.Top",
    nestingLevel: "box",
    classNames: {
      main: "plus4u5-app-top uu5-noindex",
      leftMenu: "plus4u5-app-top-left-menu",
      content: "plus4u5-app-top-left-content",
      buttonGroup: "plus4u5-app-top-left-button-group"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    plus4uButton: UU5.PropTypes.bool,
    notificationBadge: UU5.PropTypes.bool,
    languageSelector: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      plus4uButton: false,
      notificationBadge: false,
      languageSelector: true
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentDidMount() {
    if (this.props.plus4uButton) {
      UU5.Common.Tools.warning('Property "plus4uButton" is deprecated! Render the button in system layer instead.');
    }
    if (this.props.notificationBadge) {
      UU5.Common.Tools.warning('Property "notificationBadge" is deprecated! Render the badge in system layer instead.');
    }
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getDynamicContent() {
    let content = [];

    if (
      this.props.languageSelector &&
      Plus4U5.App.Page &&
      Plus4U5.App.Page.page &&
      Plus4U5.App.Page.page.props.displayedLanguages
    ) {
      content.push(
        <LanguageSelector
          key="slc-ls"
          headerMode="code"
          bodyMode="label-code"
          noIndex
          pullRight
          displayedLanguages={Plus4U5.App.Page.page.props.displayedLanguages}
          className="plus4u5-app-page-language-selector"
        />
      );
    }

    if (this.props.plus4uButton) {
      content.push(<Button key="slc-button" />);
    }

    if (this.props.notificationBadge) {
      content.push(<NotificationBadge key="slc-notificationbadge" />);
    }

    return content;
  },

  _getPropsToPass() {
    return UU5.Common.Tools.merge(this.getMainPropsToPass(), { content: null });
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    //TODO - ContainerCollection
    return this.getNestingLevel() ? (
      <UU5.Bricks.Div {...this._getPropsToPass()}>
        <UU5.Bricks.Page.MenuButton menu="left" colorSchema="custom" className={this.getClassName("leftMenu")} />

        {this._getDynamicContent()}
        <UU5.Bricks.Span className={this.getClassName("content")} content={this.props.content || this.props.children} />
      </UU5.Bricks.Div>
    ) : null;
  }
  //@@viewOff:render
});

export default Top;
