// uu_plus4u5filesg01/dist-node/config/lsi.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var Lsi = {
  LSI_LABEL_ADD_BINARY: {
    cs: "Nahrát soubor",
    en: "Upload File"
  },
  LSI_LABEL_ADD_BINARIES: {
    cs: "Nahrát soubory",
    en: "Upload Files"
  },
  LSI_LABEL_UPDATE_BINARY: {
    cs: "Upravit soubor",
    en: "Update File"
  },
  LSI_UPDATE_BINARY_META_DATA_FORM_HEADER: {
    cs: "Upravit tagy souboru",
    en: "Update File Tags"
  },
  LSI_LABEL_DELETE_BINARY: {
    cs: "Smazat soubor",
    en: "Delete File"
  },
  LSI_LABEL_FILE_MANAGER: {
    cs: "Správce souborů",
    en: "File Manager"
  },
  LSI_LABEL_FILE_MANAGER_HEADER: {
    cs: "Soubory",
    en: "Files"
  },
  LSI_LABEL_FILE_MANAGER_SEARCH_PLACEHOLDER: {
    cs: "Hledat soubor...",
    en: "Search file..."
  },
  LSI_LABEL_FILE_MANAGER_DROP_BAR_CONTENT: {
    cs: "Sem přetáhněte soubory k nahrání",
    en: "Drop files to upload here"
  },
  LSI_LABEL_FILE_MANAGER_DROP_ZONE: {
    cs: "Přetáhněte soubory k nahrání",
    en: "Drop files to upload"
  },
  LSI_LABEL_FILE_MANAGER_EMPTY_WARNING: {
    cs: "Úložiště neobsahuje žádné soubory.",
    en: "The storage does not contain any files."
  },
  LSI_LABEL_FILE_MANAGER_COPY_INFO: {
    cs: "Kód byl zkopírován do schránky.",
    en: "Code has been copied to clipboard."
  },
  LSI_LABEL_FILE_MANAGER_ACTION_FULL_TILE: {
    cs: "Zobraz velké dlaždice",
    en: "Show Full Tile"
  },
  LSI_LABEL_FILE_MANAGER_ACTION_SMALL_TILE: {
    cs: "Zobraz kompaktní dlaždice",
    en: "Show Compact Tile"
  },
  LSI_LABEL_FILE_MANAGER_FILTER_HEADER: {
    cs: "Filtry",
    en: "Filters"
  },
  LSI_LABEL_FILE_MANAGER_FILTER_RESET: {
    cs: "Zrušit filtry",
    en: "Reset filters"
  },
  LSI_LABEL_FILE_MANAGER_FILTER_IMAGE: {
    cs: "Obrázky",
    en: "Images"
  },
  LSI_LABEL_FILE_MANAGER_FILTER_TEXT: {
    cs: "Dokumenty",
    en: "Documents"
  },
  LSI_LABEL_FILE_MANAGER_FILTER_ARCHIVE: {
    cs: "Archivy",
    en: "Archives"
  },
  LSI_LABEL_FILE_MANAGER_FILTER_OTHER: {
    cs: "Ostatní",
    en: "Other"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_HEADER: {
    cs: "Sdílet soubor",
    en: "Share File"
  },
  LSI_LABEL_FILE_MANAGER_TILE_DROPDOWN_PREVIEW: {
    cs: "Náhled",
    en: "Preview"
  },
  LSI_LABEL_FILE_MANAGER_TILE_DROPDOWN_SHARE: {
    cs: "Sdílet",
    en: "Share"
  },
  LSI_LABEL_FILE_MANAGER_TILE_DROPDOWN_DOWNLOAD: {
    cs: "Stáhnout",
    en: "Download"
  },
  LSI_LABEL_FILE_MANAGER_TILE_DROPDOWN_UPDATE: {
    cs: "Nahradit",
    en: "Update"
  },
  LSI_LABEL_FILE_MANAGER_TILE_DROPDOWN_UPDATE_TAGS: {
    cs: "Upravit tagy",
    en: "Update Tags"
  },
  LSI_LABEL_FILE_MANAGER_TILE_DROPDOWN_DELETE: {
    cs: "Smazat",
    en: "Delete"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_RADIO_LABEL: {
    cs: "Sdílet",
    en: "Share"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_TYPE_CODE: {
    cs: "Kód",
    en: "Code"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_TYPE_IMAGE: {
    cs: "Komponentu obrázku",
    en: "Image Component"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_TYPE_FILE_LINK: {
    cs: "Komponentu odkazu",
    en: "Link Component"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_TYPE_URI: {
    cs: "URI",
    en: "URI"
  },
  LSI_LABEL_DELETE_BINARY_SUCCESS_CONFIRM_MSG: {
    cs: "Soubor byl úspěšně smazán.",
    en: "The file has been successfully deleted."
  },
  LSI_LABEL_SHARE_BINARY_SUCCESS_CONFIRM_MSG: {
    cs: "Zkopírováno do schránky",
    en: "Copied to clipboard"
  },
  LSI_LABEL_FILE_MANAGER_SHARE_TEXT: {
    cs: "Text ke zkopírování",
    en: "Text to copy"
  },
  LSI_LABEL_FILE_MANAGER_PREVIEW_ERROR: {
    en: "File does not exist. It was probably deleted.",
    cs: "Soubor neexistuje. Nejspíš byl smazán."
  },
  LSI_LABEL_FILE_MANAGER_UPLOAD_RESULT_SUCCESS: {
    en: "All files were successfuly uploaded.",
    cs: "Všechny soubory byly úspěšně nahrány."
  },
  LSI_LABEL_FILE_MANAGER_UPLOAD_RESULT_FAILED: {
    en: "Some files could not be uploaded.",
    cs: "Některé soubory se nepodařilo nahrát."
  },
  LSI_LABEL_ADD_BINARY_CODE: {
    cs: "Kód",
    en: "Code"
  },
  LSI_LABEL_ADD_BINARY_CODE_REQUIRED: {
    cs: "Kód je povinný",
    en: "Code is required"
  },
  LSI_LABEL_ADD_TAGS: {
    cs: "Tagy",
    en: "Tags"
  },
  LSI_LABEL_ADD_BINARY_FILE_SIZE_EXCEEDED: {
    cs: "Soubor je příliš velký. Maximální velikost souboru je ${maxSize}kB.",
    en: "The file is too large. Maximal file size is ${maxSize}kB."
  },
  LSI_LABEL_ADD_BINARY_FILE: {
    cs: "Soubor",
    en: "File"
  },
  LSI_LABEL_ADD_BINARY_FILE_REQUIRED: {
    cs: "Soubor je povinný.",
    en: "File is required."
  },
  LSI_LABEL_ADD_BINARY_ADVANCED: {
    cs: "Rozšířené",
    en: "Advanced"
  },
  LSI_LABEL_DELETE_BINARY_MSG: {
    cs: "Opravdu chcete trvale odstranit soubor ${filename}?",
    en: "Do you really want to delete the file ${filename}?"
  },
  LSI_LABEL_COPY_FILE_MANAGER_LIST_FAILED: {
    en: "An error occurred while receiving data from server.",
    cs: "Nepodařilo se načíst data ze serveru."
  },
  LSI_LABEL_BUTTON_CANCEL: {
    cs: "Zrušit",
    en: "Cancel"
  },
  LSI_LABEL_BUTTON_SAVE: {
    cs: "Uložit",
    en: "Save"
  },
  LSI_LABEL_BUTTON_DELETE: {
    cs: "Smazat",
    en: "Delete"
  },
  LSI_LABEL_BUTTON_COPY: {
    cs: "Kopírovat",
    en: "Copy"
  },
  LSI_INVALID_CODE: {
    cs: "Kód je nevalidní.",
    en: "Code is invalid."
  },
  LSI_INVALID_TAGS: {
    cs: "Tagy $1 nejsou povolené.",
    en: "Set tags $1 are not allowed."
  },
  LSI_DUPLICATE_CODE: {
    en: "Given code is not unique.",
    cs: "Zadaný kód je již obsazen."
  },
  LSI_DEFAULT_ERROR_MESSAGE: {
    en: "Sending failed!",
    cs: "Odeslání selhalo."
  },
  commandErrors: {
    "uu-app-binarystore/uuBinaryCreateBinary/duplicateCode": {
      en: "Given code is not unique.",
      cs: "Zadaný kód je již obsazen."
    },
    "uu-app-binarystore/uuBinaryCreateBinary/invalidDtoIn": {
      en: "Input data is not valid.",
      cs: "Zadaná vstupní data nejsou validní."
    },
    "uu-app-binarystore/uuBinaryDeleteBinary/invalidDtoIn": {
      en: "Input data is not valid.",
      cs: "Zadaná vstupní data nejsou validní."
    },
    resizeFailed: {
      en: "Resizing failed!",
      cs: "Změna velikosti selhala."
    },
    default: {
      en: "Sending failed!",
      cs: "Odeslání selhalo."
    }
  }
};
var _default = Lsi;
exports.default = _default;
