// uu_plus4u5g01/dist-esm/uubmldraw/image/image.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import ImageLoader from "./image-loader.js";
import ImageLegacy from "./image-legacy.js";

export const Image = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.IdentityMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.UuBmlDraw.Image",
    lsi: {
      codeError: {
        cs: "Chybí povinná vlastnost 'code'!",
        en: "Required property 'code' is missing!"
      }
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    code: UU5.PropTypes.string.isRequired
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getCorrectChild() {
    if (!this.props.code) return <UU5.Bricks.Error content={this.getLsiComponent("codeError")} />;

    const identity = this.getIdentity();
    const props = Object.assign({}, this.props, { userIdentity: identity.uuIdentity, userName: identity.name });
    return this.props.tidAwid ? <ImageLoader {...props} /> : <ImageLegacy {...props} />;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getAuthenticatedChild(() => this._getCorrectChild(), { silent: true, inline: true });
  }
  //@@viewOff:render
});

export default Image;
