// uu_plus4u5g01/dist-esm/uubmldraw/image/image-legacy.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";

import "./image-legacy.css";

export let ImageLegacy = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.UuBmlDraw.ImageLegacy_",
    classNames: {
      main: "plus4u5-uubmldraw-image plus4u5-uubmldraw-image-legacy",
      modal: "plus4u5-uubmldraw-image-modal-legacy",
      iframe: "plus4u5-uubmldraw-image-iframe-legacy"
    },
    defaults: {
      width: 250,
      height: 150,
      contentUrl:
        "https://widget.plus4u.net/%s/uu-uubmldraw/getContent?referrerUri=%s&uuid=%s&uuAT=Bearer %s&readOnly=%s",
      sheetUri: "ues:VPH-BT[29537617]:UUAPPKNH/PORTAL[44191581396036005]:50946987294721227[50946987294721227]"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    sheetUri: UU5.PropTypes.string.isRequired,
    code: UU5.PropTypes.string.isRequired,
    readOnly: UU5.PropTypes.bool,
    inProgress: UU5.PropTypes.bool,
    userIdentity: UU5.PropTypes.string,
    userName: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      readOnly: false,
      inProgress: false,
      userIdentity: null,
      userName: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentDidMount() {
    window.uuBMLDraw = window.uuBMLDraw || {};
    uuBMLDraw[this.props.code] = uuBMLDraw[this.props.code] || {};
    uuBMLDraw.initStack = uuBMLDraw.initStack || [];
    UU5.Environment.EventListener.addWindowEvent("message", this.getId(), this._receiveMessage);
  },

  componentWillUnmount() {
    UU5.Environment.EventListener.removeWindowEvent("message", this.getId());
    delete uuBMLDraw[this.props.code];
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  showImage() {
    if (uuBMLDraw[this.props.code] && uuBMLDraw[this.props.code].image) {
      this._modal.open({
        size: "auto",
        content: <UU5.Bricks.Image src={uuBMLDraw[this.props.code].image} />
      });
    }
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getIframeProps() {
    return {
      className: this.getClassName().iframe,
      src: this._buildSrc(),
      inline: true,
      resize: true,
      ref_: iframe => (this._iframe = iframe),
      iframeAttrs: {
        name: this.getId()
      }
    };
  },

  _buildSrc() {
    const sheetUri = this.props.sheetUri || this.getDefault().sheetUri;
    return UU5.Common.Tools.formatString(this.getDefault().contentUrl, [
      this._getTerritoryCode(sheetUri),
      sheetUri,
      this.props.code,
      Plus4U5.Environment.getToken(),
      this.props.readOnly
    ]);
  },

  _getTerritoryCode(uri) {
    let matcher = uri.match(/:([A-Z0-9-]+)/);
    return matcher[1];
  },

  _receiveMessage(event) {
    let message = event.data;

    let widgetData = uuBMLDraw[this.props.code] || {};

    if (message.name === this.getId()) {
      switch (message.type) {
        case "setDimensions":
        case "restoreDimensions":
          this._iframe.setSize(message.width + 1, message.height + 10);
          break;
        case "setImage":
          let img = message.base64;
          widgetData.image = img;
          this._iframe.setSize(message.width + 1, message.height + 10);
          break;
        case "showImages":
          this.showImage();
          break;
        default:
          break;
      }
    }
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <div {...this.getMainAttrs()}>
        <UU5.Bricks.Modal
          parent={this}
          header=" "
          ref_={modal => (this._modal = modal)}
          className={this.getClassName().modal}
        />
        <UU5.Bricks.Iframe {...this._getIframeProps()} />
      </div>
    );
  }
  //@@viewOff:render
});

export default ImageLegacy;
