// uu_plus4u5filesg01/dist-node/forms/autocomplete-items.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var UU5 = _interopRequireWildcard(require("uu5g04"));

require("uu5g04-bricks");

var _config = _interopRequireDefault(require("../config/config.js"));

var _css = _interopRequireDefault(require("../helpers/css.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var ClassNames = {};
var ClassDefs = {
  popover: "\n    width: 100%;\n    max-height: 350px;\n    overflow-y: auto;\n  ",
  link: "\n    && {\n      color: rgba(0,0,0, .87);\n      display: block;\n      padding: 8px 16px;\n      width: 100%;\n  \n      &:hover {\n        text-decoration: none;\n        background-color: #E0E0E0;\n      }\n    }\n  "
};

var getClassName = function getClassName(className) {
  if (!ClassNames[className]) {
    ClassNames[className] = _css.default.css(ClassDefs[className]);
  }

  return ClassNames[className];
};

var AutocompleteItems = function AutocompleteItems(props) {
  // no items to shown => render null
  if (!props.shown || !Array.isArray(props.items) || !props.items.length) return null;
  var items = props.items; // filter items by filterValue

  if (props.filterValue) {
    items = items.filter(function (item) {
      return item.indexOf(props.filterValue) !== -1;
    });
  } // filter items by filterItems


  if (Array.isArray(props.filterItems)) {
    items = items.filter(function (item) {
      return props.filterItems.indexOf(item) === -1;
    });
  } // validate if is something to show


  if (items.length === 0) {
    return null;
  }

  return UU5.Common.Element.create(UU5.Bricks.Popover, {
    shown: true,
    fitHeightToViewport: true,
    forceRender: true,
    className: getClassName("popover")
  }, items.map(function (item) {
    return UU5.Common.Element.create(UU5.Bricks.Link, {
      className: getClassName("link"),
      colorSchema: "default",
      key: item,
      mainAttrs: {
        onMouseDown: function onMouseDown(e) {
          e.preventDefault();
          props.onClick(item);
        }
      }
    }, item);
  }));
};

AutocompleteItems.displayName = _config.default.TAG + "AutocompleteItems";
AutocompleteItems.tagName = _config.default.TAG + "AutocompleteItems";
AutocompleteItems.propTypes = {
  items: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
  filterValue: UU5.PropTypes.string,
  filterItems: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
  shown: UU5.PropTypes.bool,
  onClick: UU5.PropTypes.func
};
AutocompleteItems.defaultProps = {
  filterValue: "",
  shown: false
};
var _default = AutocompleteItems;
exports.default = _default;
