// uu_plus4u5filesg01/dist-node/forms/controls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var UU5 = _interopRequireWildcard(require("uu5g04"));

var _config = _interopRequireDefault(require("../config/config.js"));

var _lsi = _interopRequireDefault(require("../config/lsi.js"));

var _lsiHelper = _interopRequireDefault(require("../helpers/lsi-helper.js"));

require("./controls.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var Controls = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "Controls",
    classNames: {
      main: _config.default.CSS + "controls",
      codeButtons: _config.default.CSS + "controls-code-buttons"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    formComponent: UU5.PropTypes.func,
    hasDelete: UU5.PropTypes.bool,
    label: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.object])
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      hasDelete: false
    };
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    return {
      disabledButton: false
    };
  },
  //@@viewOff:standardComponentLifeCycle
  //@@viewOn:interface
  setDisabledButton: function setDisabledButton(b) {
    this.setState({
      disabledButton: b
    });
  },
  //@@viewOff:interface
  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods
  //@@viewOn:componentSpecificHelpers
  _save: function _save(e) {
    if (this.props.formComponent() && this.props.formComponent().getForm() && this.props.formComponent().getForm().isValid()) {
      this.props.formComponent() && this.props.formComponent().getForm() && this.props.formComponent().getForm().save();
    }

    return this;
  },
  _getMainPropsToPass: function _getMainPropsToPass() {
    var props = this.getMainPropsToPass();
    props.disabled = false;
    return props;
  },
  _getButtons: function _getButtons() {
    var _this = this;

    var buttonCancelComp = UU5.Common.Element.create(UU5.Bricks.Button, {
      onClick: function onClick() {
        return _this.props.formComponent().getForm() && _this.props.formComponent().getForm().cancel();
      },
      content: _lsiHelper.default.getLsiComponent(_lsi.default.LSI_LABEL_BUTTON_CANCEL),
      disabled: this.isDisabled(),
      key: "Cancel"
    });
    var buttonOkComp = UU5.Common.Element.create(UU5.Bricks.Button, {
      colorSchema: this.props.hasDelete ? "danger" : "primary",
      onClick: this._save,
      content: _lsiHelper.default.getLsiComponent(this.props.label ? this.props.label : this.props.hasDelete ? _lsi.default.LSI_LABEL_BUTTON_DELETE : _lsi.default.LSI_LABEL_BUTTON_SAVE),
      disabled: this.isDisabled() || this.state.disabledButton,
      key: "Submit"
    });
    return [buttonCancelComp, buttonOkComp];
  },
  //@@viewOff:componentSpecificHelpers
  //@@viewOn:render
  render: function render() {
    return UU5.Common.Element.create(UU5.Common.Div, this._getMainPropsToPass(), this._getButtons());
  } //@@viewOff:render

});
var _default = Controls;
exports.default = _default;
