// uu_plus4u5filesg01/dist-node/forms/delete-file-form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var UU5 = _interopRequireWildcard(require("uu5g04"));

var _config = _interopRequireDefault(require("../config/config.js"));

var _calls = _interopRequireDefault(require("../calls.js"));

var _lsi = _interopRequireDefault(require("../config/lsi.js"));

var _lsiHelper = _interopRequireDefault(require("../helpers/lsi-helper.js"));

require("./delete-file-form.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DeleteFileForm = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.CcrReaderMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "DeleteFileForm",
    classNames: {
      main: _config.default.CSS + "delete-file-form"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    done: UU5.PropTypes.func,
    onSuccess: UU5.PropTypes.func,
    onCancel: UU5.PropTypes.func,
    binary: UU5.PropTypes.object,
    thumbnail: UU5.PropTypes.object,
    removeThumbnail: UU5.PropTypes.func,
    baseUri: UU5.PropTypes.string,
    cmdDeleteBinary: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    cmdDeleteThumbnail: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    shouldSaveThumbnail: UU5.PropTypes.bool,
    lsiLabels: UU5.PropTypes.object,
    currentControls: UU5.PropTypes.func
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle
  //@@viewOn:interface
  getForm: function getForm() {
    return this._form;
  },
  //@@viewOff:interface
  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods
  //@@viewOn:componentSpecificHelpers
  _onSave: function _onSave() {
    this.props.currentControls() && this.props.currentControls().setDisabledButton(true);
    var done = this.props.shouldSaveThumbnail && this.props.thumbnail ? this._deleteThumbnail : this._saveDone;
    var fail = this._sendFail;
    var data = {
      code: this.props.binary.code
    };

    if (typeof this.props.cmdDeleteBinary === "function") {
      this.props.cmdDeleteBinary(data).then(done, fail);
    } else {
      var dtoIn = {
        data: data,
        done: done,
        fail: fail
      };

      _calls.default.customPostCmd(this.props.baseUri, this.props.cmdDeleteBinary, dtoIn);
    }
  },
  _deleteThumbnail: function _deleteThumbnail(dtoOut) {
    var _this = this;

    var dtoIn = {
      done: function done() {
        _this.props.removeThumbnail(_this.props.thumbnail.code);

        _this._saveDone(dtoOut);
      },
      fail: function fail() {
        UU5.Common.Tools.warning("Delete image thumbnail failed.");

        _this._saveDone(dtoOut);
      },
      data: {
        code: this.props.thumbnail.code
      }
    };

    if (typeof this.props.cmdDeleteThumbnail === "function") {
      this.props.cmdDeleteThumbnail(dtoIn.data).then(dtoIn.done, dtoIn.fail);
    } else {
      _calls.default.customPostCmd(this.props.baseUri, this.props.cmdDeleteThumbnail, dtoIn);
    }
  },
  _sendFail: function _sendFail(opt) {
    this.props.currentControls() && this.props.currentControls().setDisabledButton(false);
    var msg = null;

    if (opt.error && opt.error["cause"] && opt.error["cause"]["uuAppErrorMap"]["uu-app-objectstore/invalidRevision"]) {
      var code = opt.error["code"] + "/invalidRevision";
      msg = _lsiHelper.default.getLsiComponent(_lsi.default.commandErrors[code] || _lsi.default.commandErrors["default"]);
    } else {
      msg = _lsiHelper.default.getLsiComponent(_lsi.default.commandErrors[opt.error["code"]] || _lsi.default.commandErrors["default"]);
    }

    this._form.getAlertBus().setAlert({
      content: msg,
      colorSchema: "danger"
    });
  },
  _saveDone: function _saveDone(dtoOut, justSave) {
    var _this2 = this;

    this.props.onSuccess && this.props.onSuccess(this.props.binary.code);
    this.props.done && this.props.done(dtoOut, justSave);
    this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE) && this.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE).getAlertBus().clearAlerts(function () {
      _this2.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE) && _this2.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE).getAlertBus && _this2.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE).getAlertBus() && _this2.getCcrComponentByKey(UU5.Environment.CCRKEY_PAGE).getAlertBus().setAlert({
        content: _lsiHelper.default.getLsiComponent(_lsi.default.LSI_LABEL_DELETE_BINARY_SUCCESS_CONFIRM_MSG),
        colorSchema: "success",
        closeTimer: 3000
      });
    });
  },
  _onValidate: function _onValidate(component) {},
  _getChild: function _getChild() {
    var _this3 = this;

    var content;

    if (typeof this.props.lsiLabels.deleteBinaryFormContent === "string") {
      content = this.props.lsiLabels.deleteBinaryFormContent;
    } else {
      content = _objectSpread({}, this.props.lsiLabels.deleteBinaryFormContent);
    }

    if (typeof content === "string") {
      content = content.replace("${filename}", this.props.binary.filename);
    } else if (_typeof(content) === "object") {
      for (var key in content) {
        if (content.hasOwnProperty(key)) {
          content[key] = content[key].replace("${filename}", this.props.binary.filename);
        }
      }
    }

    var text = _lsiHelper.default.getLsiComponent(content);

    return UU5.Common.Element.create(UU5.Forms.Form, _extends({}, this.getMainPropsToPass(), {
      onSave: function onSave(_ref) {
        var component = _ref.component,
            values = _ref.values;
        return _this3._onSave(component, values);
      },
      onCancel: this.props.onCancel,
      ref_: function ref_(form) {
        return _this3._form = form;
      }
    }), UU5.Common.Element.create(UU5.Bricks.P, _extends({}, this.getMainPropsToPass(), {
      content: text
    })));
  },
  //@@viewOff:componentSpecificHelpers
  //@@viewOn:render
  render: function render() {
    return this._getChild();
  } //@@viewOff:render

});
var _default = DeleteFileForm;
exports.default = _default;
