// uu_plus4u5filesg01/dist-node/forms/preview-file-form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var UU5 = _interopRequireWildcard(require("uu5g04"));

var _calls = _interopRequireDefault(require("../calls.js"));

var _config = _interopRequireDefault(require("../config/config.js"));

var _lsi = _interopRequireDefault(require("../config/lsi.js"));

var _lsiHelper = _interopRequireDefault(require("../helpers/lsi-helper.js"));

require("./preview-file-form.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var PreviewFileForm = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "PreviewFileForm",
    classNames: {
      main: _config.default.CSS + "preview-file-form"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    code: UU5.PropTypes.string,
    done: UU5.PropTypes.func,
    onCancel: UU5.PropTypes.func,
    rev: UU5.PropTypes.oneOfType([UU5.PropTypes.string, UU5.PropTypes.number]),
    baseUri: UU5.PropTypes.string,
    cmdGetBinaryData: UU5.PropTypes.string
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {
    if (this.props.cmdGetBinaryData && this.props.code) {
      this._updateImageSrc();
    }

    return {
      error: false,
      src: null
    };
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.cmdGetBinaryData !== this.props.cmdGetBinaryData || nextProps.code !== this.props.code) {
      this._updateImageSrc(nextProps);
    }
  },
  //@@viewOff:standardComponentLifeCycle
  //@@viewOn:interface
  getForm: function getForm() {
    return this._form;
  },
  hasUnsavedChanges: function hasUnsavedChanges() {
    return false;
  },
  //@@viewOff:interface
  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods
  //@@viewOn:componentSpecificHelpers
  _updateImageSrcDone: function _updateImageSrcDone(data) {
    this.setState(function (state) {
      if (state.src && state.src.match(/^blob:/)) {
        // release old image url
        window.URL.revokeObjectURL(state.src);
      } // create new url to blob


      return {
        src: data.match(/^blob:/) ? window.Url.createObjectURL(data) : data
      };
    });
  },
  _updateImageSrc: function _updateImageSrc() {
    var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.props;
    var src;
    var data = {
      code: props.code
    };

    if (typeof props.cmdGetBinaryData === "function") {
      props.cmdGetBinaryData(data).then(this._updateImageSrcDone, this._updateImageSrcDone);
    } else {
      src = _calls.default.getCommandUri(props.baseUri, props.cmdGetBinaryData, data);
      src = src + "&" + props.rev;

      this._updateImageSrcDone(src);
    }
  },
  _onSave: function _onSave() {
    this.props.done && this.props.done();
  },
  _renderImage: function _renderImage() {
    var _this = this;

    return UU5.Common.Element.create(UU5.Bricks.Image, {
      mainAttrs: {
        onError: function onError() {
          return _this.setState({
            error: true
          });
        }
      },
      responsive: true,
      src: this.state.src,
      authenticate: true
    });
  },
  _renderError: function _renderError() {
    return UU5.Common.Element.create(UU5.Bricks.Div, {
      content: _lsiHelper.default.getLsiComponent(_lsi.default.LSI_LABEL_FILE_MANAGER_PREVIEW_ERROR)
    });
  },
  _getChild: function _getChild() {
    return UU5.Common.Element.create(UU5.Bricks.Div, this.getMainPropsToPass(), this.state.error ? this._renderError() : this._renderImage());
  },
  //@@viewOff:componentSpecificHelpers
  //@@viewOn:render
  render: function render() {
    return this._getChild();
  } //@@viewOff:render

});
var _default = PreviewFileForm;
exports.default = _default;
