// uu_plus4u5filesg01/dist-node/forms/update-file-form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var UU5 = _interopRequireWildcard(require("uu5g04"));

var _config = _interopRequireDefault(require("../config/config.js"));

var _calls = _interopRequireDefault(require("../calls.js"));

var _lsi = _interopRequireDefault(require("../config/lsi.js"));

var _lsiHelper = _interopRequireDefault(require("../helpers/lsi-helper.js"));

var _file = require("../helpers/file.js");

require("./update-file-form.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var UpdateFileForm = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "UpdateFileForm",
    classNames: {
      main: _config.default.CSS + "update-file-form"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    done: UU5.PropTypes.func,
    onSuccess: UU5.PropTypes.func,
    onCancel: UU5.PropTypes.func,
    addThumbnail: UU5.PropTypes.func,
    removeThumbnail: UU5.PropTypes.func,
    thumbnail: UU5.PropTypes.object,
    binary: UU5.PropTypes.object,
    file: UU5.PropTypes.object,
    baseUri: UU5.PropTypes.string,
    cmdUpdateBinaryData: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    cmdCreateThumbnail: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    cmdUpdateThumbnailData: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    cmdDeleteThumbnail: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    shouldSaveThumbnail: UU5.PropTypes.bool,
    lsiLabels: UU5.PropTypes.object,
    currentControls: UU5.PropTypes.func,
    outputMaxHeight: UU5.PropTypes.number,
    outputMaxWidth: UU5.PropTypes.number,
    outputQuality: UU5.PropTypes.number
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:standardComponentLifeCycle
  getInitialState: function getInitialState() {},
  //@@viewOff:standardComponentLifeCycle
  //@@viewOn:interface
  getForm: function getForm() {
    return this._form;
  },
  //@@viewOff:interface
  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods
  //@@viewOn:componentSpecificHelpers
  _callUpdateBinary: function _callUpdateBinary(values, imageFile) {
    var isImage = values.data.type && values.data.type.match(/image\//);
    var done;

    if (!this.props.shouldSaveThumbnail || !isImage && !this.props.thumbnail) {
      done = this._saveDone;
    } else if (!isImage && this.props.thumbnail) {
      done = this._deleteThumbnail;
    } else {
      done = this._saveThumbnail;
    }

    var fail = this._sendFail;
    var data = {
      code: this.props.binary.code,
      data: imageFile
    };
    data.sys = {
      rev: this.props.binary.sys.rev
    };

    if (typeof this.props.cmdUpdateBinaryData === "function") {
      this.props.cmdUpdateBinaryData(data).then(done, fail);
    } else {
      var dtoIn = {
        data: data,
        done: done,
        fail: fail
      };

      _calls.default.customPostMultipartCmd(this.props.baseUri, this.props.cmdUpdateBinaryData, dtoIn);
    }
  },
  _onSave: function _onSave() {
    var _this = this;

    var values = this._form.getValues();

    (0, _file.adjustFile)(values.data, this.props).then(function (file) {
      _this._callUpdateBinary(values, file);
    }, function (error) {
      _this._sendFail(error);
    });
  },
  _saveThumbnail: function _saveThumbnail(dtoOut) {
    var _this2 = this;

    var values = this._form.getValues();

    (0, _file.adjustFile)(values.data, {
      outputMaxHeight: 512,
      outputMaxWidth: 512,
      outputQuality: 0.85
    }).then(function (file) {
      var dtoIn = {
        done: function done(thumbnailDtoOut) {
          _this2.props.addThumbnail(thumbnailDtoOut);

          _this2._saveDone(dtoOut);
        },
        fail: function fail() {
          UU5.Common.Tools.warning("Saving image thumbnail failed.");

          _this2._saveDone(dtoOut);
        },
        data: {
          code: (dtoOut.code.length > 61 ? dtoOut.code.substr(0, 61) : dtoOut.code) + "_th",
          data: file
        }
      };

      if (_this2.props.thumbnail) {
        dtoIn.data.sys = {
          rev: _this2.props.thumbnail.sys.rev
        };

        if (typeof _this2.props.cmdUpdateThumbnailData === "function") {
          _this2.props.cmdUpdateThumbnailData(dtoIn.data).then(dtoIn.done, dtoIn.fail);
        } else {
          _calls.default.customPostMultipartCmd(_this2.props.baseUri, _this2.props.cmdUpdateThumbnailData, dtoIn);
        }
      } else {
        if (typeof _this2.props.cmdCreateThumbnail === "function") {
          _this2.props.cmdCreateThumbnail(dtoIn.data).then(dtoIn.done, dtoIn.fail);
        } else {
          if (typeof _this2.props.cmdCreateThumbnail === "function") {
            _this2.props.cmdCreateThumbnail(dtoIn.data).then(dtoIn.done, dtoIn.fail);
          } else {
            _calls.default.customPostMultipartCmd(_this2.props.baseUri, _this2.props.cmdCreateThumbnail, dtoIn);
          }
        }
      }
    });
  },
  _deleteThumbnail: function _deleteThumbnail(dtoOut) {
    var _this3 = this;

    var dtoIn = {
      done: function done() {
        _this3.props.removeThumbnail(_this3._getSafeThumbnailCode(_this3.props.binary.code));

        _this3._saveDone(dtoOut);
      },
      fail: function fail() {
        UU5.Common.Tools.warning("Delete image thumbnail failed.");

        _this3._saveDone(dtoOut);
      },
      data: {
        code: this._getSafeThumbnailCode(this.props.binary.code)
      }
    };

    if (typeof this.props.cmdDeleteThumbnail === "function") {
      this.props.cmdDeleteThumbnail(dtoIn.data).then(dtoIn.done, dtoIn.fail);
    } else {
      _calls.default.customPostCmd(this.props.baseUri, this.props.cmdDeleteThumbnail, dtoIn);
    }
  },
  _getSafeThumbnailCode: function _getSafeThumbnailCode(code) {
    return (code.length > 61 ? code.substr(0, 61) : code) + "_th";
  },
  _sendFail: function _sendFail(opt) {
    this.props.currentControls() && this.props.currentControls().setDisabledButton(false);
    var msg = null;

    if (!opt.error) {
      msg = opt.toString();
    } else if (opt.error && opt.error["cause"] && opt.error["cause"]["uuAppErrorMap"]["uu-app-objectstore/invalidRevision"]) {
      var code = opt.error["code"] + "/invalidRevision";
      msg = _lsiHelper.default.getLsiComponent(_lsi.default.commandErrors[code] || _lsi.default.commandErrors["default"]);
    } else {
      msg = _lsiHelper.default.getLsiComponent(_lsi.default.commandErrors[opt.error["code"]] || _lsi.default.commandErrors["default"]);
    }

    this._form.getAlertBus().setAlert({
      content: msg,
      colorSchema: "danger"
    });
  },
  _saveDone: function _saveDone(dtoOut, justSave) {
    this.props.onSuccess && this.props.onSuccess(dtoOut);
    this.props.done && this.props.done(dtoOut, justSave);

    this._fileForm.disable();
  },
  _registerFileInput: function _registerFileInput(input) {
    this._fileForm = input;

    this._validateFileSize({
      value: this.props.file
    });
  },
  _validateFileSize: function _validateFileSize(_ref) {
    var _this4 = this;

    var value = _ref.value;
    var props = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : this.props;

    if (value) {
      // check file size after adjustment
      (0, _file.adjustFile)(value, props).then(function (file) {
        if (file && file.size > (0, _file.getMaxFileSize)(props.sizeLimit) * 1024) {
          _this4._fileForm.setError(_lsiHelper.default.getLsiComponent(props.lsiLabels.fileSizeExceededError, {
            maxSize: (0, _file.getMaxFileSize)(props.sizeLimit) + "KB"
          }), value);
        } else {
          _this4._fileForm.setInitial(null, value);
        }
      });
    } else {
      this._fileForm.setInitial(null, value);
    }
  },
  _getChild: function _getChild() {
    var _this5 = this;

    return UU5.Common.Element.create(UU5.Forms.Form, _extends({}, this.getMainPropsToPass(), {
      onSave: function onSave(_ref2) {
        var component = _ref2.component,
            values = _ref2.values;
        return _this5._onSave(component, values);
      },
      onCancel: this.props.onCancel,
      ref_: function ref_(form) {
        return _this5._form = form;
      }
    }), UU5.Common.Element.create(UU5.Forms.File, {
      label: _lsiHelper.default.getLsiComponent(this.props.lsiLabels.addBinaryFormFile),
      controlled: false,
      required: true,
      name: "data",
      requiredMessage: _lsiHelper.default.getLsiComponent(this.props.lsiLabels.addBinaryFormFileRequired),
      ref_: this._registerFileInput,
      onChange: this._validateFileSize
    }));
  },
  //@@viewOff:componentSpecificHelpers
  //@@viewOn:render
  render: function render() {
    return this._getChild();
  } //@@viewOff:render

});
var _default = UpdateFileForm;
exports.default = _default;
