// uu_plus4u5filesg01/dist-node/forms/update-binary-meta-data-form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var UU5 = _interopRequireWildcard(require("uu5g04"));

var _config = _interopRequireDefault(require("../config/config.js"));

var _calls = _interopRequireDefault(require("../calls.js"));

var _lsi = _interopRequireDefault(require("../config/lsi.js"));

var _lsiHelper = _interopRequireDefault(require("../helpers/lsi-helper.js"));

var _tagsSelect = _interopRequireDefault(require("./tags-select.js"));

require("./add-file-form.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

var UpdateBinaryMetaDataForm = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "UpdateBinaryForm",
    classNames: {
      main: _config.default.CSS + "update-binary-form"
    },
    opt: {
      nestingLevelWrapper: true
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    done: UU5.PropTypes.func,
    onSuccess: UU5.PropTypes.func,
    onCancel: UU5.PropTypes.func,
    binary: UU5.PropTypes.object.isRequired,
    baseUri: UU5.PropTypes.string,
    cmdUpdateBinary: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]).isRequired,
    currentControls: UU5.PropTypes.func.isRequired,
    ignoreTags: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
    availableTags: UU5.PropTypes.arrayOf(UU5.PropTypes.string),
    allowCustomTags: UU5.PropTypes.bool,
    onTagCreate: UU5.PropTypes.func
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle
  //@@viewOn:interface
  getForm: function getForm() {
    return this._form;
  },
  //@@viewOff:interface
  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods
  //@@viewOn:componentSpecificHelpers
  _callUpdateBinary: function _callUpdateBinary(values) {
    var data = {
      code: this.props.binary.code,
      tagList: values.tagList || [],
      sys: this.props.binary.sys
    };
    var done = this._saveDone;
    var fail = this._sendFail;

    if (typeof this.props.cmdUpdateBinary === "function") {
      this.props.cmdUpdateBinary(data).then(done, fail);
    } else {
      var dtoIn = {
        data: data,
        done: done,
        fail: fail
      };

      _calls.default.customPostCmd(this.props.baseUri, this.props.cmdUpdateBinary, dtoIn);
    }
  },
  _onSave: function _onSave(_ref) {
    var component = _ref.component,
        values = _ref.values;
    this.props.currentControls() && this.props.currentControls().setDisabledButton(true);

    this._callUpdateBinary(values);
  },
  _sendFail: function _sendFail(opt) {
    this.props.currentControls() && this.props.currentControls().setDisabledButton(false);
    var msg = null;

    if (!opt.error) {
      msg = opt.toString();
    } else if (opt.error && opt.error["cause"] && opt.error["cause"]["uuAppErrorMap"]["uu-app-objectstore/invalidRevision"]) {
      var code = opt.error["code"] + "/invalidRevision";
      msg = _lsiHelper.default.getLsiComponent(_lsi.default.commandErrors[code] || _lsi.default.commandErrors["default"]);
    } else {
      msg = _lsiHelper.default.getLsiComponent(_lsi.default.commandErrors[opt.error["code"]] || _lsi.default.commandErrors["default"]);
    }

    this._form.getAlertBus().setAlert({
      content: msg,
      colorSchema: "danger"
    });
  },
  _saveDone: function _saveDone(dtoOut, justSave) {
    // check new tags and call onTagCreate
    if (this.props.onTagCreate && dtoOut.tagList) {
      var newTags = [];
      var originalTagList = this.props.binary.tagList || [];
      var availableTags = this.props.availableTags || [];

      if (!originalTagList && !availableTags.length) {
        newTags = _toConsumableArray(dtoOut.tagList);
      } else {
        dtoOut.tagList.forEach(function (tag) {
          if (originalTagList.indexOf(tag) === -1 && availableTags.indexOf(tag) === -1) {
            newTags.push(tag);
          }
        });
      }

      if (newTags.length) {
        this.props.onTagCreate(newTags);
      }
    }

    this.props.onSuccess && this.props.onSuccess(dtoOut);
    this.props.done && this.props.done(dtoOut, justSave);
  },
  //@@viewOff:componentSpecificHelpers
  //@@viewOn:render
  render: function render() {
    var _this = this;

    return UU5.Common.Element.create(UU5.Forms.Form, _extends({}, this.getMainPropsToPass(), {
      onSave: this._onSave,
      onCancel: this.props.onCancel,
      ref_: function ref_(form) {
        return _this._form = form;
      },
      spacing: 8
    }), UU5.Common.Element.create(_tagsSelect.default, {
      label: UU5.Common.Element.create(UU5.Bricks.Lsi, {
        lsi: _lsi.default.LSI_LABEL_ADD_TAGS
      }),
      labelAlignment: "left",
      availableTags: this.props.availableTags,
      ignoreTags: this.props.ignoreTags,
      allowCustomTags: this.props.allowCustomTags,
      name: "tagList",
      value: this.props.binary.tagList,
      labelColWidth: "xs-12",
      inputColWidth: "xs-12",
      tagListPosition: "bottom"
    }));
  } //@@viewOff:render

});
var _default = UpdateBinaryMetaDataForm;
exports.default = _default;
