// uu5imagingg01/dist-esm/imaging/internal/image-loading.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./image-loading.less";

import Cfg from "../../config/config.js";

export const ImageLoading = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.TAG + "ImageLoading",
    classNames: {
      main: Cfg.CSS + "image-loading"
    },
    opt: {
      pureRender: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    value: UU5.PropTypes.number,
    iconAttrs: UU5.PropTypes.object
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      value: 0,
      iconAttrs: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let iconAttrs = UU5.Common.Tools.merge(this.props.iconAttrs, { icon: "mdi-window-close" });

    return (
      <div className="radial-progress" data-progress={this.props.value}>
        <div className="circle">
          <div className="mask full">
            <div className="fill" />
          </div>
          <div className="mask half">
            <div className="fill" />
            <div className="fill fix" />
          </div>
          <div className="shadow" />
        </div>
        <div className="inset">
          <UU5.Bricks.Icon {...iconAttrs} />
        </div>
      </div>
    );
  }
  //@@viewOn:render
});

export default ImageLoading;
