// uu5imagingg01/dist-esm/imaging/internal/image-input-item.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./image-input-item.less";

// import Loading from "./image-loading.js";
import Cfg from "../../config/config.js";

const TRANSPARENT_IMAGE_DATA_URI = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7";

export const ImageInputItem = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.TAG + "ImageInputItem",
    classNames: {
      main: Cfg.CSS + "image-input-item",
      image: Cfg.CSS + "image-input-item-image",
      imageWrapper: Cfg.CSS + "image-input-item-image-wrapper",
      buttonBar: Cfg.CSS + "image-input-item-button-bar",
      name: Cfg.CSS + "image-input-item-name",
      nameError: Cfg.CSS + "image-input-item-name-error",
      error: Cfg.CSS + "image-input-item-error",
      errorIcon: Cfg.CSS + "image-input-item-error-icon",
      loading: Cfg.CSS + "image-input-item-loading",
      loadingComponent: Cfg.CSS + "image-input-item-loading-component"
    },
    defaults: {
      aspectRatioShapesRegexp: /^((rounded)?square|circle|(rounded)?rect\d+x\d+)$/i
    },
    opt: {
      pureRender: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    imageSrc: UU5.PropTypes.string,
    imageName: UU5.PropTypes.string,
    imageId: UU5.PropTypes.string,
    onRemove: UU5.PropTypes.func,
    onAbort: UU5.PropTypes.func,
    progress: UU5.PropTypes.number,
    readOnly: UU5.PropTypes.bool,
    error: UU5.PropTypes.any,
    previewWidth: UU5.PropTypes.number,
    previewHeight: UU5.PropTypes.number,
    previewShape: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      imageSrc: null,
      imageName: null,
      onRemove: null,
      onAbort: null,
      progress: null,
      readOnly: null,
      error: null,
      previewWidth: null,
      previewHeight: null,
      previewShape: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getButtonBar() {
    return !this.props.readOnly ? (
      <div className={this.getClassName("buttonBar")}>
        <UU5.Bricks.Button
          onClick={() => this.props.onRemove(this.props.imageId)}
          content={<UU5.Bricks.Icon icon="mdi-delete" />}
        />
      </div>
    ) : null;
  },

  _getLoading() {
    return (
      // <Loading value={this.props.progress} iconAttrs={{ mainAttrs: { onClick: this.props.onAbort } }} />
      <UU5.Bricks.Loading className={this.getClassName("loadingComponent")} />
    );
  },

  _getError() {
    return (
      <div className={this.getClassName("error")}>
        <UU5.Bricks.Icon icon="mdi-alert-circle" className={this.getClassName("errorIcon")} />
      </div>
    );
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { imageSrc, progress, imageName, error, previewShape } = this.props;
    let isLoading = !error && typeof progress === "number" && progress < 100;
    let mainAttrs = this.getMainAttrs();
    if (isLoading) mainAttrs.className = (mainAttrs.className || "") + " " + this.getClassName("loading");
    let width = this.props.previewWidth;
    let height = this.props.previewHeight;
    if (height == null && previewShape && !previewShape.match(this.getDefault("aspectRatioShapesRegexp"))) {
      height = width;
    }
    return (
      <span {...mainAttrs} style={{ width }}>
        <div className={this.getClassName("imageWrapper")}>
          {(imageSrc || error) && (
            <UU5.Imaging.Image
              authenticate
              blackBackground={false}
              src={imageSrc || TRANSPARENT_IMAGE_DATA_URI}
              className={this.getClassName("image")}
              width={width}
              height={height}
              shape={this.props.previewShape}
              fit="coverNoScaleUp"
            />
          )}
          {error && this._getError()}
          {this._getButtonBar()}
          {isLoading ? this._getLoading() : null}
        </div>
        <div
          className={this.getClassName("name") + (error ? " " + this.getClassName("nameError") : "")}
          title={imageName}
        >
          {imageName || "\xa0"}
        </div>
      </span>
    );
  }
  //@@viewOn:render
});

export default ImageInputItem;
