// uu_contentkitg01/dist-esm/tables/table.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import Tools from "../core/tools.js";
import AbstractTable from "./abstract-table.js";
import Lsi from "../config/lsi.js";

import Cfg from "../core/_config.js";
import "./table.less";

const Table = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.NestingLevelMixin, UU5.Common.SectionMixin, UU5.Common.EditableMixin],

  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: Cfg.tag("Tables.Table"),
    nestingLevelList: UU5.Environment.getNestingLevelList("bigBoxCollection", "box"),
    classNames: {
      main: `${Cfg.CSS} ${Cfg.css("table")}`,
      table: Cfg.css("table-cover-table"),
      auto: Cfg.css("table-auto"),
      fixed: Cfg.css("table-fixed")
    },
    defaults: {
      editableComponent: Cfg.tag("TablesEditable.Table")
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    data: UU5.PropTypes.arrayOf(UU5.PropTypes.arrayOf(UU5.PropTypes.any)),
    transpose: UU5.PropTypes.bool,
    responsive: UU5.PropTypes.bool,
    rowHeader: UU5.PropTypes.bool,
    colHeader: UU5.PropTypes.bool,
    colWidth: UU5.PropTypes.arrayOf(UU5.PropTypes.number),
    header: UU5.PropTypes.string,
    footer: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      data: null,
      transpose: false,
      responsive: true,
      rowHeader: false,
      colHeader: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return {
      data: this._getData(this.props)
    };
  },
  componentWillReceiveProps(nextProps) {
    if (this.props.data !== nextProps.data || this.props.children !== nextProps.children) {
      this.setState({ data: this._getData(nextProps) });
    }
  },
  shouldComponentUpdate(nextProps, nextState) {
    if (this.state.editation) {
      return !nextState.editation;
    } else {
      return true;
    }
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  getData() {
    return Tools.deepArrayCopy(this.state.data);
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  onBeforeForceEndEditation_() {
    return this._component && this._component.getFormValues();
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getData(props) {
    let content = null;
    if (!props.data && props.children) {
      let children = Tools.assertType(props.children, Array) ? props.children.join("") : props.children;
      content = UU5.Common.Tools.parseFromUu5JSON(children);
    }
    return props.data || content || [];
  },

  _getChild() {
    if (this.state.editation) return Tools.getEditableComponent(this.getDefault("editableComponent"), this);

    const data = this.state.data;
    if (!data || (Tools.assertType(data, Array) && data.length === 0)) {
      return <UU5.Bricks.Text>{this.getLsiItem(Lsi.LSI_LABEL_TABLE_NO_DATA)}</UU5.Bricks.Text>;
    } else if (!Tools.assertType(data, Array)) {
      return <UU5.Bricks.Text>{this.getLsiItem(Lsi.LSI_LABEL_INVALID_DATA)}</UU5.Bricks.Text>;
    } else {
      return this._getTable(data);
    }
  },
  _isEmpty(array) {
    if (!array) return true;
    if (array.constructor === Array) {
      let empty = true;
      if (array.length === 0) return true;

      for (let i = 0; i < array.length; i++) {
        if (array[i]) {
          empty = false;
          break;
        }
      }
      return empty;
    }
  },

  _getTable(data) {
    const { colHeader, rowHeader, transpose, responsive, colWidth } = this.props;
    let layout = "auto";
    if (this._isEmpty(colWidth)) {
      layout = "fixed";
    }

    return [
      this.getHeader() ? this.getHeaderChild() : null,
      <AbstractTable
        key={"table"}
        className={`${this.getClassName("table")} ${this.getClassName(layout)}`}
        data={data}
        colHeader={colHeader}
        rowHeader={rowHeader}
        transpose={transpose}
        responsive={responsive}
        colWidth={colWidth}
      />,
      this.getFooter() ? this.getFooterChild() : null
    ];
  }, //
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return this.getNestingLevel() ? <UU5.Bricks.Div {...this.getMainPropsToPass()} content={this._getChild()} /> : null;
  }
  //@@viewOff:render
});

export default Table;
