// uu_contentkitg01/dist-esm/images/internal/displayed-images-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import "uu5g04-bricks";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../../core/_config.js";
import Lsi from "../../config/lsi.js";
import Css from "./css";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.DisplayedImagesInput"),
  classNames: {
    marginRow: () => Css.css`margin: 0px -8px; & .uu5-forms-input { margin-bottom: 0; }`
  }
};
//@@viewOff:statics

const DisplayedImagesInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Bricks.Row {...UU5.Common.VisualComponent.getAttrs(props, STATICS.classNames.marginRow())}>
        <UU5.Bricks.Column colWidth="xs-6">
          <UU5.Forms.Number
            label={Lsi.displayedImages}
            onChange={opt => props.onChangeProps({ displayedItems: +opt.value })}
            min={1}
            max={100}
            value={props.componentProps.displayedItems}
          />
        </UU5.Bricks.Column>
        <UU5.Bricks.Column colWidth="xs-6">
          <UU5.Forms.Checkbox
            type={2}
            label={Lsi.stepByOne}
            onChange={opt => props.onChangeProps({ stepByOne: opt.value })}
            value={props.componentProps.stepByOne}
            disabled={props.componentProps.displayedItems === 1}
          />
        </UU5.Bricks.Column>
      </UU5.Bricks.Row>
    );
    //@@viewOff:render
  }
});

export default DisplayedImagesInput;
