// uu_contentkitg01/dist-esm/images/internal/effect-caption-select.js
//@@viewOn:imports
import UU5 from "uu5g04";
import "uu5g04-bricks";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../../core/_config.js";
import Lsi from "../../config/lsi.js";
import Css from "./css";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.EffectCaptionSelect"),
  classNames: {
    marginBottom: () => Css.css` &.uu5-forms-input { margin-bottom: 0; }`,
    marginRow: () => Css.css`margin: 0 -8px;`,
    padding: () => Css.css`padding: 0 8px;`
  }
};
//@@viewOff:statics

function generateEffectOptionsItems() {
  return Cfg.EFFECT_OPTIONS.map(effect => {
    return <UU5.Forms.Select.Option key={effect} value={effect} />;
  });
}

function generateShapeOptionsItems() {
  return Cfg.SHAPE.map(shape => {
    return <UU5.Forms.Select.Option key={shape} value={shape} />;
  });
}

const EffectCaptionSelect = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    const marginClassName = STATICS.classNames.marginBottom();
    return (
      <UU5.Bricks.Row {...UU5.Common.VisualComponent.getAttrs(props, STATICS.classNames.marginRow())}>
        <UU5.Bricks.Column colWidth="xs-12 s-6">
          <UU5.Forms.Select
            className={marginClassName}
            key="shape"
            label={Lsi.shape}
            value={props.componentProps.shape}
            onChange={opt => props.onChangeProps({ shape: opt.value })}
            inputColWidth="xs-12"
          >
            {generateShapeOptionsItems()}
          </UU5.Forms.Select>
        </UU5.Bricks.Column>
        <UU5.Bricks.Column colWidth="xs-12 s-6">
          <UU5.Forms.Select
            className={marginClassName}
            key="effect"
            label={Lsi.effect}
            value={props.componentProps.effect}
            onChange={opt => props.onChangeProps({ effect: opt.value })}
            inputColWidth="xs-12"
          >
            {generateEffectOptionsItems()}
          </UU5.Forms.Select>
        </UU5.Bricks.Column>
        <UU5.Bricks.Column
          style={props.componentProps.effect !== "blur" ? "display: none;" : ""}
          colWidth="xs-12"
          noSpacing
        >
          <UU5.Forms.Text
            className={marginClassName + " " + STATICS.classNames.padding()}
            key="caption"
            label="Caption"
            disabled={props.componentProps.effect !== "blur"}
            value={props.componentProps.caption}
            onChange={opt => props.onChangeProps({ caption: opt.value })}
          />
        </UU5.Bricks.Column>
      </UU5.Bricks.Row>
    );
    //@@viewOff:render
  }
});

export default EffectCaptionSelect;
