// uu_plus4u5filesg01/dist-node/bricks/file-manager-tile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _config = _interopRequireDefault(require("../config/config.js"));

var UU5 = _interopRequireWildcard(require("uu5g04"));

require("uu5imagingg01");

var _calls = _interopRequireDefault(require("../calls.js"));

var _lsiHelper = _interopRequireDefault(require("../helpers/lsi-helper.js"));

var _css = _interopRequireDefault(require("../helpers/css.js"));

var _tagList = _interopRequireDefault(require("./tag-list.js"));

var _imageFromCall = _interopRequireDefault(require("./image-from-call.js"));

require("./file-manager-tile.less");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; if (obj != null) { var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var LIGHTBOX_PLACEHOLDER_DATA = [{
  id: "image",
  src: "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7",
  authenticate: true
}];
var MemoizedLightbox = UU5.Common.Component.memo(function (props) {
  return UU5.Common.Element.create(UU5.Imaging.Lightbox, props);
});
var FileManagerTile = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.PureRenderMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "FileManagerTile",
    classNames: {
      main: _config.default.CSS + "file-manager-tile",
      mainSelected: _config.default.CSS + "file-manager-tile-selected",
      selectableOnly: _config.default.CSS + "file-manager-selectable-only",
      innerTile: _config.default.CSS + "file-manager-inner-tile",
      smallTile: "small-tile",
      previewField: "preview-field",
      image: "image",
      previewContentWrapper: "preview-content-wrapper",
      infoField: "info-field",
      tileCard: "tile-card",
      infoBox: "info-box",
      infoFieldRow: "info-field-row",
      infoFieldCol: "info-field-col",
      primaryAttribute: "primary-attribute",
      secondaryAttribute: "secondary-attribute",
      buttonPanel: "button-panel",
      selectButton: "select-button",
      flex: function flex(props) {
        return props.tileType === "small" ? "" : _css.default.css("\n              display: flex;\n              align-items: stretch;\n\n              & > .plus4u5-files-file-manager-tile {\n                display: flex;\n                flex-direction: column;\n                align-items: stretch;\n\n                & > .tile-card {\n                  flex: 1 0 auto;\n                  display: flex;\n                  flex-direction: column;\n                  align-items: stretch;\n\n                  & > .info-box {\n                    flex: 1 0 auto;\n                    display: flex;\n                    flex-direction: column;\n                    align-items: stretch;\n\n                    & > .info-field {\n                      flex: 1 0 auto;\n                      display: flex;\n                      flex-direction: column;\n                      align-items: stretch;\n                    }\n                  }\n                }\n              }\n            ");
      }
    },
    defaults: {
      maxButtons: 3
    },
    opt: {
      nestingLevelWrapper: true,
      pureRender: true
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    fileManager: UU5.PropTypes.object,
    tileType: UU5.PropTypes.string,
    getThumbnail: UU5.PropTypes.func,
    isS: UU5.PropTypes.bool,
    uuCmdGetBinaryData: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    uuCmdGetBinaryDataUri: UU5.PropTypes.func,
    uuCmdGetThumbnailData: UU5.PropTypes.oneOfType([UU5.PropTypes.func, UU5.PropTypes.string]),
    uuCmdGetThumbnailDataUri: UU5.PropTypes.func,
    onFileClick: UU5.PropTypes.func,
    shouldShowThumbnail: UU5.PropTypes.bool,
    fileButtons: UU5.PropTypes.object,
    customSelectable: UU5.PropTypes.string,
    selectableOnly: UU5.PropTypes.bool,
    onFilenameSet: UU5.PropTypes.func
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      selectableOnly: false
    };
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:reactLifeCycle
  getInitialState: function getInitialState() {
    return _objectSpread({
      loadError: false,
      renderButtons: this.props.isS,
      src: undefined,
      srcMeta: this._updateSrc(this.props.uuCmdGetBinaryData, this.props.uuCmdGetBinaryDataUri),
      thumbSrc: undefined,
      thumbThumbSrc: undefined
    }, this._updateThumbnails(), {
      icon: this.props.data.icon || "mdi-file",
      thumbnailType: this.props.data.thumbnailType || "icon",
      binaryType: this.props.data.type || "other"
    });
  },
  componentDidMount: function componentDidMount() {
    this._updateLightbox();
  },
  componentDidUpdate: function componentDidUpdate(prevProps, prevState) {
    var _this = this;

    this.setState(function (prevState) {
      if (!prevState.renderButtons && _this.props.isS) {
        return {
          renderButtons: true
        };
      }
    });

    this._updateLightbox(prevState.src);
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    var newState = {};

    if (nextProps.baseUri !== this.props.baseUri || nextProps.uuCmdGetBinaryData !== this.props.uuCmdGetBinaryData || nextProps.code !== this.props.code) {
      newState.srcMeta = this._updateSrc(nextProps.uuCmdGetBinaryData, nextProps.uuCmdGetBinaryDataUri, {
        baseUri: nextProps.baseUri,
        binary: nextProps.data.binary,
        data: {
          code: nextProps.code
        }
      });
    }

    var updatedThumbnailData = this._updateThumbnails(nextProps);

    if (!UU5.Common.Tools.deepEqual(updatedThumbnailData.thumbSrcMeta, this.state.thumbSrcMeta)) {
      newState.thumbSrcMeta = updatedThumbnailData.thumbSrcMeta;
    }

    if (!UU5.Common.Tools.deepEqual(updatedThumbnailData.thumbThumbSrcMeta, this.state.thumbThumbSrcMeta)) {
      newState.thumbThumbSrcMeta = updatedThumbnailData.thumbThumbSrcMeta;
    }

    if (nextProps.data.icon !== this.props.data.icon || nextProps.data.thumbnailType !== this.props.data.thumbnailType || nextProps.data.binaryType !== this.props.data.binaryType) {
      newState = _objectSpread({}, newState, {
        icon: nextProps.data.icon || "mdi-file",
        thumbnailType: nextProps.data.thumbnailType || "icon",
        binaryType: nextProps.data.type || "other"
      });
    }

    if (Object.keys(newState).length) {
      this.setState(newState);
    }
  },
  //@@viewOff:reactLifeCycle
  //@@viewOn:interface
  //@@viewOff:interface
  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods
  //@@viewOn:private
  _updateLightbox: function _updateLightbox(prevSrc) {
    if (this._lightbox && prevSrc !== this.state.src) {
      if (this.state.src !== undefined) {
        this._lightbox.replaceImage(this.state.src || LIGHTBOX_PLACEHOLDER_DATA[0].src, "image");
      }
    } // NOTE Currently it's not possible to change data while lightbox is opened => that's why
    // we waited for new "src" and we'll open the lightbox now.


    if (this._lightbox && this._lastOpenId !== this.state.lightboxOpenId && this.state.src !== undefined) {
      this._lastOpenId = this.state.lightboxOpenId;

      this._lightbox.open();
    }
  },
  _updateSrc: function _updateSrc(uuCmd, uuCmdUri) {
    var _ref = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {},
        baseUri = _ref.baseUri,
        binary = _ref.binary,
        stringParams = _ref.stringParams,
        data = _ref.data,
        _ref$download = _ref.download,
        download = _ref$download === void 0 ? false : _ref$download;

    var callback = arguments.length > 3 ? arguments[3] : undefined;
    data = data || {
      code: this.props.data.binary.code
    };
    baseUri = baseUri || this.props.baseUri;
    var stringSrc;

    if (typeof uuCmdUri === "function") {
      stringSrc = uuCmdUri(binary || this.props.data.binary, download);
    } else if (typeof uuCmd !== "function") {
      stringSrc = _calls.default.getCommandUri(baseUri, uuCmd, data);

      if (stringParams) {
        stringSrc += "&" + stringParams;
      }

      if (download) stringSrc += "&contentDisposition=attachment";
    }

    if (download && stringSrc && UU5.Environment.isTrustedDomain(stringSrc)) {
      var session = UU5.Environment.getSession();
      stringSrc += "&access_token=" + (session ? (session.getCallToken() || {}).token : "");
    }

    return {
      data: data,
      uuCmd: uuCmd,
      stringSrc: stringSrc,
      callback: callback
    };
  },
  _updateThumbnails: function _updateThumbnails() {
    var props = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : this.props;
    var code = props.data.binary.code;
    var thumbCode = code.length > 61 ? code.substr(0, 61) : code;
    thumbCode = thumbCode + "_th";
    var thumbnail = props.getThumbnail(thumbCode);
    var thumbSrcMeta;
    var thumbThumbSrcMeta;

    if (thumbnail) {
      var stringParams = thumbnail.sys.mts;
      stringParams = stringParams.replace(/[:.-]/g, ""); // TODO ??? Why using uuCmdGetThumbnailData for getting props.data.binary?

      thumbSrcMeta = this._updateSrc(props.uuCmdGetThumbnailData, props.uuCmdGetThumbnailDataUri, {
        data: {
          code: code
        },
        binary: props.data.binary,
        stringParams: stringParams
      });
      thumbThumbSrcMeta = this._updateSrc(props.uuCmdGetThumbnailData, props.uuCmdGetThumbnailDataUri, {
        data: {
          code: thumbCode
        },
        binary: thumbnail,
        stringParams: stringParams
      });
    }

    return {
      thumbSrcMeta: thumbSrcMeta,
      thumbThumbSrcMeta: thumbThumbSrcMeta
    };
  },
  _buildPreview: function _buildPreview(renderImages) {
    var _this2 = this;

    var preview;

    if (this.state.thumbnailType === "original") {
      // no thumbnails, show original images
      if (renderImages) {
        var mainPropsToPass = this.getMainPropsToPass();

        if (this.props.customSelectable !== "none" && this.props.selectableOnly) {
          preview = UU5.Common.Element.create(_imageFromCall.default, {
            srcMeta: this.state.srcMeta,
            onSrcReady: function onSrcReady(_ref2) {
              var src = _ref2.src;
              return _this2.setState({
                src: src
              });
            },
            style: {
              maxWidth: mainPropsToPass.style && mainPropsToPass.style.width ? mainPropsToPass.style.width - 4 : 200
            },
            authenticate: true
          });
        } else {
          preview = UU5.Common.Element.create(_imageFromCall.default, {
            srcMeta: this.state.srcMeta,
            onSrcReady: function onSrcReady(_ref3) {
              var src = _ref3.src;
              return _this2.setState({
                src: src
              });
            },
            imageComponent: UU5.Imaging.Image,
            fit: "coverNoScaleUp",
            authenticate: true,
            onClick: this.props.onFileClick ? function () {
              return _this2.props.onFileClick(_this2.props.data.binary);
            } : undefined
          });
        }
      } else {
        preview = UU5.Common.Element.create(UU5.Bricks.Icon, {
          icon: this.state.icon
        });
      }
    } else if (this.state.thumbnailType === "thumbnail") {
      // show thumbnails
      if (this.state.thumbSrcMeta && renderImages && !this.state.loadError) {
        var _mainPropsToPass = this.getMainPropsToPass();

        if (this.props.customSelectable !== "none" && this.props.selectableOnly) {
          preview = UU5.Common.Element.create(_imageFromCall.default, {
            srcMeta: this.state.thumbThumbSrcMeta,
            onSrcReady: function onSrcReady(_ref4) {
              var src = _ref4.src;
              return _this2.setState({
                thumbThumbSrc: src
              });
            },
            style: {
              maxWidth: _mainPropsToPass.style && _mainPropsToPass.style.width ? _mainPropsToPass.style.width - 4 : 200
            },
            authenticate: true
          });
        } else {
          preview = UU5.Common.Element.create(_imageFromCall.default, {
            srcMeta: this.state.thumbSrcMeta,
            thumbSrcMeta: this.state.thumbThumbSrcMeta,
            imageComponent: UU5.Imaging.Image,
            onSrcReady: function onSrcReady(_ref5) {
              var src = _ref5.src,
                  thumbSrc = _ref5.thumbSrc;
              return _this2.setState({
                thumbSrc: src,
                thumbThumbSrc: thumbSrc
              });
            },
            lightbox: this.props.data.id,
            fit: "coverNoScaleUp",
            authenticate: true,
            onClick: this.props.onFileClick ? function () {
              return _this2.props.onFileClick(_this2.props.data.binary);
            } : undefined
          });
        }
      } else {
        preview = UU5.Common.Element.create(UU5.Bricks.Icon, {
          icon: this.state.icon
        });
      }
    } else {
      // show icons
      preview = UU5.Common.Element.create(UU5.Bricks.Icon, {
        icon: this.state.icon
      });
    } // TODO :-( Currently we need this.state.src because of Lightbox so we download both thumbnail & full image
    // right away.


    if (this.state.lightboxSrcMeta) {
      preview = UU5.Common.Element.create(UU5.Common.Fragment, null, preview, UU5.Common.Element.create(_imageFromCall.default, {
        style: {
          display: "none"
        },
        srcMeta: this.state.lightboxSrcMeta,
        onSrcReady: function onSrcReady(_ref6) {
          var src = _ref6.src;
          return _this2.setState({
            src: src
          });
        },
        srcReadyKey: this.state.lightboxOpenId + "",
        authenticate: true
      }));
    }

    return preview;
  },
  _openPreviewForm: function _openPreviewForm(binaryCode, type, event) {
    event.preventDefault();
    event.stopPropagation();
    this.props.onFileClick(binaryCode, type);
  },
  _getMainAttrValue: function _getMainAttrValue() {
    return this.props.onFilenameSet ? _lsiHelper.default.getLsiComponent(this.props.onFilenameSet(this.props.data.binary)) : this.props.data.binary.filename;
  },
  _toggleSelected: function _toggleSelected() {
    if (this.props.selected) {
      this.props.fileManager.customRemoveSelectedItem(this.props.id);
    } else {
      this.props.fileManager.customAddSelectedItem(this.props.id);
    }
  },
  _onDownloadLinkClick: function _onDownloadLinkClick() {
    var _this$_updateSrc = this._updateSrc(this.props.uuCmdGetBinaryData, this.props.uuCmdGetBinaryDataUri, {
      download: true
    }),
        data = _this$_updateSrc.data,
        stringSrc = _this$_updateSrc.stringSrc,
        uuCmd = _this$_updateSrc.uuCmd;

    if (!stringSrc && typeof this.props.uuCmdGetBinaryData === "function") {
      // blob cannot be downloaded directly - we must download manually and change content type to octet/stream
      var filename = this.props.binary.filename;
      uuCmd(data).then(function (dtoOut) {
        var fileBlob = new Blob([dtoOut], {
          type: "octet/stream"
        }); // save blob in IE and Edge - standard click on a link doesn't work in this browsers

        if (typeof navigator.msSaveBlob === "function") {
          navigator.msSaveBlob(fileBlob, filename);
          return;
        }

        var url = window.URL.createObjectURL(fileBlob); // to set name for the file we need to create link element and set name into download attribute

        var link = document.createElement("a");
        link.href = url;
        link.download = filename;
        link.style = "display: none;";
        document.body.appendChild(link);
        link.click();
        document.body.removeChild(link);
        setTimeout(function () {
          window.URL.revokeObjectURL(url);
        }, 60000);
      });
    } else {
      window.open(stringSrc, "_self");
    }
  },
  _getDownloadLink: function _getDownloadLink(opt, i) {
    return UU5.Common.Element.create(UU5.Bricks.Link, {
      key: "dropdownItem" + i,
      className: "uu5-bricks-dropdown-item-link",
      content: _lsiHelper.default.getLsiComponent(opt.label),
      onClick: this._onDownloadLinkClick,
      download: true
    });
  },
  _renderButtonPanel: function _renderButtonPanel() {
    var _this3 = this;

    var buttons = [];
    var dropdownItems = [];
    var _this$props$fileButto = this.props.fileButtons,
        itemList = _this$props$fileButto.itemList,
        hideDropdown = _this$props$fileButto.hideDropdown;

    if (this.props.customSelectable !== "none" && this.props.selectableOnly && this.props.data.clickable) {
      buttons.push(UU5.Common.Element.create(UU5.Bricks.Button, {
        key: "magnifyButton",
        onClick: function onClick(opt, event) {
          event.preventDefault();
          event.stopPropagation();

          if (_this3.props.onFileClick) {
            _this3.props.onFileClick(_this3.props.data.binary);
          } else if (_this3._lightbox) {
            _this3.setState(function (state) {
              return {
                lightboxSrcMeta: state.srcMeta,
                lightboxOpenId: (state.lightboxOpenId || 0) + 1
              };
            });
          }
        },
        colorSchema: "primary"
      }, UU5.Common.Element.create(UU5.Bricks.Icon, {
        icon: "mdi-magnify-plus"
      })));
    }

    itemList = itemList.filter(Boolean);
    var dropdownItemsAreDuplicitOnly = true;

    var _loop = function _loop(i) {
      var item = itemList[i];
      var alsoAsButton = false;

      if (item.active && (_this3.props.customSelectable === "none" || !_this3.props.selectableOnly) && !(_this3.props.tileType === "small" && _this3.props.isS)) {
        alsoAsButton = true;
      }

      if (alsoAsButton) {
        buttons.push(UU5.Common.Element.create(UU5.Bricks.Button, {
          key: "button" + i,
          onClick: function onClick(opt, event) {
            event.preventDefault();
            event.stopPropagation();
            if (typeof item.onClick === "function") item.onClick(_this3.props.data.binary, _this3.state.binaryType);
          },
          colorSchema: "primary",
          tooltip: item.label ? _this3.getLsiItem(item.label) : undefined
        }, UU5.Common.Element.create(UU5.Bricks.Icon, {
          icon: item.icon
        })));
      }

      if ((!hideDropdown || _this3.props.tileType === "small" && _this3.props.isS) && (_this3.props.customSelectable === "none" || !_this3.props.selectableOnly)) {
        if (item.isDownload) {
          if (dropdownItemsAreDuplicitOnly) dropdownItemsAreDuplicitOnly = alsoAsButton;
          dropdownItems.push(_this3._getDownloadLink(item, i));
        } else {
          if (item.icon !== "mdi-magnify-plus" || item.icon === "mdi-magnify-plus" && _this3.props.data.clickable) {
            if (dropdownItemsAreDuplicitOnly) dropdownItemsAreDuplicitOnly = alsoAsButton;
            dropdownItems.push(UU5.Common.Element.create(UU5.Bricks.Dropdown.Item, {
              label: _lsiHelper.default.getLsiComponent(item.label),
              onClick: function onClick() {
                return item.onClick(_this3.props.data.binary, _this3.state.binaryType);
              },
              key: "dropdownItem" + i
            }));
          }
        }
      }
    };

    for (var i = 0; i < itemList.length; i++) {
      _loop(i);
    }

    if (dropdownItemsAreDuplicitOnly) dropdownItems = [];
    var remainingButtons = this.getDefault("maxButtons") - (dropdownItems.length > 0 ? 1 : 0);
    if (buttons.length > remainingButtons) buttons.splice(remainingButtons);
    return this.state.renderButtons && UU5.Common.Element.create(UU5.Bricks.Div, {
      className: this.getClassName("buttonPanel")
    }, buttons, dropdownItems.length > 0 && UU5.Common.Element.create(UU5.Bricks.Dropdown, {
      label: "",
      ref_: function ref_(opt) {
        return _this3._dropdown = opt;
      },
      iconOpen: "mdi-dots-vertical",
      iconClosed: "mdi-dots-vertical",
      pullRight: true,
      colorSchema: "primary",
      allowTags: ["UU5.Bricks.Link"]
    }, dropdownItems));
  },
  _renderSelectButton: function _renderSelectButton() {
    var _this4 = this;

    return this.props.customSelectable !== "none" && UU5.Common.Element.create(UU5.Bricks.Button, {
      key: "selectButton",
      className: this.getClassName("selectButton") + " " + (this.props.selected ? this.getClassName("selectButtonSelected") : ""),
      tooltip: this.getLsiItem({
        cs: "Vybrat soubor",
        en: "Select File"
      }),
      onClick: function onClick(opt, event) {
        return _this4._toggleSelected(event);
      },
      colorSchema: this.props.isS ? this.props.selected ? "orange" : "blue" : "orange"
    }, UU5.Common.Element.create(UU5.Bricks.Icon, {
      icon: "mdi-checkbox-marked-circle",
      style: {
        color: "#ffffff",
        fontSize: this.props.isS ? null : "32px"
      }
    }));
  },
  _getFullInfoBox: function _getFullInfoBox() {
    var _this5 = this;

    return UU5.Common.Element.create("div", {
      className: this.getClassName("infoBox")
    }, UU5.Common.Element.create("div", {
      className: this.getClassName("previewField") + (this.props.data.clickable ? " " + this.getClassName("image") : "")
    }, UU5.Common.Element.create("div", {
      className: this.getClassName("previewContentWrapper")
    }, this._buildPreview(this.props.shouldShowThumbnail))), UU5.Common.Element.create("div", {
      className: this.getClassName("infoField")
    }, UU5.Common.Element.create("div", {
      className: this.getClassName("primaryAttribute"),
      title: this.props.data.binary.filename
    }, this._getMainAttrValue()), UU5.Common.Element.create("div", {
      className: this.getClassName("secondaryAttribute"),
      title: this.props.data.binary.code
    }, this.props.data.binary.code), UU5.Common.Element.create(UU5.Bricks.Resize, null, function (size) {
      return UU5.Common.Element.create(_tagList.default, {
        key: "tagList",
        items: _this5.props.data.binary.tagList,
        onTagExpand: _this5.props.handleMeasure,
        className: "tag-list",
        width: size.width,
        height: size.height
      });
    }), UU5.Common.Element.create("div", {
      className: this.getClassName("secondaryAttribute")
    }, new Date(this.props.data.binary.sys.mts).toLocaleString().slice(0, -3).replace(/:$/, ""))), this._renderSelectButton(), !this.props.isS && this._renderButtonPanel());
  },
  _getSmallInfoBox: function _getSmallInfoBox() {
    return UU5.Common.Element.create("div", null, UU5.Common.Element.create("div", {
      className: this.getClassName("previewField") + (this.props.data.clickable ? " " + this.getClassName("image") : "")
    }, this._buildPreview(false)), UU5.Common.Element.create("div", {
      className: this.getClassName("infoField")
    }, UU5.Common.Element.create("div", {
      className: this.getClassName("infoFieldRow")
    }, UU5.Common.Element.create("div", {
      className: this.getClassName("infoFieldCol"),
      title: this.props.data.binary.filename
    }, this._getMainAttrValue()), UU5.Common.Element.create("div", {
      className: this.getClassName("infoFieldCol")
    }, new Date(this.props.data.binary.sys.mts).toLocaleString().slice(0, -3))), UU5.Common.Element.create("div", {
      className: this.getClassName("infoFieldRow")
    }, UU5.Common.Element.create("div", {
      className: this.getClassName("infoFieldCol"),
      title: this.props.data.binary.code
    }, this.props.data.binary.code))), this._renderSelectButton(), !this.props.isS && this._renderButtonPanel());
  },
  _setLightboxRef: function _setLightboxRef(ref) {
    this._lightbox = ref;
  },
  //@@viewOff:private
  //@@viewOn:render
  render: function render() {
    var _this6 = this;

    var mainPropsToPass = this.getMainPropsToPass();
    if (this.props.selected) mainPropsToPass.className += " " + this.getClassName("mainSelected");
    if (this.props.customSelectable !== "none" && this.props.selectableOnly) mainPropsToPass.className += " " + this.getClassName("selectableOnly");
    if (this.props.tileType === "small") mainPropsToPass.className += " " + this.getClassName("smallTile");
    return UU5.Common.Element.create(UU5.Bricks.Div, {
      style: this.props.wrapperStyle,
      className: this.getClassName("flex")
    }, UU5.Common.Element.create(UU5.Bricks.Div, _extends({}, mainPropsToPass, {
      mainAttrs: {
        onMouseOver: function onMouseOver() {
          return _this6.setState({
            renderButtons: true
          });
        },
        onClick: function onClick(event) {
          event.preventDefault();
          event.stopPropagation();
          _this6.props.customSelectable !== "none" && _this6.props.selectableOnly ? _this6._toggleSelected() : null;
        }
      }
    }), UU5.Common.Element.create(UU5.Bricks.Card, {
      className: this.getClassName("tileCard"),
      elevation: 1,
      elevationHover: 2,
      colorSchema: this.props.selected ? "orange" : null,
      bgStyle: "outline",
      style: {
        borderWidth: this.props.selected ? "2px" : "1px",
        margin: "0"
      }
    }, this.props.isS && this._renderButtonPanel(), this.props.tileType === "small" ? this._getSmallInfoBox() : this._getFullInfoBox()), UU5.Common.Element.create("div", {
      onClick: function onClick(e) {
        return e.stopPropagation();
      }
    }, this.props.selectableOnly ? // our re-render always causes the lightbox to close => use memo-ized component so that it doesn't re-render at all
    // (we'll control it using API methods)
    UU5.Common.Element.create(MemoizedLightbox, {
      name: this.props.data.id,
      ref_: this._setLightboxRef,
      data: LIGHTBOX_PLACEHOLDER_DATA
    }) : null)));
  } //@@viewOff:render

});
var _default = FileManagerTile;
exports.default = _default;
