// eslint-plugin-uu5/src/index.js
/**
 * @fileoverview Basic linting for uu5 library
 * @author Milan Martinek
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

var requireIndex = require("requireindex");

//------------------------------------------------------------------------------
// Plugin Definition
//------------------------------------------------------------------------------

module.exports = {
  // import all rules in lib/rules
  rules: requireIndex(__dirname + "/rules"),
  configs: {
    recommended: {
      plugins: ["uu5"],
      parserOptions: {
        ecmaVersion: 2017,
        ecmaFeatures: {
          jsx: true
        },
        sourceType: "module"
      },
      rules: {
        "uu5/base-mixin": "error",
        "uu5/component-comments": "error",
        "uu5/import-order": "warn",
        "uu5/tag-name": "error",
        "uu5/no-deprecated-mixins": "warn",
        "uu5/hooks-rules": "error",
        "uu5/hooks-exhaustive-deps": "warn"
      }
    }
  }
};
