// uu_appg01_devkit-common/src/templates/__nodejs-service/app/api/errors/project-error.js
"use strict";
const <%= context.projectClassName %>UseCaseError = require("./<%= context.projectClassFile %>-use-case-error.js");

const Init = {
  UC_CODE: `${<%= context.projectClassName %>UseCaseError.ERROR_PREFIX}init/`,

  InvalidDtoIn: class extends <%= context.projectClassName %>UseCaseError {
    constructor() {
      super(...arguments);
      this.code = `${Init.UC_CODE}invalidDtoIn`;
      this.message = "DtoIn is not valid.";
    }
  },

  SchemaDaoCreateSchemaFailed: class extends <%= context.projectClassName %>UseCaseError {
    constructor() {
      super(...arguments);
      this.status = 500;
      this.code = `${Init.UC_CODE}schemaDaoCreateSchemaFailed`;
      this.message = "Create schema by Dao createSchema failed.";
    }
  },

  SetProfileFailed: class extends <%= context.projectClassName %>UseCaseError {
    constructor() {
      super(...arguments);
      this.code = `${Init.UC_CODE}sys/setProfileFailed`;
      this.message = "Set profile failed.";
    }
  },

  CreateAwscFailed: class extends <%= context.projectClassName %>UseCaseError {
    constructor() {
      super(...arguments);
      this.code = `${Init.UC_CODE}createAwscFailed`;
      this.message = "Create uuAwsc failed.";
    }
  },
};

module.exports = {
  Init,
};
