// uu_appg01_devkit/src/scripts/cmd_generator/test-generator.js
const path = require("path");
const { fileExists } = require("../../tools/helpers");
const util = require("util");
const fs = require("fs-extra");
const fs_mkdir = util.promisify(fs.mkdir);
const { processFile } = require("uu_appg01_devkit-common/src/tools/template-helpers.js");

class TestGenerator {

  constructor(context, executionQueue) {
    this.context = JSON.parse(JSON.stringify(context));
    let dirname = path.dirname(this.context.sourceCodePath);
    this.testsFolder = path.join(dirname, "test");
    this.entityTestFolder = path.join(dirname, "test", this.context.entity);
    this.testFile = path.join(this.entityTestFolder, `${context.testFilePrefix}.test.js`);
    this.testTemplate = path.join(__dirname, "templates", "test-template.js");
    this.executionQueue = executionQueue;
  }

  async execute() {
    if (!await fileExists(this.testsFolder)) {
      console.log(`Test folder ${this.testsFolder} does not exit.`);
      process.exit(1);
    }
    if (!await fileExists(this.entityTestFolder)) {
      console.log(`Creating ${this.entityTestFolder}`);
      await fs_mkdir(this.entityTestFolder);
    }
    if (await fileExists(this.testFile)) {
      console.log(`Test file ${this.testFile} already exists`);
      process.exit(1);
    }
    this.executionQueue.addEvent(async () => {
      console.log(`Creating ${this.testFile}`);
      await processFile(this.testTemplate, this.testFile, this.context);
    });
  }

}

module.exports = TestGenerator;
