// uu_appg01_core-perflog/src/logger.js
"use strict";

const { Config } = require("uu_appg01_core-utils");
const LoggerFactory = require("uu_appg01_core-logging").LoggerFactory;
const BaseLogger = require("uu_appg01_core-logging").Logger;
const LogLevel = require("uu_appg01_core-logging").LogLevel;
const MessageFormatter = require("./message-formatter.js");

const LOGGER_NAME = "UuApp.Perflog.Logger";

/**
 * Logger component internally used by Perflog for providing gathered performance
 * statistics data. Logger is transparently managed via LoggerFactory and
 * therefore supports external configuration via following configuration parameters:
 *
 * * <b>uuapp.perflog.logger.log_level</b><br>
 *   Defines log level. Level can be defined either via name, or its numeric representation. Defaults to +INFO+.
 *
 * * <b>uuapp.perflog.logger.log_message_format</b><br>
 *   Allows to define custom message format. See {UuApp::Perflog::MessageFormatter#message_format} for details.
 */
class Logger {
  /**
   * Overrides default message format.
   * @param messageFormat Format string.
   */
  static setMessageFormat(messageFormat) {
    Logger._getLogger().setMessageFormat(messageFormat);
  }

  /**
   * Logging severity threshold. Possible values are:
   # * OFF - Logger is disabled.
   # * INFO - Logger logs data of first level measured sections only.
   # * DEBUG - Logger logs data of all measured sections.
   * @param level Log level.
   */
  static setLevel(level) {
    Logger._getLogger().setLevel(level);
  }

  /**
   * Checks if performance log is enabled (log level is set to anything other than "OFF").
   * @returns {boolean} True if log is enabled, else false.
   */
  static isEnabled() {
    return Logger._getLogger().getLevel() < LogLevel.OFF;
  }

  /**
   * Checks verbosity of performance log. If log is in verbose mode (log
   * level is set to "DEBUG") it prints data of all measured sections. Else
   * it logs only first level sections.
   * @returns {boolean} True if perfomance log is in verbose mode, else false.
   */
  static isVerbose() {
    return Logger._getLogger().getLevel() < LogLevel.INFO;
  }

  /**
   * Transforms given section to log message and prints it.
   * @param section Section to be logged.
   */
  static log(section) {
    if (!section) {
      return;
    }
    Logger._getLogger().log(section);
  }

  static _getLogger() {
    if (!this._logger) {
      let logger = new BaseLogger(LOGGER_NAME);
      if (!Config.get(`${LOGGER_NAME.toLowerCase()}.log_level`, false)) {
        Config.set(`${LOGGER_NAME.toLowerCase()}.log_level`, "INFO");
      }
      // TODO Should be settable via Logger API
      logger._formatter = new MessageFormatter();
      LoggerFactory.configure(logger, false);
      this._logger = logger;
    }
    return this._logger;
  }
}

module.exports = Logger;
