// uu_plus4u5g01/dist-esm/bricks/cmd-viewer.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import Calls from "./calls.js";

import "./cmd-viewer.less";

export const CmdViewer = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.LoadMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.Bricks.CmdViewer",
    classNames: {
      main: "plus4u5-bricks-cmd-viewer"
    },
    calls: {
      onLoad: "loadCmd"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    url: UU5.PropTypes.string.isRequired,
    params: UU5.PropTypes.object,
    theme: UU5.PropTypes.string,
    minRows: UU5.PropTypes.number,
    maxRows: UU5.PropTypes.number,
    rows: UU5.PropTypes.number,
    hideMarks: UU5.PropTypes.bool,
    loadMethod: UU5.PropTypes.oneOf(["get", "post"])
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps: function() {
    return {
      url: null,
      params: null,
      theme: "chrome",
      minRows: null,
      maxRows: 50,
      rows: null,
      hideMarks: false,
      loadMethod: "get"
    };
  },

  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  componentWillMount() {
    this.setCalls(Calls);
  },
  componentWillReceiveProps(nextProps) {
    if (this.props.hideMarks !== nextProps.hideMarks) {
      this._hideMarks(nextProps);
    }
  },

  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  getOnLoadData_(props) {
    return {
      url: props.url,
      params: props.params,
      loadMethod: props.loadMethod.toLowerCase()
    };
  },
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _hideMarks(props = this.props) {
    if (!this._editor) return;
    setTimeout(() => {
      let aceEditor = this._editor._textInput.editor;
      aceEditor.getSession().setFoldStyle(props.hideMarks ? "manual" : "markbegin");
    }, 1000);
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function() {
    return this.getLoadFeedbackChildren(dtoOut => {
      return UU5.Common.Tools.findComponent(
        "UU5.CodeKit.JsonEditor",
        UU5.Common.Tools.merge({}, this.getMainPropsToPass(), {
          value: dtoOut,
          format: "pretty",
          readOnly: true,
          theme: this.props.theme,
          minRows: this.props.minRows,
          maxRows: this.props.rows ? null : this.props.maxRows,
          rows: this.props.rows,
          ref_: editor => {
            this._editor = editor;
            this._hideMarks();
          }
        })
      );
    });
  }
  //@@viewOff:render
});

export default CmdViewer;
