// uu_plus4u5g01/dist-esm/bricks/error.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";

export const Error = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.Bricks.Error",
    classNames: {
      main: "plus4u5-bricks-error"
    },
    lsi: () => Plus4U5.Environment.Lsi.Bricks.error
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    errorData: UU5.PropTypes.object,
    error: UU5.PropTypes.shape({
      stack: UU5.PropTypes.string,
      name: UU5.PropTypes.string,
      message: UU5.PropTypes.string
    }),
    silent: UU5.PropTypes.bool,
    inline: UU5.PropTypes.bool,
    errorDataSpace: UU5.PropTypes.number // TODO: set deprecated
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      errorData: null,
      error: null,
      silent: false,
      inline: false,
      errorDataSpace: null
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers

  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render: function() {
    let props = this.getMainPropsToPass();
    let children;

    if (!this.props.children && !this.getContent()) {
      props.content = this.getLsiComponent("message");
      if (this.props.inline) {
        props.content = [
          <span key="message" style={{ marginRight: "4px" }}>
            {this.getLsiComponent("messageFull")}
          </span>,
          <Plus4U5.Bricks.LinkHelpdesk colorSchema="default" key="LinkHelpDesk" />
        ];
      }
    } else if (this.props.children) {
      children = UU5.Common.Children.toArray(this.props.children);
    }

    return (
      <UU5.Bricks.Error
        {...props}
        silent={this.props.silent}
        errorData={this.props.errorData}
        inline={this.props.inline}
        error={this.props.error}
        moreInfo
        errorInfo={
          <div>
            {this.getLsiComponent("content")} <Plus4U5.Bricks.LinkHelpdesk colorSchema="default" />
          </div>
        }
      >
        {children}
      </UU5.Bricks.Error>
    );
  }
  //@@viewOff:render
});

export default Error;
