// uu_plus4u5g01/dist-esm/bricks/link.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

import "./link.less";

export const Link = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ColorSchemaMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.Bricks.Link",
    classNames: {
      main: "plus4u5-bricks-link"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    href: UU5.PropTypes.string,
    onClick: UU5.PropTypes.func,
    onCtrlClick: UU5.PropTypes.func,
    onWheelClick: UU5.PropTypes.func,
    smoothScroll: UU5.PropTypes.number,
    offset: UU5.PropTypes.number,
    target: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Link
        {...this.getMainPropsToPass()}
        {...this.props}
        authenticate={this.props.href && !this.props.href.startsWith("#")}
        content={
          (this.props.content == null && !this.props.children) || this.props.content === ""
            ? this.props.href
            : this.props.content
        } // fallback to displaying original href (without token)
      />
    );
  }
  //@@viewOff:render
});

export default Link;
