// uu_plus4u5g01/dist-esm/bricks/plus4u-link.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";

import "./plus4u-link.less";

export const Plus4ULink = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.ColorSchemaMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.Bricks.Plus4ULink",
    classNames: {
      main: "plus4u5-bricks-plus4u-link"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uri: UU5.PropTypes.string,
    target: UU5.PropTypes.string
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      uri: null,
      target: "_blank"
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _onClick(e, target) {
    window.open(Plus4U5.Common.Tools.getLegacyUrl(this.props.uri), target || this.props.target);
    e.preventDefault();
    return this;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    return (
      <UU5.Bricks.Link
        {...this.getMainPropsToPass()}
        href={Plus4U5.Common.Tools.getLegacyUrl(this.props.uri)}
        onClick={(component, e) => this._onClick(e)}
        onCtrlClick={(component, e) => this._onClick(e, "_blank")}
        onWheelClick={(component, e) => this._onClick(e, "_blank")}
        content={this.props.content}
        colorSchema={this.props.colorSchema}
      >
        {this.props.children}
      </UU5.Bricks.Link>
    );
  }
  //@@viewOff:render
});

export default Plus4ULink;
