// uu_plus4u5g01/dist-esm/bricks/user-photo.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";
import * as Plus4U5 from "uu_plus4u5g01";

import "./user-photo.less";
import RichImage from "./rich-image.js";

export const UserPhoto = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.ElementaryMixin, UU5.Common.IdentityMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.Bricks.UserPhoto",
    classNames: {
      main: "plus4u5-bricks-user-photo"
    },
    defaults: {
      anonymousSrc: Plus4U5.Environment.basePath + "core/assets/img/anonymous.png"
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uuIdentity: UU5.PropTypes.string,
    src: UU5.PropTypes.string,
    type: UU5.PropTypes.oneOf(["rounded", "circle", "thumbnail"]),
    width: UU5.PropTypes.oneOfType([UU5.PropTypes.number, UU5.PropTypes.string]),
    shadow: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      uuIdentity: null,
      src: null,
      type: null,
      width: 50,
      shadow: true
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _getUserPhoto(uuIdentity) {
    let uri = this.props.src;

    if (uuIdentity !== "0-0" && !uri && this.isAuthenticated()) {
      let url =
        "https://uuappg01-eu-w-1.plus4u.net/uu-plus4upeople-maing01/56ac93ddb0034de8b8e4f4b829ff7d0f/people/getPersonPhoto?uuIdentity=%s";
      uri = UU5.Common.Tools.formatString(url, [uuIdentity || this.getIdentity().uuIdentity]);
    }

    if (!uri) uri = this.getDefault("anonymousSrc");

    return uri;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    const props = this.getMainPropsToPass();
    props.mainAttrs = props.mainAttrs || {};
    props.mainAttrs.style = props.mainAttrs.style || {};
    if (typeof this.props.width == "number") {
      props.mainAttrs.style.width = this.props.width + "px";
    } else if (typeof this.props.width == "string") {
      props.mainAttrs.style.width = this.props.width;
    } else if (this.props.width != null) {
      props.mainAttrs.style.width = "50px";
    }
    let userPhoto = this._getUserPhoto(this.props.uuIdentity);
    let isAnonymous = userPhoto === this.getDefault("anonymousSrc");

    return (
      <RichImage
        {...props}
        name={this.props.uuIdentity}
        src={userPhoto}
        loadingImageSrc={isAnonymous ? null : this.getDefault("anonymousSrc")}
        errorImageSrc={this.getDefault("anonymousSrc")}
        shape={this.props.type}
        effect="none"
        shadow={this.props.shadow}
        authenticate={!isAnonymous}
      />
    );
  }
  //@@viewOff:render
});

export default UserPhoto;
