// uu_plus4u5g01/dist-esm/bricks/user-data.js
import * as UU5 from "uu5g04";
import "uu5g04-bricks";

export const UserData = UU5.Common.Component.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "Plus4U5.Bricks.UserData"
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {
    uuIdentity: UU5.PropTypes.string.isRequired,
    detail: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  getDefaultProps() {
    return {
      uuIdentity: null,
      detail: false
    };
  },
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  _filterData(inputData) {
    const formatKey = key => {
      key = UU5.Common.Tools.getCamelCase(key.replace(/_/g, "-"));
      return key.charAt(0).toLowerCase() + key.slice(1);
    };

    const filterLevel = data => {
      let filteredData = {};

      for (let key in data) {
        if (key === "p4u_id") {
          filteredData.uuIdentity = data[key];
        } else if (key === "contact_list") {
          filteredData.contactList = [];
          data[key].forEach(contact => {
            filteredData.contactList.push(filterLevel(contact));
          });
        } else if (key !== "uri") {
          filteredData[formatKey(key)] = data[key];
        }
      }

      return filteredData;
    };

    return filterLevel(inputData);
  },

  _formatOutput(data) {
    data = { ...data };

    if (!data.isLoading && !data.isError) {
      if (data.data) {
        if (Array.isArray(data.data.data)) {
          data.data = data.data.data[0];
        } else {
          data.data = data.data.data;
        }
      }
    }

    return data;
  },
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    if (this.props.uuIdentity) {
      return (
        <UU5.Common.Loader
          uri="https://cmd.plus4u.net/PLUS4U-BT/uu-plus4uppl/PersonalCard/getCardList/exec"
          data={{ p4u_id: this.props.uuIdentity }}
          authenticate
        >
          {opt => {
            let outputData = this._formatOutput(opt);

            // save for later use
            let uri = outputData.data ? outputData.data.uri : undefined;
            let name = outputData.data ? outputData.data.name : undefined;

            outputData.data = this._filterData(outputData.data);

            if (!opt.isLoading && !opt.isError && this.props.detail) {
              return (
                <UU5.Common.Loader
                  uri="https://cmd.plus4u.net/uu-plus4uppl/PersonalCard/getAttributes/exec"
                  data={{ uuUri: uri }}
                  authenticate
                >
                  {opt => {
                    if (!opt.isLoading && !opt.isError) {
                      opt.data && opt.data.data && (opt.data.data.name = name);
                      outputData = this._formatOutput(opt);
                      outputData.data = this._filterData(outputData.data);
                    } else {
                      outputData = this._formatOutput(opt);
                    }

                    return this.props.children(outputData);
                  }}
                </UU5.Common.Loader>
              );
            } else if ((!opt.isError && !this.props.detail) || opt.isError) {
              return this.props.children(outputData);
            } else {
              return null;
            }
          }}
        </UU5.Common.Loader>
      );
    } else {
      return this.props.children({ isLoading: false, isError: false, data: {} });
    }
  }
  //@@viewOff:render
});

export default UserData;
