// uu_plus4u5filesg01/dist-node/core/internal/list-wrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _uu5g = _interopRequireDefault(require("uu5g04"));

require("uu5g04-bricks");

require("uu5tilesg01");

var _config = _interopRequireDefault(require("../../config/config"));

var _css = _interopRequireDefault(require("../../helpers/css.js"));

var _lsi = _interopRequireDefault(require("../../config/lsi.js"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _templateObject6() {
  var data = _taggedTemplateLiteral(["\n        position: absolute;\n        left: 0;\n        right: 0;\n        top: ", ";\n        display: flex;\n        flex-direction: column;\n        align-items: center;\n      "]);

  _templateObject6 = function _templateObject6() {
    return data;
  };

  return data;
}

function _templateObject5() {
  var data = _taggedTemplateLiteral(["\n        font-size: 24px;\n      "]);

  _templateObject5 = function _templateObject5() {
    return data;
  };

  return data;
}

function _templateObject4() {
  var data = _taggedTemplateLiteral(["\n        font-size: 80px;\n        color: #616161;\n        width: 80px;\n        height: 80px;\n        display: flex;\n        align-items: center;\n      "]);

  _templateObject4 = function _templateObject4() {
    return data;
  };

  return data;
}

function _templateObject3() {
  var data = _taggedTemplateLiteral(["\n        position: absolute;\n        left: 0;\n        right: 0;\n        top: ", ";\n        display: flex;\n        flex-direction: column;\n        align-items: center;\n      "]);

  _templateObject3 = function _templateObject3() {
    return data;
  };

  return data;
}

function _templateObject2() {
  var data = _taggedTemplateLiteral(["\n        position: absolute;\n        top: 0;\n        bottom: 0;\n        left: 0;\n        right: 0;\n        background: rgba(250, 250, 250, 0.95);\n        border: 4px dashed rgba(0, 0, 0, 0.17);\n      "]);

  _templateObject2 = function _templateObject2() {
    return data;
  };

  return data;
}

function _templateObject() {
  var data = _taggedTemplateLiteral(["\n        position: relative;\n      "]);

  _templateObject = function _templateObject() {
    return data;
  };

  return data;
}

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

//@@viewOff:imports
var ListWrapper = _uu5g.default.Common.Reference.forward(function (props, ref) {
  return _uu5g.default.Common.Element.create(ListWrapperBody, _extends({}, props, {
    listRef: ref
  }));
});

ListWrapper.displayName = _config.default.TAG + "ListWrapper";
ListWrapper.hocFor = _uu5g.default.Tiles.List.hocFor;

var ListWrapperBody = _uu5g.default.Common.VisualComponent.create({
  displayName: "ListWrapperBody",
  //@@viewOn:mixins
  mixins: [_uu5g.default.Common.BaseMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    classNames: {
      main: function main() {
        return _css.default.css(_templateObject());
      },
      emptyListWarning: _config.default.CSS + "file-manager-empty-list-warning",
      dndHoverArea: function dndHoverArea() {
        return _css.default.css(_templateObject2());
      },
      hintWrapper: function hintWrapper() {
        return _css.default.css(_templateObject3(), typeof state.positionY === "number" ? state.positionY + "px" : "0");
      },
      hintIcon: function hintIcon() {
        return _css.default.css(_templateObject4());
      },
      hintText: function hintText() {
        return _css.default.css(_templateObject5());
      }
    }
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    wrapperAttrs: _uu5g.default.PropTypes.object,
    isEmpty: _uu5g.default.PropTypes.bool,
    isEmptyMessage: _uu5g.default.PropTypes.node
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {};
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:reactLifeCycle
  getInitialState: function getInitialState() {
    this._hasRendered = false;
    return {
      positionY: null,
      height: null,
      hintHeight: null
    };
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (this.props.isHovered !== nextProps.isHovered) {
      this._hasRendered = false;
    }
  },
  componentDidUpdate: function componentDidUpdate() {
    if (this._hoverArea && this._hintWrapper && !this._hasRendered) {
      this._hasRendered = true;
      this.forceUpdate();
    }
  },
  //@@viewOff:reactLifeCycle
  //@@viewOn:interface
  //@@viewOff:interface
  //@@viewOn:overriding
  //@@viewOff:overriding
  //@@viewOn:private
  _registerHoverArea: function _registerHoverArea(ref) {
    this._hoverArea = ref;
  },
  _registerHintWrapper: function _registerHintWrapper(ref) {
    this._hintWrapper = ref;
  },
  _getHintWrapperClassName: function _getHintWrapperClassName() {
    if (this._hasRendered) {
      var hoverAreaRect = this._hoverArea.getBoundingClientRect();

      var hintWrapperRect = this._hintWrapper.getBoundingClientRect();

      var positionY;
      var visibleArea = {
        top: 0,
        height: hoverAreaRect.height
      };

      if (hoverAreaRect.top < 0) {
        visibleArea.top -= hoverAreaRect.top;
        visibleArea.height -= visibleArea.top;
      }

      if (hoverAreaRect.bottom > window.innerHeight) {
        visibleArea.height -= hoverAreaRect.bottom - window.innerHeight;
      }

      positionY = visibleArea.top + (visibleArea.height - hintWrapperRect.height) / 2;
      return _css.default.css(_templateObject6(), typeof positionY === "number" ? positionY + "px" : "0");
    }
  },
  //@@viewOff:private
  //@@viewOn:render
  render: function render() {
    var _this$props = this.props,
        wrapperAttrs = _this$props.wrapperAttrs,
        isHovered = _this$props.isHovered,
        listRef = _this$props.listRef,
        isEmpty = _this$props.isEmpty,
        isEmptyMessage = _this$props.isEmptyMessage,
        listProps = _objectWithoutProperties(_this$props, ["wrapperAttrs", "isHovered", "listRef", "isEmpty", "isEmptyMessage"]);

    return _uu5g.default.Common.Element.create(_uu5g.default.Common.Fragment, null, this.props.children, _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Div, {
      mainAttrs: wrapperAttrs,
      className: this.getClassName("main")
    }, _uu5g.default.Common.Element.create(_uu5g.default.Tiles.List, _extends({}, listProps, {
      ref: listRef
    })), isEmpty && _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Div, {
      className: this.getClassName("emptyListWarning")
    }, _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Icon, {
      icon: "mdi-package-variant"
    }), _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Div, {
      content: isEmptyMessage
    })), isHovered ? _uu5g.default.Common.Element.create("div", {
      className: this.getClassName("dndHoverArea"),
      ref: this._registerHoverArea
    }, _uu5g.default.Common.Element.create("div", {
      className: this._getHintWrapperClassName(),
      ref: this._registerHintWrapper
    }, _uu5g.default.Common.Element.create("span", {
      className: this.getClassName("hintIcon")
    }, _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Icon, {
      icon: "mdi-upload"
    })), _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Lsi, {
      className: this.getClassName("hintText"),
      lsi: _lsi.default.LSI_LABEL_FILE_MANAGER_DROP_ZONE
    }))) : null));
  } //@@viewOff:render

});

var _default = ListWrapper;
exports.default = _default;
