// uu_appg01_core/src/index.js
"use strict";

// polyfills (pretty much only for IE11)
if (typeof Symbol === "undefined") {
  require("core-js/fn/symbol");
  require("core-js/fn/array/iterator.js");
  require("core-js/fn/array/from.js");
  require("core-js/fn/array/find.js");
  require("core-js/fn/array/values.js");
  require("core-js/fn/array/includes.js");
  require("core-js/fn/reflect/construct.js");
}
if (!Object.assign) {
  require("core-js/fn/object/assign.js");
}
if (!Object.values) {
  require("core-js/fn/object/values.js");
}

require("regenerator-runtime/runtime");

module.exports = {
  get AppClient() {
    let AppClient = require("uu_appg01_core-appclient");
    // support legacy API
    AppClient.Client = AppClient.AppClient;
    return AppClient;
  },
  get Uri() {
    return require("uu_appg01_core-uri");
  },
  get Logging() {
    return require("uu_appg01_core-logging");
  },
  get Util() {
    return require("uu_appg01_core-utils");
  },
  get Utils() {
    return require("uu_appg01_core-utils");
  },
  get Perflog() {
    return require("uu_appg01_core-perflog");
  },
  get Authentication() {
    return require("uu_appg01_core-authentication");
  },
  get Error() {
    let Utils = require("uu_appg01_core-utils");
    return { Error: Utils.BaseError };
  }
};
