// uu_appg01_core-appclient/src/error/remote-error.js
"use strict";

const { BaseError } = require("uu_appg01_core-utils");

class RemoteError extends BaseError {
  constructor(message = null, response = null) {
    if (typeof message === "string") {
      super(message);
      this.paramMap = {};
      this.dtoOut = {};
    } else if (message.uuAppErrorMap) {
      let keys = Object.keys(message.uuAppErrorMap);
      if (keys.length === 0) {
        // Error in response format (serialized error is missing)
        super(JSON.stringify(message));
        this.paramMap = {};
        this.dtoOut = {};
      } else {
        for (let key of keys) {
          let error = message.uuAppErrorMap[key];
          if (error.type === "error") {
            super(error.message);
            this.code = key;
            if (error.id) {
              this.id = error.id;
            }
            this.paramMap = error.paramMap;
            break;
          }
        }
        this.dtoOut = message;
      }
    } else {
      super(JSON.stringify(message));
      this.paramMap = {};
      this.dtoOut = {};
    }
    if (response) {
      this.response = response;
      this.status = response.status;
    }
  }
}

module.exports = RemoteError;
