// uu_contentkitg01/dist-esm/bricks/inject-block.js
import Block from "./block.js";
import UU5 from "uu5g04";
import "uu5g04-bricks";

import Cfg from "../core/_config.js";

const injectBlock = type => {
  let InjectBlock = UU5.Common.VisualComponent.create({
    //@@viewOn:mixins
    mixins: [UU5.Common.BaseMixin, UU5.Common.SectionMixin],
    //@@viewOff:mixins

    //@@viewOn:statics
    statics: {
      tagName: `${Cfg.CSS} ${Cfg.tag("Bricks.Block" + type.charAt(0).toUpperCase() + type.slice(1))}`,
      opt: {
        hoc: true
      }
    },
    //@@viewOff:statics

    //@@viewOn:propTypes
    propTypes: {
      content: UU5.PropTypes.oneOfType([UU5.PropTypes.array, UU5.PropTypes.string])
    },
    //@@viewOff:propTypes

    //@@viewOn:standardComponentLifeCycle
    //@@viewOff:standardComponentLifeCycle

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:overridingMethods
    //@@viewOff:overridingMethods

    //@@viewOn:componentSpecificHelpers
    //@@viewOff:componentSpecificHelpers

    //@@viewOn:render
    render() {
      const content = this.props.content || this.props.children;
      return <Block {...this.props} content={content} type={type} />;
    }
    //@@viewOff:render
  });

  return InjectBlock;
};

export default injectBlock;
