// uu_contentkitg01/dist-esm/links/link-editable.js
//@@viewOn:imports
import UU5, { createVisualComponent } from "uu5g04";
// import "uu5g04-bricks-editable"; // not because of lazy external dependencies

import Cfg from "../core/_config.js";
import Lsi from "../config/lsi.js";

import LinkView from "./link-view.js";
import LinkInput from "./internal/link-input.js";
import LinkTypeInput from "./internal/link-type-input.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Links.LinkEditable")
};

const componentPropsForm = [
  {
    name: Lsi.componentProperties,
    setup: [
      LinkInput,
      {
        name: "target",
        type: "switchSelector",
        label: Lsi.target,
        getProps: () => ({
          items: [{ value: "_self", content: Lsi.self }, { value: "_blank", content: Lsi.blank }]
        })
      }
    ]
  },
  {
    name: Lsi.display,
    setup: [LinkTypeInput]
  },
  {
    name: Lsi.advance,
    setup: [
      {
        name: "fragment",
        type: "text",
        label: "Element ID",
        getProps: () => ({
          message: Lsi.help.fragment
        })
      }
    ]
  }
];

//@@viewOff:statics
const LinkEditable = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    props: UU5.PropTypes.shape(LinkView.propTypes).isRequired,
    onClose: UU5.PropTypes.func.isRequired,
    componentName: UU5.PropTypes.string.isRequired
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  //@@viewOff:defaultProps

  getPropsToSave() {
    return this._modalRef ? this._modalRef.getPropsToSave() : undefined;
  },

  _ref(modal) {
    this._modalRef = modal;
  },

  render() {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.BricksEditable.Modal
        shown
        header={Lsi.editation.link}
        onClose={this.props.onClose}
        componentProps={this.props.props}
        componentName={this.props.componentName}
        componentPropsForm={componentPropsForm}
        menuWidth="250px"
        ref_={this._ref}
      />
    );
    //@@viewOff:render
  }
});

export default LinkEditable;
