// uu_contentkitg01/dist-esm/links/link.js
//@@viewOn:imports
import UU5, { createComponent } from "uu5g04";

import Cfg from "../core/_config.js";
import LinkView from "./link-view.js";

const LinkEditable = UU5.Common.Component.lazy(async () => {
  if (typeof SystemJS !== undefined) {
    await SystemJS.import("uu5g04-forms");
    await SystemJS.import("uu5g04-bricks-editable");
  }
  return import("./link-editable.js");
});
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Links.Link"),
  nestingLevel: UU5.Environment.getNestingLevelList("smallBox", "inline")
};
//@@viewOff:statics
const { id, className, style, mainAttrs, noIndex, ...defaultProps } = LinkView.defaultProps;

export const Link = createComponent({
  ...STATICS,

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.EditableMixin],
  //@@viewOff:mixins

  //@@viewOn:propTypes
  propTypes: LinkView.propTypes,
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: defaultProps,
  //@@viewOff:defaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  onBeforeForceEndEditation_() {
    return this._editRef ? this._editRef.getPropsToSave() : undefined;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _ref(ref) {
    this._editRef = ref;
  },
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <>
        {this.isInlineEdited() && (
          <UU5.Common.Suspense fallback={this.getEditingLoading()}>
            <LinkEditable
              props={this.getEditablePropsValues(Object.keys(defaultProps))}
              onClose={this.endEditation}
              componentName={STATICS.displayName}
              ref={this._ref}
            />
          </UU5.Common.Suspense>
        )}
        <LinkView {...this.props} />
      </>
    );
  }
  //@@viewOff:render
});

Link.View = LinkView;
Link.Editable = LinkEditable;

export default Link;
