// uu_contentkitg01/dist-esm/images/image.js
//@@viewOn:imports
import UU5, { createComponent } from "uu5g04";

import Cfg from "../core/_config.js";

import ImageView from "./image-view.js";
import LazyLibraries from "../core/lazy-libraries";
const ImageEditable = UU5.Common.Component.lazy(async () => {
  if (typeof SystemJS !== undefined) {
    await SystemJS.import("uu5g04-forms");
    await SystemJS.import("uu5g04-bricks-editable");

    /* eslint-disable no-undef */
    LazyLibraries["uu_plus4u5filesg01"] = await SystemJS.import("uu_plus4u5filesg01");
    /* eslint-enable no-undef */
  }
  return import("./image-editable.js");
});
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Images.Image")
};
//@@viewOff:statics
const { id, className, style, mainAttrs, noIndex, ...defaultProps } = ImageView.defaultProps;

const Image = createComponent({
  ...STATICS,

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.EditableMixin],
  //@@viewOff:mixins

  //@@viewOn:propTypes
  propTypes: ImageView.propTypes,
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: defaultProps,
  //@@viewOff:defaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  onBeforeForceEndEditation_() {
    return this._editRef ? this._editRef.getPropsToSave() : undefined;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _ref(ref) {
    this._editRef = ref;
  },
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <>
        {this.isInlineEdited() && (
          <UU5.Common.Suspense fallback={this.getEditingLoading()}>
            <ImageEditable
              props={this.getEditablePropsValues(Object.keys(defaultProps))}
              onClose={this.endEditation}
              componentName={STATICS.displayName}
              ref={this._ref}
            />
          </UU5.Common.Suspense>
        )}
        <ImageView {...this.props} />
      </>
    );
  }
  //@@viewOff:render
});

export default Image;
