// uu_contentkitg01/dist-esm/images/image-column-editable.js
//@@viewOn:imports
import UU5, { createHoc, createVisualComponent } from "uu5g04";

import Cfg from "../core/_config.js";
import Lsi from "../config/lsi.js";

import ImageRowView from "./image-column-view.js";
import ImageListEditable from "./internal/image-list-editable.js";
import InfoMessage from "../editable/info-message.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Images.ImageColumnEditable")
};

const HocInfoMessage = createHoc({
  getProps(props) {
    return {
      ...props,
      message: Lsi.help.imageColumn
    };
  },
  component: InfoMessage
});

const componentPropsForm = [
  {
    name: Lsi.componentProperties,
    setup: [
      {
        name: "effect",
        type: "switchSelector",
        label: Lsi.effectImage,
        getProps: () => ({
          items: Cfg.EFFECT
        })
      },
      {
        name: "alignment",
        type: "switchSelector",
        label: Lsi.alignment,
        getProps: () => ({
          items: [
            { value: "block", content: Lsi.block },
            { value: "top", content: Lsi.top },
            { value: "center", content: Lsi.center },
            { value: "bottom", content: Lsi.bottom }
          ]
        })
      },
      [
        {
          name: "minWidth",
          type: "number",
          label: Lsi.minWidth,
          getProps: () => ({
            buttonHidden: true
          })
        },
        {
          name: "maxWidth",
          type: "number",
          label: Lsi.maxWidth,
          getProps: () => ({
            buttonHidden: true
          })
        }
      ],
      [
        {
          name: "height",
          type: "number",
          label: Lsi.columnHeight,
          getProps: () => ({
            buttonHidden: true
          })
        },
        {
          name: "randomOrder",
          type: "bool",
          label: Lsi.randomOrder
        }
      ],
      HocInfoMessage
    ]
  },
  {
    name: Lsi.advance,
    setup: [
      {
        name: "lightbox",
        type: "switchSelector",
        label: "Lightbox",
        getProps: () => ({
          items: [
            { value: "local", content: Lsi.local },
            { value: "global", content: Lsi.global },
            { value: "none", content: Lsi.noneLightbox }
          ],
          message: Lsi.help.lightboxGallery
        })
      }
    ]
  }
];

//@@viewOff:statics
const ImageColumnEditable = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    props: UU5.PropTypes.shape(ImageRowView.propTypes).isRequired,
    onClose: UU5.PropTypes.func.isRequired,
    saveEditation: UU5.PropTypes.func.isRequired,
    componentName: UU5.PropTypes.string.isRequired
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  //@@viewOff:defaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  getPropsToSave() {
    return this._editRef ? this._editRef.getPropsToSave() : undefined;
  },
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _ref(ref) {
    this._editRef = ref;
  },
  //@@viewOff:private

  render() {
    //@@viewOn:render
    return (
      <ImageListEditable
        ref={this._ref}
        props={this.props.props}
        onClose={this.props.onClose}
        saveEditation={this.props.saveEditation}
        componentName={this.props.componentName}
        header={Lsi.editation.imageColumn}
        componentPropsForm={componentPropsForm}
      />
    );
    //@@viewOff:render
  }
});

export default ImageColumnEditable;
