// uu5richtextg01/dist-esm/richtext/editor-lazy.js
//@@viewOn:imports
import UU5 from "uu5g04";

import { EDITOR_BUTTONS } from "./bricks/defaults.js";

import { Suspense, lazy } from "./internal/suspense.js";

const Editor = lazy(() => import("./editor.js"));
//@@viewOff:imports

export const EditorLazy = UU5.Common.VisualComponent.create({
  displayName: "UU5.RichText.EditorLazy",
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    getDerivedStateFromError(error) {
      return { error };
    },
    opt: {
      hoc: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return { error: null };
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { error } = this.state;
    return error ? (
      <UU5.Common.Error error={error} content="UU5.RichText.Editor failed to render." moreInfo />
    ) : (
      <Suspense fallback={<UU5.Bricks.Loading />}>
        <Editor {...this.props} />
      </Suspense>
    );
  }
  //@@viewOff:render
});

EditorLazy.isMultiline = () => false;
EditorLazy.buttons = [...EDITOR_BUTTONS];

export default EditorLazy;
