// uu5richtextg01/dist-esm/richtext/editor-block-lazy.js
//@@viewOn:imports
import UU5 from "uu5g04";

import { Suspense, lazy } from "./internal/suspense.js";

const EditorBlock = lazy(() => import("./editor-block.js"));
//@@viewOff:imports

export const EditorBlockLazy = UU5.Common.VisualComponent.create({
  displayName: "UU5.RichText.BlockLazy",
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "UU5.RichText.BlockLazy",
    getDerivedStateFromError(error) {
      return { error };
    },
    opt: {
      hoc: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return { error: null };
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { error } = this.state;
    return error ? (
      <UU5.Common.Error error={error} content="EditorBlock failed to render." moreInfo />
    ) : (
      <Suspense fallback={<UU5.Bricks.Loading />}>
        <EditorBlock {...this.props} />
      </Suspense>
    );
  }
  //@@viewOff:render
});
export default EditorBlockLazy;
