// uu5richtextg01/dist-esm/richtext/expandable-editor-input-lazy.js
//@@viewOn:imports
import UU5 from "uu5g04";

import { Suspense, lazy } from "./internal/suspense.js";

const ExpandableEditorInputWrapper = lazy(() => import("./expandable-editor-input-wrapper.js"));
//@@viewOff:imports

export const ExpandableEditorInputLazy = UU5.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "UU5.RichText.ExpandableEditorInputLazy",
    getDerivedStateFromError(error) {
      return { error };
    },
    opt: {
      hoc: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:reactLifeCycle
  getInitialState() {
    return { error: null };
  },
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { error } = this.state;
    return error ? (
      <UU5.Common.Error error={error} content="UU5.RichText.Editor failed to render." moreInfo />
    ) : (
      <Suspense fallback={<UU5.Bricks.Loading />}>
        <ExpandableEditorInputWrapper {...this.props} />
      </Suspense>
    );
  }
  //@@viewOff:render
});
export default ExpandableEditorInputLazy;
