// uu5richtextg01/dist-esm/richtext/html-editor-lazy.js
//@@viewOn:imports
import UU5 from "uu5g04";

import { Suspense, lazy } from "./internal/suspense.js";
import { HTML_EDITOR_BUTTONS } from "./bricks/defaults.js";

const HtmlEditor = lazy(() => import("./html-editor"));
//@@viewOff:imports

export const HtmlEditorLazy = UU5.Common.VisualComponent.create({
  displayName: "UU5.RichText.HtmlEditorLazy",
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins

  //@@viewOn:statics
  statics: {
    tagName: "UU5.RichText.HtmlEditorLazy",
    getDerivedStateFromError(error) {
      return { error };
    },
    opt: {
      hoc: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return { error: null };
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { error } = this.state;
    return error ? (
      <UU5.Common.Error error={error} content="UU5.RichText.Area failed to render." moreInfo />
    ) : (
      <Suspense fallback={<UU5.Bricks.Loading />}>
        <HtmlEditor {...this.props} />
      </Suspense>
    );
  }
  //@@viewOff:render
});

HtmlEditorLazy.buttons = [...HTML_EDITOR_BUTTONS];

export default HtmlEditorLazy;
