// uu5richtextg01/dist-esm/richtext/richtext-lsi.js
export const RichTextLsi = {
  // components
  editor: {
    editorError: {
      cs:
        "Při běhu RichText editoru došlo k nečekané chybě. Aby bylo zabráněno ztrátě dat, poslední změna byla vrácena zpět.",
      en:
        "An unexpected error occurred while running the RichText editor. To prevent data loss, the last change was rolled back."
    },
    uu5stringToEditorError: {
      cs: "Nepodařilo se převést zadanou hodnotu do stavu editoru.",
      en: "Cannot parse value to editor state."
    },
    editorStateToUu5stringError: {
      cs: "Nepodařilo se převést stav editoru na data.",
      en: "Cannot get value from editor state."
    },
    htmlParseError: {
      cs: "Nepodařilo se zpracovat vložené HTML.",
      en: "Cannot parse pasted HTML."
    },
    mandatoryMethodParams: {
      cs: "Volání metody neobsahuje všechny povinné parametry.",
      en: "Method call does not have all mandatory parameters."
    }
  },
  editorInput: {
    invalidUu5String: {
      cs: "Zadaný uu5string není validní.",
      en: "Invalid uu5string.",
      sk: "Zadaný uu5string nie je validný."
    }
  },
  block: {
    saving: {
      cs: "Ukládám...",
      en: "Saving..."
    },
    endEditation: {
      cs: "Ukončit úpravy",
      en: "Stop editing"
    },
  },
  // html plugins
  htmlPlugins: {
    image: {
      label: {
        cs: "Obrázek",
        en: "Image"
      }
    },
    link: {
      label: {
        cs: "Odkaz",
        en: "Link"
      },
      editationFormHeader: {
        cs: "Editace odkazu",
        en: "Edit Link"
      },
      addFormHeader: {
        cs: "Vložit odkaz",
        en: "Insert Link"
      }
    },
    linkForm: {
      additionalProps: {
        cs: "Doplňující vlastnosti",
        en: "Additional Props"
      },
      hrefAttributeMessage: {
        cs: "Uložení prázdné hodnoty způsobí ostranění komponenty.",
        en: "Submit empty value will cause to remove link component."
      }
    }
  },
  // plugins
  plugins: {
    blockIndent: {
      increaseIndentLabel: {
        cs: "Zvětšit odsazení",
        en: "Increase Indent"
      },
      decreaseIndentLabel: {
        cs: "Zmenšit odsazení",
        en: "Decrease Indent"
      }
    },
    blockquote: {
      label: {
        cs: "Citace",
        en: "Blockquote"
      }
    },
    bold: {
      label: {
        cs: "Tučné",
        en: "Bold"
      }
    },
    button: {
      label: {
        cs: "Tlačítko",
        en: "Button"
      }
    },
    code: {
      label: {
        cs: "Kód",
        en: "Code"
      }
    },
    color: {
      textColorLabel: {
        cs: "Barva textu",
        en: "Text Color"
      },
      backgroundColorLabel: {
        cs: "Barva pozadí",
        en: "Background Color"
      },
      defaultColor: {
        cs: "Výchozí barva",
        en: "Default Color"
      }
    },
    fontFamily: {
      label: {
        cs: "Písmo",
        en: "Font Family"
      }
    },
    fontSize: {
      default: {
        cs: " (výchozí)",
        en: " (default)"
      },
      label: {
        cs: "Velikost písma",
        en: "Font Size"
      }
    },
    icon: {
      label: {
        cs: "Ikona",
        en: "Icon"
      }
    },
    index: {
      subscriptLabel: {
        cs: "Dolní index",
        en: "Subscript"
      },
      superscriptLabel: {
        cs: "Horní index",
        en: "Superscript"
      }
    },
    italic: {
      label: {
        cs: "Kurzíva",
        en: "Italic"
      }
    },
    link: {
      label: {
        cs: "Odkaz",
        en: "Link"
      },
      editationFormHeader: {
        cs: "Editace odkazu",
        en: "Edit Link"
      },
      addFormHeader: {
        cs: "Vložit odkaz",
        en: "Insert Link"
      }
    },
    linkForm: {
      additionalProps: {
        cs: "Doplňující vlastnosti",
        en: "Additional Props"
      },
      hrefAttributeMessage: {
        cs: "Uložení prázdné hodnoty způsobí ostranění komponenty.",
        en: "Submit empty value will cause to remove link component."
      }
    },
    list: {
      orderedListLabel: {
        cs: "Číslovaný seznam",
        en: "Ordered List"
      },
      unorderedListLabel: {
        cs: "Odrážkový seznam",
        en: "Unordered List"
      }
    },
    p: {
      label: {
        cs: "Odstavec",
        en: "Paragraph"
      }
    },
    removeFormat: {
      label: {
        cs: "Odstranit formátování",
        en: "Remove Format"
      }
    },
    strike: {
      label: {
        cs: "Přeškrtnutí",
        en: "Strike"
      }
    },
    textAlign: {
      label: {
        center: {
          cs: "Zarovnat na střed",
          en: "Align Center"
        },
        left: {
          cs: "Zarovnat doleva",
          en: "Align Left"
        },
        right: {
          cs: "Zarovnat doprava",
          en: "Align Right"
        },
        all: {
          cs: "Zarovnání textu",
          en: "Text Align"
        }
      }
    },
    underline: {
      label: {
        cs: "Podtržení",
        en: "Underline"
      }
    }
  }
};

export default RichTextLsi;
