// uu5richtextg01/dist-esm/richtext/toolbar-lazy.js
import UU5 from "uu5g04";

import { Suspense, lazy } from "./internal/suspense.js";

const Toolbar = lazy(() => import("./toolbar.js"));

export const ToolbarLazy = UU5.Common.VisualComponent.create({
  displayName: "UU5.RichText.ToolbarLazy",
  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin],
  //@@viewOff:mixins
  statics: {
    tagName: "UU5.RichText.ToolbarLazy",
    getDerivedStateFromError(error) {
      return { error };
    },
    opt: {
      hoc: true
    }
  },
  //@@viewOff:statics

  //@@viewOn:propTypes
  //@@viewOff:propTypes

  //@@viewOn:getDefaultProps
  //@@viewOff:getDefaultProps

  //@@viewOn:standardComponentLifeCycle
  getInitialState() {
    return { error: null };
  },
  //@@viewOff:standardComponentLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  //@@viewOff:overridingMethods

  //@@viewOn:componentSpecificHelpers
  //@@viewOff:componentSpecificHelpers

  //@@viewOn:render
  render() {
    let { error } = this.state;
    return error ? (
      <UU5.Common.Error error={error} content="UU5.RichText.Toolbar failed to render." moreInfo />
    ) : (
      <Suspense fallback={<div />}>
        <Toolbar {...this.props} />
      </Suspense>
    );
  }
  //@@viewOff:render
});
export default ToolbarLazy;
