// uu_appg01_devkit/src/scripts/uu_cloud/task-uu-cloud-logstore-export.js
/* eslint-disable no-redeclare */
const CmdHelper = require("uu_appg01_devkit-common/src/scripts/misc/cmd-helper.js");
const AppClient = require("uu_appg01_devkit-common/src/scripts/misc/app-client.js");

const DEFAULT_CMD_BASE_PATH = "uu-logstore/Log/getRecordList/exec";

const HEADERS = {
  Accept: "application/json",
  "Content-type": "application/json"
};

class uuLogStoreExportTask {
  constructor(cloudConfig, params) {
    this.config = cloudConfig.uuLogStoreConfig();
    this.appClient = new AppClient(this.config.oidcToken);
    this.params = params;
    this.params.appDeploymentUri = this.config.appDeploymentUri;
    this.params.env = this.config.env;
  }

  async process() {
    let logs = await this._getExport();

    const saveAndFormatLogs = require("uu_appg01_devkit-common/src/scripts/save-and-format-logs.js");
    saveAndFormatLogs(logs, this.params);
  }

  async _getExport() {
    console.log("Exporting logs for app deployment " + this.config.appDeploymentUri);

    let from = this.params.from;
    if (!from) {
      from = new Date();
      from.setHours(from.getHours() - 1);
      from = from.toISOString();
      this.params.from = from;
    }

    let to = this.params.to;
    if (!to) {
      to = new Date();
      to = to.toISOString();
      this.params.to = to;
    }

    const processedIds = new Set();
    let exportCmdUri = this._buildExportCmdUri(this.config.appDeploymentUri);
    let result;
    let firstLoop = true;

    while (true) {
      let query = new Map();
      var maxPageSize = 1000;
      query.set("from", from);
      query.set("to", to);

      //Testing, validating and saving all logstore parameters
      if (["FATAL", "ERROR", "WARNING", "INFO", "DEBUG"].includes(this.params.logLevel)) {
        query.set("logLevel", this.params.logLevel);
      }
      if (typeof this.params.recordType === "string") {
        query.set("recordType", this.params.recordType);
      }
      if (this.params.runtimeStackCode) {
        query.set("runtimeStackCode", this.params.runtimeStackCode);
      }
      if (typeof this.params.nodeImageName === "string") {
        query.set("nodeImageName", this.params.nodeImageName);
      }
      if (typeof this.params.nodeName === "string") {
        query.set("nodeName", this.params.nodeName);
      }
      if (typeof this.params.hostName === "string") {
        query.set("hostName", this.params.hostName);
      }
      if (typeof this.params.correlationId === "string") {
        query.set("correlationId", this.params.correlationId);
      }
      if (typeof this.params.traceId === "string") {
        query.set("traceId", this.params.traceId);
      }
      if (/^([0-9]\.[0-9]\.[0-9]\-?.*)$/.test(this.params.appVersion)) {
        query.set("appVersion", this.params.appVersion);
      }
      if (typeof this.params.pageSize === "number") {
        var maxPageSize = Math.trunc(this.params.pageSize);
      }

      if (firstLoop == true) {
        console.log("This is the query being used:", Array.from(query));
        firstLoop = false;
      } else {
        console.log("> Processing query: from=" + from + ", to=" + to);
      }
      let partialResult = (await this.appClient.exchange(
        CmdHelper.buildCmd2Url(exportCmdUri, this.config.appDeploymentUri, query),
        "get",
        null,
        HEADERS
      )).body;
      partialResult = JSON.parse(partialResult);

      // remove duplicates
      partialResult.pageEntries = partialResult.pageEntries.filter(item => {
        return processedIds.has(item.id) ? false : processedIds.add(item.id);
      });
      partialResult.totalSize = partialResult.pageEntries.length;

      if (result) {
        result.totalSize += partialResult.totalSize;
        result.pageEntries = result.pageEntries.concat(partialResult.pageEntries);
      } else {
        result = partialResult;
      }

      if (partialResult.totalSize < maxPageSize) {
        break;
      } else {
        to = partialResult.pageEntries[partialResult.pageEntries.length - 1].time;
        if (new Date(to) < new Date(from)) {
          break;
        }
      }
    }

    return result;
  }

  /**
   * Parses territory code or id from appDeploymentUri and creates uuLogStoreExport uri with it.
   */
  _buildExportCmdUri(appDeploymentUri) {
    let terrMatch = appDeploymentUri.match(/ues:([^:\[]*)(\[([^:]+)\])?:/);
    if (terrMatch == null) {
      throw new Error("Wrong format of appDeploymentUri: " + appDeploymentUri);
    }

    let territory_part = terrMatch[3] != null ? terrMatch[3] : terrMatch[1];

    return `${this.appClient.logstoreBaseUri}/${territory_part}/${DEFAULT_CMD_BASE_PATH}`;
  }
}

module.exports = uuLogStoreExportTask;
