// uu_appg01_devkit/src/scripts/start.nodejs-app.js
const shell = require("shelljs");
const path = require("path");
const fs = require("fs");
const os = require("os");
const net = require("net");
const child_process = require("child_process");
const Pack = require("uu_appg01_devkit-common/src/tools/package.js");
const helpers = require("../tools/helpers.js");
const configLoader = require("../tools/simple-config-loader.js");
const connectionHelper = require("../tools/connection-helper.js");
const Muri = require("muri");

let TestMongo = null;
try {
  TestMongo = require("uu_appg01_datastore-test").TestMongo;
} catch (e) {
  console.warn("Library uu_appg01_datastore-test is not installed, embedded database will not be started.", e);
}

module.exports = class Start {

  constructor(config) {
    this.config = config || {};
  }

  async process() {
    // Ensure project is built before starting
    // Build server itself
    shell.exec("npm run --loglevel=error build -- --no-build-client " + this.config.toCommandLineArgs().join(" ")); // TODO Escaping.
    Pack.getSingletonSync("package.json").loadSync(); // reload as version might have changed

    // Trigger rebuilding of client project
    let serverDir = process.cwd();
    let buildCheckFile = path.join(serverDir, "public", ".finished");
    let clientDir = serverDir.match(/-server$/) ? path.join("..", path.basename(serverDir).replace(/-server$/, "-hi")) : null;
    if (clientDir && !fs.existsSync(clientDir)) {
      clientDir = path.join("..", path.basename(serverDir).replace(/-server$/, "-client"));
    }
    let hasClient = clientDir && fs.existsSync(clientDir) && fs.existsSync(path.join(clientDir, "package.json"));
    if (hasClient) {
      console.log("Building client project (with automatic rebuild on changes).");
      if (fs.existsSync(buildCheckFile)) {
        // Delete file used for checking of build finish, so that server is not started
        // before build process is started (causing delete of public during server start)
        fs.unlinkSync(buildCheckFile);
      }
      child_process.spawn("npm", ["run", "build", "--", "--watch", "--mode=development"],
        {stdio: ["ignore", "ignore", "inherit"], shell: true, cwd: path.resolve(clientDir)}
      );
    }

    // Load project configuration
    let runtimeEnvironment = this.config.environment || "development,production";
    console.log(`\nLoading configuration for \"${runtimeEnvironment}\" runtime environment`);
    process.env["NODE_ENV"] = runtimeEnvironment;
    let loadedConfig = configLoader.loadConfiguration(runtimeEnvironment);

    if (TestMongo) {
      // Process database configurations
      let dataStoreMap = loadedConfig["uuSubAppDataStoreMap"] || {};
      let projectName = path.basename(path.resolve(".")).replace(/\-[a-zA-Z0-9]+$/, "");
      let dbBaseName = projectName.replace(/(_\w)/g, function (match) {
        return match[1].toUpperCase();
      });
      let embeddedDataStores = {};
      for (let dataStore in dataStoreMap) {
        let connectionUri = dataStoreMap[dataStore];

        let parsedUri = null;
        try {
          parsedUri = Muri(connectionUri);
        } catch (e) {}

        if (parsedUri) {
          let host = parsedUri.hosts[0].host;
          let port = parsedUri.hosts[0].port;
          if (parsedUri.db === "admin") parsedUri.db = null;
          let dbName = parsedUri.db || (`/${dbBaseName}${dataStore.replace(/\\b\\w/g, l => l.toUpperCase())}`).substr(1);

          if (!(await connectionHelper.isServiceListening(port, host))) {
            embeddedDataStores[dataStore] = dbName;
          }
        } else {
          let dbName = `${dbBaseName}${dataStore.replace(/\\b\\w/g, l => l.toUpperCase())}`;
          embeddedDataStores[dataStore] = dbName;
        }
      }

      // Start embedded database if necessary
      if (Object.keys(embeddedDataStores).length != 0) {
        console.log(`\nStarting embedded database for datastores [${Object.keys(embeddedDataStores)}]...`);
        let targetPath = path.join(path.resolve("."), 'target');
        if (!fs.existsSync(targetPath)) {
          fs.mkdirSync(targetPath);
        }
        let storagePath = path.join(path.resolve("."), 'target', 'db_data');
        if (!fs.existsSync(storagePath)) {
          fs.mkdirSync(storagePath);
        }
        let embeddedDb = new TestMongo({
          db: "admin",
          port: await connectionHelper.findFreePort(27017),
          dbPath: storagePath,
          storageEngine: "wiredTiger"
        });
        await embeddedDb.start();
        let dbPort = embeddedDb.getPort();
        for (let dataStore of Object.keys(embeddedDataStores)) {
          embeddedDataStores[dataStore] = `mongodb://127.0.0.1:${dbPort}/${embeddedDataStores[dataStore]}`;
        }
        dataStoreMap = Object.assign(dataStoreMap, embeddedDataStores);
        process.env["SERVER_CFG"] = JSON.stringify({uuSubAppDataStoreMap: dataStoreMap});
        console.log(`Embedded database running at mongodb://127.0.0.1:${dbPort}`);
      }
    }

    // Start application server
    let serverPort = loadedConfig["port"] || (await connectionHelper.findFreePort(6221));
    let pckg = Pack.getSingletonSync("package.json").get();
    let baseName = pckg.name.split("-").slice(0, -1).join("-");
    let appRoute = baseName.toLowerCase().replace(/[\-_]/g, '-');
    process.env["PORT"] = serverPort;
    let startServer = () => {
      console.log(`\nStarting application server with automatic restart on changes...\n`);
      let nodemon = helpers.npmScriptBinFile("nodemon");
      let watchedSources = ["--watch", "app.js", "--watch", "app", "--watch", "src"];
      child_process.spawn(nodemon, ["app.js", ...watchedSources, ...(this.config._commandLineArgs || [])],
        {stdio: ["inherit", "inherit", "inherit"], shell: true}
      );
      this._waitForServer(serverPort, () => {
        console.log(`\nApplication URI: http://localhost:${serverPort}/${appRoute}/${loadedConfig.tid ? `${loadedConfig.tid}-` : ""}${loadedConfig.asid}\n`);
      });
    };
    // Start either immediately or wait for client project rebuild
    if (hasClient && !fs.existsSync(buildCheckFile)) {
      console.log("\nWaiting for finish of client build...");
      this._waitForFile(buildCheckFile, startServer);
    } else {
      startServer();
    }
  }

  _waitForFile(path, callback) {
    let waitOrRun = (err) => {
      if (err) {
        setTimeout(() => {
          fs.access(path, fs.constants.F_OK, waitOrRun);
        }, 250);
      } else {
        callback();
      }
    };
    fs.access(path, fs.constants.F_OK, waitOrRun);
  }

  _waitForServer(port, callback) {
    let socket = null;
    let waitOrRun = (err) => {
      if (err) {
        setTimeout(() => {
          socket = net.createConnection(port).on("connect", waitOrRun).on("error", waitOrRun);
        }, 250);
      } else {
        socket.end();
        callback();
      }
    };
    socket = net.createConnection(port).on("connect", waitOrRun).on("error", waitOrRun);
  }

};
