// uu_contentkitg01/dist-esm/layout/background-image.js
//@@viewOn:imports
import UU5, { createComponent } from "uu5g04";

import Cfg from "../core/_config.js";

import BackgroundImageView from "./background-image-view.js";
import LazyLibraries from "../core/lazy-libraries";
import ImagePanelView from "../images/image-panel-view";
const BackgroundImageEditable = UU5.Common.Component.lazy(async () => {
  if (typeof SystemJS !== undefined) {
    await SystemJS.import("uu5g04-forms");
    await SystemJS.import("uu5g04-bricks-editable");

    /* eslint-disable no-undef */
    LazyLibraries["uu_plus4u5filesg01"] = await SystemJS.import("uu_plus4u5filesg01");
    /* eslint-enable no-undef */
  }
  return import("./background-image-editable.js")
});
//@@viewOff:imports

//@@viewOn:statics
const editMode = {
  name: { en: "Background image", cs: "Obrázek na pozadí" },
  backgroundColor: "rgba(0,0,0,.2)",
  color: "rgba(0,0,0,.87)",
  highlightColor: "#CCCCCC"
};

const STATICS = {
  displayName: Cfg.tag("Layout.BackgroundImage"),
  editMode
};
//@@viewOff:statics
// const { id, className, style, mainAttrs, noIndex, content, ...defaultProps } = BackgroundImageView.defaultProps;
const { id, mainAttrs, noIndex, ...editableProps } = BackgroundImageView.defaultProps;
const { className, style, ...defaultProps } = editableProps;

const BackgroundImage = createComponent({
  ...STATICS,

  statics: {
    editMode
  },

  //@@viewOn:mixins
  mixins: [UU5.Common.BaseMixin, UU5.Common.EditableMixin, UU5.Common.ContentMixin],
  //@@viewOff:mixins

  //@@viewOn:propTypes
  propTypes: BackgroundImageView.propTypes,
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: defaultProps,
  //@@viewOff:defaultProps

  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle

  //@@viewOn:interface
  //@@viewOff:interface

  //@@viewOn:overridingMethods
  onBeforeForceEndEditation_() {
    return this._editRef ? this._editRef.getPropsToSave() : undefined;
  },
  //@@viewOff:overridingMethods

  //@@viewOn:private
  _ref(ref) {
    this._editRef = ref;
  },
  //@@viewOff:private

  //@@viewOn:render
  render() {
    return (
      <div {...this.getMainAttrs()}>
        {this.isInlineEdited() && (
          <UU5.Common.Suspense fallback={this.getEditingLoading()}>
            <BackgroundImageEditable
              props={this.getEditablePropsValues(Object.keys(editableProps))}
              onClose={this.endEditation}
              saveEditation={this.saveEditation}
              componentName={STATICS.displayName}
              component={this}
              ref={this._ref}
            />
          </UU5.Common.Suspense>
        )}
        {(this.isNotInlineEdited ? this.isNotInlineEdited() : !this.isInlineEdited()) && (
          <BackgroundImageView {...this.props} />
        )}
      </div>
    );
  }
  //@@viewOff:render
});

export default BackgroundImage;
