// uu5tilesg01/dist-node/tiles/simple-filter-panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SimpleFilterPanel = void 0;

var _uu5g = _interopRequireDefault(require("uu5g04"));

require("uu5g04-bricks");

require("uu5g04-forms");

var _config = _interopRequireDefault(require("../config/config.js"));

var _lsi = _interopRequireDefault(require("../config/lsi.js"));

var _util = require("./util.js");

require("./simple-filter-panel.less");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

//@@viewOff:imports
var SimpleFilterPanel = _uu5g.default.Common.VisualComponent.create({
  //@@viewOn:mixins
  mixins: [_uu5g.default.Common.BaseMixin, _uu5g.default.Common.ElementaryMixin, _uu5g.default.Common.PureRenderMixin, _uu5g.default.Common.LsiMixin, _uu5g.default.Common.ScreenSizeMixin],
  //@@viewOff:mixins
  //@@viewOn:statics
  statics: {
    tagName: _config.default.TAG + "SimpleFilterPanel",
    classNames: {
      main: _config.default.CSS + "simple-filter-panel",
      formRow: _config.default.CSS + "simple-filter-panel-form-row"
    },
    opt: {
      pureRender: true // avoid re-render from parent

    },
    lsi: _lsi.default.simpleFilterPanel
  },
  //@@viewOff:statics
  //@@viewOn:propTypes
  propTypes: {
    appliedFilters: _uu5g.default.PropTypes.object,
    filters: _uu5g.default.PropTypes.arrayOf(_uu5g.default.PropTypes.shape({
      key: _uu5g.default.PropTypes.string,
      label: _uu5g.default.PropTypes.oneOfType([_uu5g.default.PropTypes.string, _uu5g.default.PropTypes.object]),
      filterFn: _uu5g.default.PropTypes.func
    })),
    addFilter: _uu5g.default.PropTypes.func,
    removeFilter: _uu5g.default.PropTypes.func,
    filterBar: _uu5g.default.PropTypes.object,
    buttonProps: _uu5g.default.PropTypes.oneOfType([_uu5g.default.PropTypes.object, _uu5g.default.PropTypes.func])
  },
  //@@viewOff:propTypes
  //@@viewOn:getDefaultProps
  getDefaultProps: function getDefaultProps() {
    return {
      appliedFilters: undefined,
      filters: undefined,
      addFilter: undefined,
      removeFilter: undefined,
      filterBar: undefined,
      buttonProps: undefined
    };
  },
  //@@viewOff:getDefaultProps
  //@@viewOn:reactLifeCycle
  //@@viewOff:reactLifeCycle
  //@@viewOn:interface
  //@@viewOff:interface
  //@@viewOn:overriding
  //@@viewOff:overriding
  //@@viewOn:private
  _registerForm: function _registerForm(cmp) {
    this._form = cmp;
  },
  _save: function _save() {
    var _this = this;

    var values = this._form.getValues();

    var type = values.type || "";
    var value = values.value || "";

    if (type.length > 0 && value.length > 0) {
      var _this$props$filters$t = this.props.filters[type],
          key = _this$props$filters$t.key,
          label = _this$props$filters$t.label,
          filterFn = _this$props$filters$t.filterFn;
      this.props.addFilter(key, label, value, filterFn, null, function () {
        return _this.props.filterBar.collapse();
      });
    }
  },
  _getOptions: function _getOptions() {
    var _this2 = this;

    return Object.keys(this.props.filters).map(function (idx) {
      return _uu5g.default.Common.Element.create(_uu5g.default.Forms.Select.Option, {
        key: idx,
        value: idx,
        disabled: _this2.props.appliedFilters.hasOwnProperty(_this2.props.filters[idx].key)
      }, _this2.getLsiItem(_this2.props.filters[idx].label));
    });
  },
  //@@viewOff:private
  //@@viewOn:render
  render: function render() {
    return _uu5g.default.Common.Element.create(_uu5g.default.Forms.Form, _extends({}, this.getMainPropsToPass(), {
      ref_: this._registerForm,
      saveOnEnter: true,
      onSave: this._save
    }), _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Row, {
      className: this.getClassName().formRow
    }, _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Column, {
      colWidth: "xs-12 s-6 m-5 l-4 xl-3"
    }, _uu5g.default.Common.Element.create(_uu5g.default.Forms.Select, {
      name: "type",
      labelColWidth: "xs-12 s-11 m-5",
      inputColWidth: "xs-12 s-11 m-7",
      label: this.getLsiValue("filterTypeLabel"),
      placeholder: this.getLsiValue("filterTypePlaceholder")
    }, this._getOptions())), _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Column, {
      colWidth: "xs-12 s-6 m-5 l-4 xl-3"
    }, _uu5g.default.Common.Element.create(_uu5g.default.Forms.Text, {
      name: "value",
      labelColWidth: "xs-12 m-5",
      inputColWidth: "xs-12 m-7",
      label: this.getLsiValue("filterValueLabel"),
      placeholder: this.getLsiValue("filterValuePlaceholder")
    })), this.isS() && _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Column, {
      colWidth: "xs-0 s-6 m-0",
      content: "\xa0"
    }), _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Column, {
      colWidth: "xs-12 s-6 m-2 xl-1"
    }, _uu5g.default.Common.Element.create(_uu5g.default.Bricks.Button, (0, _util.mergeButtonProps)({
      onClick: this._save,
      colorSchema: this.props.colorSchema || "primary"
    }, this.props.buttonProps, "submitFilter"), this.getLsiValue("addButton")))));
  } //@@viewOff:render

});

exports.SimpleFilterPanel = SimpleFilterPanel;
var _default = SimpleFilterPanel;
exports.default = _default;
