// uu_contentkitg01/dist-esm/bricks/internal/color-picker.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../../core/_config.js";
import Lsi from "../../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.ColorPicker")
};
//@@viewOff:statics

const ColorPicker = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Forms.ColorPicker
        name={"bgColor"}
        inputColWidth={{ xs: 12, s: 4 }}
        label={Lsi.bgColor}
        value={props.componentProps.bgColor}
        openToContent
        simplePalette
      />
    );
    //@@viewOff:render
  }
});

export default ColorPicker;
