// uu_contentkitg01/dist-esm/bricks/internal/content-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";
import "uu5richtextg01";

import Cfg from "../../core/_config.js";
import Lsi from "../../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.ContentInput")
};
//@@viewOff:statics

const ContentInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.RichText.EditorInput
        {...UU5.Common.VisualComponent.getAttrs(props)}
        inputColWidth="xs-12"
        label={Lsi.richText}
        value={props.componentProps.content}
        onBlur={({ value }) => props.onChangeProps({ content: value })}
      />
    );
    //@@viewOff:render
  }
});

export default ContentInput;
