// uu_contentkitg01/dist-esm/bricks/internal/slider-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../../core/_config.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.SliderInput")
};
//@@viewOff:statics

const SliderInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func,
    disable: UU5.PropTypes.bool,
    propName: UU5.PropTypes.string,
    label: UU5.PropTypes.object,
    min: UU5.PropTypes.number,
    max: UU5.PropTypes.number,
    step: UU5.PropTypes.number
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined,
    disabled: false,
    propName: null,
    label: null,
    min: 0,
    max: 1,
    step: 0.01
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <UU5.Forms.Slider
        {...UU5.Common.VisualComponent.getAttrs(props)}
        label={props.label}
        step={props.step}
        min={props.min}
        max={props.max}
        value={props.componentProps[props.propName]}
        onChange={opt => props.onChangeProps({ [props.propName]: +opt.value })}
      />
    );
    //@@viewOff:render
  }
});

export default SliderInput;
