// uu_contentkitg01/dist-esm/links/internal/content-disposition-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../../core/_config.js";
import Lsi from "../../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Links.ContentDispositionInput")
};
//@@viewOff:statics

const ContentDispositionInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <>
        <UU5.Forms.SwitchSelector
          label="Content disposition"
          value={props.componentProps.contentDisposition}
          onChange={opt =>
            props.onChangeProps({ contentDisposition: opt.value, target: opt.value === "inline" ? "_blank" : "_self" })
          }
          items={[{ value: "attachment", content: Lsi.attachment }, { value: "inline", content: Lsi.inline }]}
          message={Lsi.help.contentDisposition}
        />
        <UU5.Forms.SwitchSelector
          label={Lsi.target}
          disabled={props.componentProps.contentDisposition === "attachment"}
          value={props.componentProps.target}
          onChange={opt => props.onChangeProps({ target: opt.value })}
          items={[{ value: "_self", content: Lsi.self }, { value: "_blank", content: Lsi.blank }]}
        />
      </>
    );
    //@@viewOff:render
  }
});

export default ContentDispositionInput;
