// uu_appg01_server-client/src/index.js
"use strict";

module.exports = {
  get AppClient() {
    return require("./client.js");
  },
  get RemoteError() {
    return require("uu_appg01_core-appclient").RemoteError;
  },
  get ApplicationError() {
    return require("uu_appg01_core-appclient").ApplicationError;
  },
  get SystemError() {
    return require("uu_appg01_core-appclient").SystemError;
  },
  get Uri() {
    return require("uu_appg01_core-uri").Uri;
  },
  get UriBuilder() {
    return require("uu_appg01_core-uri").UriBuilder;
  },
  get InvalidUriPartError() {
    return require("uu_appg01_core-uri").InvalidUriPartError;
  },
  get AuthenticationService() {
    let { Config } = require("uu_appg01_core-utils");
    let Authn = require("uu_appg01_core-authentication");
    if (!Config.hasKey(Authn.AuthenticationService.DEFAULT_PROVIDER_PARAM)) {
      // Configure default authentication service
      require("uu_appg01_oidc").OidcAuthenticationService;
      Config.set(Authn.AuthenticationService.DEFAULT_PROVIDER_PARAM, "uu_appg01_oidc");
    }
    return Authn.AuthenticationService;
  },
  get AuthenticationError() {
    return require("uu_appg01_core-authentication").AuthenticationError;
  },
  get InvalidCredentials() {
    return require("uu_appg01_core-authentication").InvalidCredentials;
  },
  get UntrustedCredentials() {
    return require("uu_appg01_core-authentication").UntrustedCredentials;
  },
  get OidcAuthenticationService() {
    return require("uu_appg01_oidc").OidcAuthenticationService;
  }
};
