// uu_contentkitg01/dist-esm/layout/internal/image-size-input.js
//@@viewOn:imports
import UU5 from "uu5g04";
import { createVisualComponent } from "uu5g04-hooks";

import Cfg from "../../core/_config.js";
import Lsi from "../../config/lsi.js";
//@@viewOff:imports

//@@viewOn:statics
const STATICS = {
  displayName: Cfg.tag("Bricks.ImageSizeInput")
};
//@@viewOff:statics

const ImageSizeInput = createVisualComponent({
  ...STATICS,

  //@@viewOn:propTypes
  propTypes: {
    componentProps: UU5.PropTypes.object,
    onChangeProps: UU5.PropTypes.func
  },
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {
    componentProps: undefined,
    onChangeProps: undefined
  },
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:hooks
    //@@viewOff:hooks

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:private
    //@@viewOff:private

    //@@viewOn:render
    return (
      <>
        <UU5.Forms.SwitchSelector
          label={Lsi.imageSize}
          value={props.componentProps.imageSize}
          onChange={({ value }) => props.onChangeProps({ imageSize: value })}
          items={[{ value: "cover" }, { value: "original" }]}
        />
        <UU5.Forms.SwitchSelector
          label={Lsi.imageRepeat}
          value={props.componentProps.imageRepeat}
          onChange={({ value }) => props.onChangeProps({ imageRepeat: value })}
          items={[
            { value: "repeat", content: Lsi.repeat },
            { value: "no-repeat", content: Lsi.noRepeat }
          ]}
          disabled={props.componentProps.imageSize === "cover"}
        />
      </>
    );
    //@@viewOff:render
  }
});

export default ImageSizeInput;
